// @geotab/zenith/dist/filters/filtersHelper.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getArrayOfElements = exports.useCombineData = exports.MAX_TOTAL_RESULTS = exports.serializeEntriesToId = exports.getTotalChangedNumber = exports.getNumberOfChangedFilters = exports.createStateObject = exports.getFiltersState = exports.getChildren = exports.prepareDefaultDropdownState = exports.getDropdownStateType = exports.stabilizeChildrenWithId = exports.createPropsHash = exports.createStateHash = void 0;
const react_1 = require("react");
const filtersGroups_1 = require("./components/filtersGroups");
const react_2 = __importDefault(require("react"));
const isArraysEqualDeeply_1 = require("../utils/compareElementsUtils/isArraysEqualDeeply");
const isElementsEqual_1 = require("../utils/compareElementsUtils/isElementsEqual");
const groupsFilter_1 = require("../groupsFilter/groupsFilter");
const filtersChip_1 = require("./components/filtersChip");
const filtersDropdown_1 = require("./components/filtersDropdown");
const filtersSearchList_1 = require("./components/filtersSearchList");
const createStateHash = (state) => {
    if (state === undefined) {
        return "undefined";
    }
    if (state === null) {
        return "null";
    }
    try {
        const jsonString = JSON.stringify(state);
        if (jsonString.length > 100) {
            return `hash:${jsonString.length}:${jsonString.slice(0, 50)}`;
        }
        return jsonString;
    }
    catch (_a) {
        return `[object ${typeof state}]`;
    }
};
exports.createStateHash = createStateHash;
const createPropsHash = (props) => {
    const relevantKeys = Object.keys(props)
        .filter(key => key !== "children")
        .sort();
    return relevantKeys
        .map(key => {
        const value = props[key];
        if (typeof value === "function") {
            return `${key}:fn(${value.name || value.toString().slice(0, 30)})`;
        }
        if (key === "defaultState" || key === "onChange") {
            try {
                return `${key}:${JSON.stringify(value)}`;
            }
            catch (_a) {
                return `${key}:${typeof value}`;
            }
        }
        return `${key}:${typeof value}`;
    })
        .join(",");
};
exports.createPropsHash = createPropsHash;
const stabilizeChildrenWithId = (children) => {
    const childrenArray = react_2.default.Children.toArray(children).filter(Boolean);
    return childrenArray.map((child, index) => {
        if (!react_2.default.isValidElement(child)) {
            return child;
        }
        const childId = child.props.id;
        if (!childId) {
            console.error(`FiltersContainer: Child component at index ${index} is missing required 'id' prop. ` +
                `Component will not be rendered. All filter components must have a unique 'id' for proper performance optimization.`, child);
            return null;
        }
        return react_2.default.cloneElement(child, { key: childId });
    });
};
exports.stabilizeChildrenWithId = stabilizeChildrenWithId;
const getDropdownStateType = (dropdownState) => {
    if (Array.isArray(dropdownState)) {
        return "base";
    }
    if (dropdownState.hasOwnProperty("isChecked") && dropdownState.hasOwnProperty("isAllSelected")) {
        return "fullSelectionWithCheckbox";
    }
    if (dropdownState.hasOwnProperty("isAllSelected")) {
        return "fullSelection";
    }
    return "withCheckbox";
};
exports.getDropdownStateType = getDropdownStateType;
const prepareDefaultDropdownState = (type) => {
    if (type === "fullSelectionWithCheckbox") {
        return { selected: [], isAllSelected: false, isChecked: false };
    }
    if (type === "withCheckbox") {
        return { selected: [], isChecked: false };
    }
    if (type === "fullSelection") {
        return { selected: [], isAllSelected: false };
    }
    return [];
};
exports.prepareDefaultDropdownState = prepareDefaultDropdownState;
const getChildren = (children) => (!children
    ? []
    : (react_1.Children.count(children) === 1 ? [children] : children)).filter(c => !!c);
exports.getChildren = getChildren;
const getFiltersState = (children) => {
    const state = {};
    react_1.Children.forEach(children, (child) => {
        state[child.props.id] = {};
        state[child.props.id].state = child.props.state;
        if (child.type === filtersGroups_1.FiltersGroups) {
            state[child.props.id].defaultState = groupsFilter_1.groupsFilterTotalStateDefault;
        }
        else if (child.type === filtersChip_1.FiltersChip) {
            state[child.props.id].defaultState = false;
        }
        else if (child.type === filtersDropdown_1.FiltersDropdown) {
            state[child.props.id].defaultState = child.props.defaultState || (0, exports.prepareDefaultDropdownState)((0, exports.getDropdownStateType)(child.props.state));
        }
        else {
            state[child.props.id].defaultState = child.props.defaultState;
        }
    });
    return state;
};
exports.getFiltersState = getFiltersState;
const createStateObject = (children, field) => {
    const state = {};
    react_1.Children.forEach(children, (child) => {
        state[child.props.id] = {};
        if (child.type === filtersGroups_1.FiltersGroups && field === "defaultState") {
            state[child.props.id] = groupsFilter_1.groupsFilterTotalStateDefault;
        }
        else if (child.type === filtersChip_1.FiltersChip && field === "defaultState") {
            state[child.props.id] = false;
        }
        else if (child.type === filtersDropdown_1.FiltersDropdown && field === "defaultState") {
            state[child.props.id] = child.props.defaultState || (0, exports.prepareDefaultDropdownState)((0, exports.getDropdownStateType)(child.props.state));
        }
        else {
            state[child.props.id] = child.props[field];
        }
    });
    return state;
};
exports.createStateObject = createStateObject;
function getNumberOfChangedFilters(state) {
    let changes = 0;
    Object.keys(state).forEach(key => {
        const defaultState = state[key].defaultState;
        const currentState = state[key].state;
        if (!state[key].hasOwnProperty("state") || !state[key].hasOwnProperty("defaultState")) {
            changes++;
            return;
        }
        if (Array.isArray(defaultState) && Array.isArray(currentState)) {
            if (!(0, isArraysEqualDeeply_1.isArraysEqualDeeply)(defaultState, currentState)) {
                changes++;
            }
            return;
        }
        if (typeof defaultState === "object" && typeof currentState === "object") {
            if (!(0, isElementsEqual_1.isElementsEqual)(defaultState, currentState)) {
                changes++;
            }
            return;
        }
        if (defaultState !== currentState) {
            changes++;
        }
    });
    return changes;
}
exports.getNumberOfChangedFilters = getNumberOfChangedFilters;
const getTotalChangedNumber = (selectState, filters, isModalCondition) => {
    const stateObject = (0, exports.createStateObject)(filters, "state");
    const filteredFilters = filters.filter(el => typeof el.props.visible === "boolean" ? el.props.visible
        : (el.props.visible === undefined ? true
            : el.props.visible(Array.from(selectState), isModalCondition ? stateObject : undefined)));
    const filteredChildrenState = (0, exports.getFiltersState)(filteredFilters);
    const numberOfChangedFilters = getNumberOfChangedFilters(filteredChildrenState);
    return numberOfChangedFilters;
};
exports.getTotalChangedNumber = getTotalChangedNumber;
const serializeEntriesToId = (entries) => {
    const serializedEntries = entries.map(({ id, name, type }) => `${type || filtersSearchList_1.NO_TYPED_SEARCH_ITEM}:${id}:${name}`);
    return serializedEntries.join("_");
};
exports.serializeEntriesToId = serializeEntriesToId;
exports.MAX_TOTAL_RESULTS = 500;
const useCombineData = (searchTypes) => {
    var _a;
    const combinedData = (0, react_1.useMemo)(() => {
        const finalResults = [];
        const totalRelevance = searchTypes.reduce((sum, item) => sum + (item.relevance !== undefined ? item.relevance : 1.0), 0);
        if (totalRelevance === 0) {
            return [];
        }
        searchTypes.forEach(item => {
            const fetchedData = item.data.data;
            const rawRelevance = item.relevance !== undefined ? item.relevance : 1.0;
            if (fetchedData && Array.isArray(fetchedData)) {
                const effectiveShare = rawRelevance / totalRelevance;
                const limit = Math.round(exports.MAX_TOTAL_RESULTS * effectiveShare);
                const takeCount = Math.min(limit, fetchedData.length);
                const subset = fetchedData
                    .slice(0, takeCount)
                    .map((el) => (Object.assign(Object.assign({}, el), { type: item.id })));
                finalResults.push(...subset);
            }
        });
        return finalResults.slice(0, exports.MAX_TOTAL_RESULTS);
    }, [searchTypes]);
    const isLoading = searchTypes.some(item => item.data.isLoading);
    return {
        data: combinedData,
        isLoading,
        error: ((_a = searchTypes.find(item => item.data.error)) === null || _a === void 0 ? void 0 : _a.data.error) || null
    };
};
exports.useCombineData = useCombineData;
function getArrayOfElements(elem) {
    if (!elem) {
        return [];
    }
    return Array.isArray(elem) ? elem : [elem];
}
exports.getArrayOfElements = getArrayOfElements;
