// @geotab/zenith/dist/chart/plugins/tooltip/tooltipPlugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TooltipPlugin = void 0;
const getBarTooltipItems_1 = require("../../barChart/getBarTooltipItems");
const getPieTooltipItems_1 = require("../../pieChart/getPieTooltipItems");
const renderTooltip_1 = require("./renderTooltip");
const getScrollableParent_1 = require("../../../utils/getScrollableParent");
const getCenterOfSector = (element) => {
    const midAngle = (element.startAngle + element.endAngle) / 2;
    const centerRadius = (element.innerRadius + element.outerRadius) / 2;
    const centerX = element.x + Math.cos(midAngle) * centerRadius;
    const centerY = element.y + Math.sin(midAngle) * centerRadius;
    return { x: centerX, y: centerY };
};
const TooltipPlugin = (containerId, options, alignment = "top", isHidden, isDark) => ({
    id: "zenithTooltip",
    afterInit: (chart) => {
        if (options === false) {
            return;
        }
        chart.zenithTooltip = {
            x: null,
            y: null,
            enabled: false,
            index: undefined,
            datasetIndex: undefined,
            mouseMove: event => {
                // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
                if (!chart.canvas) {
                    chart.zenithTooltip.enabled = false;
                    chart.update();
                    return;
                }
                let elements;
                try {
                    elements = chart.getElementsAtEventForMode(event, "nearest", { intersect: true }, true);
                }
                catch (_) {
                    elements = [];
                }
                const element = elements[0];
                if (!element) {
                    chart.zenithTooltip.enabled = false;
                    chart.update();
                    return;
                }
                chart.zenithTooltip.index = element.index;
                chart.zenithTooltip.datasetIndex = element.datasetIndex;
                if (element.element !== undefined && element.element.hasOwnProperty("startAngle") && element.element.hasOwnProperty("endAngle")) {
                    const pos = getCenterOfSector(element.element);
                    chart.zenithTooltip.x = pos.x;
                    chart.zenithTooltip.y = pos.y;
                    chart.zenithTooltip.enabled = true;
                    chart.update();
                    return;
                }
                if (element.element !== undefined && element.element.hasOwnProperty("x") && element.element.hasOwnProperty("y")) {
                    chart.zenithTooltip.x = element.element.x;
                    chart.zenithTooltip.y = element.element.y;
                    chart.zenithTooltip.enabled = true;
                    chart.update();
                    return;
                }
                chart.zenithTooltip.enabled = false;
                chart.update();
            }, mouseOut: () => {
                chart.zenithTooltip.enabled = false;
                chart.update();
            }
        };
        const canvas = chart.canvas;
        canvas.addEventListener("mousemove", chart.zenithTooltip.mouseMove);
        chart.canvas.addEventListener("mouseout", () => {
            chart.zenithTooltip.enabled = false;
        });
    },
    afterDraw: (chart) => {
        if (options === false) {
            return;
        }
        // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
        if (chart && chart.zenithTooltip && chart.zenithTooltip.enabled) {
            const x = chart.zenithTooltip.x;
            const y = chart.zenithTooltip.y;
            const index = chart.zenithTooltip.index;
            const datasetIndex = chart.zenithTooltip.datasetIndex;
            // eslint-disable-next-line
            const chartType = chart.config.type;
            let values = [];
            let title = undefined;
            const subtitle = undefined;
            if (chartType === "bar") {
                values = (0, getBarTooltipItems_1.getBarTooltipItems)(chart.data, datasetIndex, index, options, isHidden);
                title = (0, getBarTooltipItems_1.getBarTooltipTitle)(chart.data, datasetIndex, index);
            }
            if (chartType === "pie") {
                values = (0, getPieTooltipItems_1.getPieTooltipItems)(chart.data, datasetIndex, index, options);
            }
            const chartRect = chart.canvas.getBoundingClientRect();
            document.body.getBoundingClientRect();
            const scrollableParent = (0, getScrollableParent_1.getScrollableParent)(chart.canvas);
            const scrollTop = scrollableParent ? scrollableParent.scrollTop : 0;
            const scrollLeft = scrollableParent ? scrollableParent.scrollLeft : 0;
            const top = chartRect.top + y + scrollTop;
            const left = chartRect.left + x + scrollLeft;
            if (values.length > 0) {
                (0, renderTooltip_1.renderTooltip)(containerId, left, top, alignment, values, (options === null || options === void 0 ? void 0 : options.title) || title, (options === null || options === void 0 ? void 0 : options.subtitle) || subtitle, (options === null || options === void 0 ? void 0 : options.note) || undefined, isDark);
            }
            else {
                (0, renderTooltip_1.hideTooltip)(containerId);
            }
        }
        else {
            (0, renderTooltip_1.hideTooltip)(containerId);
        }
    },
    beforeDestroy: (chart) => {
        var _a, _b;
        // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
        if (chart.canvas !== null) {
            // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
            ((_a = chart.zenithTooltip) === null || _a === void 0 ? void 0 : _a.mouseMove) && chart.canvas.removeEventListener("mousemove", chart.zenithTooltip.mouseMove);
            // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
            ((_b = chart.zenithTooltip) === null || _b === void 0 ? void 0 : _b.mouseOut) && chart.canvas.removeEventListener("mouseout", chart.zenithTooltip.mouseOut);
        }
        (0, renderTooltip_1.destroyTooltip)(containerId);
    }
});
exports.TooltipPlugin = TooltipPlugin;
