// @geotab/zenith/dist/nav/navFooter/navFooter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NavFooter = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const classNames_1 = require("../../commonHelpers/classNames/classNames");
const react_1 = require("react");
const navItem_1 = require("../navItem/navItem");
const useLanguage_1 = require("../../utils/localization/useLanguage");
const useMobile_1 = require("../../commonHelpers/hooks/useMobile");
const iconNavigationFooterCollapsed_1 = require("../../icons/iconNavigationFooterCollapsed");
const iconNavigationFooterExpanded_1 = require("../../icons/iconNavigationFooterExpanded");
const navFooterAction_1 = require("./navFooterAction/navFooterAction");
const iconEdit_1 = require("../../icons/iconEdit");
const nav_context_1 = require("../context/nav.context");
const InnerContext = (0, react_1.createContext)({
    collapsed: false,
    onCollapseToggle: (_) => { },
    onEditNavClick: () => { }
});
const NavFooterComponent = ({ className, onEditNavClick, children }) => {
    const { collapsed, onCollapseToggle } = (0, nav_context_1.useNavContext)();
    const footerAction = react_1.Children.toArray(children).find(child => {
        const element = child;
        return element.type && element.type.displayName === "NavFooterAction";
    });
    const bottom = react_1.Children.toArray(children).find(child => {
        const element = child;
        return element.type && element.type.displayName === "NavFooterBottom";
    });
    return (0, jsx_runtime_1.jsxs)("div", { className: (0, classNames_1.classNames)(["zen-nav-footer", collapsed ? "zen-nav-footer--collapsed" : "", className || ""]), children: [(0, jsx_runtime_1.jsx)("div", { className: "zen-nav-footer__top", children: footerAction }), (0, jsx_runtime_1.jsx)(InnerContext.Provider, { value: { collapsed, onCollapseToggle, onEditNavClick }, children: bottom })] });
};
const NavFooterBottom = ({ className, children }) => {
    const { translate } = (0, useLanguage_1.useLanguage)();
    const { collapsed, onCollapseToggle, onEditNavClick } = (0, react_1.useContext)(InnerContext);
    const isMobile = (0, useMobile_1.useMobile)();
    const childrenArray = react_1.Children.toArray(children).filter(child => {
        const element = child;
        return element.type && element.type.displayName === "NavItem";
    });
    return (0, jsx_runtime_1.jsxs)("div", { className: (0, classNames_1.classNames)(["zen-nav-footer__bottom", isMobile ? "zen-nav-footer__bottom--mobile" : "", className || ""]), children: [childrenArray.map(child => {
                const childElement = child;
                return (0, react_1.cloneElement)(childElement, {
                    collapsed: !isMobile,
                    className: (0, classNames_1.classNames)([
                        childElement.props.className || "",
                        "zen-nav-footer__button"
                    ]),
                    menuAlignment: "right-bottom"
                });
            }), (0, jsx_runtime_1.jsx)(navItem_1.NavItem, { className: "zen-nav-footer__button", title: translate("Edit Navigation"), collapsed: !isMobile, primaryIcon: iconEdit_1.IconEdit, onClick: () => onEditNavClick && onEditNavClick(), menuAlignment: "right-bottom" }, "help-button"), isMobile ? null : (0, jsx_runtime_1.jsx)(navItem_1.NavItem, { className: "zen-nav-footer__button", title: collapsed ? translate("Expand Navigation") : translate("Collapse Navigation"), collapsed: !isMobile, primaryIcon: collapsed ? iconNavigationFooterCollapsed_1.IconNavigationFooterCollapsed : iconNavigationFooterExpanded_1.IconNavigationFooterExpanded, onClick: () => onCollapseToggle && onCollapseToggle(!collapsed), menuAlignment: "right-bottom" }, "collapse-button")] });
};
NavFooterBottom.displayName = "NavFooterBottom";
/**
 * @beta This component is not fully ready yet and may change in future releases.
 */
exports.NavFooter = Object.assign(NavFooterComponent, {
    Action: navFooterAction_1.NavFooterAction,
    Bottom: NavFooterBottom
});
exports.NavFooter.displayName = "NavFooter";
