// @geotab/zenith/dist/filtersBar/filtersBarReducer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.filtersBarReducer = exports.getInitialFiltersBarState = exports.FiltersBarStateActionType = void 0;
const getNumberOfChangedFilters_1 = require("./utils/getNumberOfChangedFilters");
var FiltersBarStateActionType;
(function (FiltersBarStateActionType) {
    FiltersBarStateActionType["ChangeChildrenState"] = "CHANGE_CHILDREN_STATE";
    FiltersBarStateActionType["UpdateNumber"] = "UPDATE_NUMBER";
})(FiltersBarStateActionType || (exports.FiltersBarStateActionType = FiltersBarStateActionType = {}));
const getInitialFiltersBarState = (childrenState, getDefaultFiltersState, allowCurrentNumberUpdates) => {
    const numberOfChangedFilters = (0, getNumberOfChangedFilters_1.getNumberOfChangedFilters)(getDefaultFiltersState(), childrenState);
    return {
        childrenState: childrenState,
        prevChildrenState: childrenState,
        allowCurrentNumberUpdates: allowCurrentNumberUpdates,
        numberOfChangedFilters: numberOfChangedFilters,
        currentNumberOfChangedFilters: numberOfChangedFilters,
        getDefaultFiltersState: getDefaultFiltersState
    };
};
exports.getInitialFiltersBarState = getInitialFiltersBarState;
function filtersBarReducer(state, action) {
    if (action.type === FiltersBarStateActionType.ChangeChildrenState) {
        const newChildrenState = action.payload.state;
        const newDefaultFiltersStateFn = action.payload.defaultStateFn;
        const newNumberOfChangedFilters = (0, getNumberOfChangedFilters_1.getNumberOfChangedFilters)(newDefaultFiltersStateFn(), newChildrenState);
        return Object.assign(Object.assign({}, state), { childrenState: newChildrenState, getDefaultFiltersState: newDefaultFiltersStateFn, currentNumberOfChangedFilters: newNumberOfChangedFilters, numberOfChangedFilters: state.allowCurrentNumberUpdates ? newNumberOfChangedFilters : state.numberOfChangedFilters });
    }
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
    if (action.type === FiltersBarStateActionType.UpdateNumber) {
        const newChildrenState = action.payload;
        const newNumberOfChangedFilters = (0, getNumberOfChangedFilters_1.getNumberOfChangedFilters)(state.getDefaultFiltersState(), newChildrenState);
        return Object.assign(Object.assign({}, state), { childrenState: newChildrenState, prevChildrenState: newChildrenState, currentNumberOfChangedFilters: newNumberOfChangedFilters, numberOfChangedFilters: newNumberOfChangedFilters });
    }
    return state;
}
exports.filtersBarReducer = filtersBarReducer;
