// @geotab/zenith/dist/header/components/collapsedItemsControl/collapsedItemsControl.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TRANSLATIONS = exports.PageToolbarCollapsedItemsControl = exports.HeaderCustomContentProvider = exports.HeaderCustomContent = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const menu_1 = require("../../../menu/menu");
const buttonType_1 = require("../../../button/buttonType");
const svgIconSize_1 = require("../../../svgIcon/svgIconSize");
const textIconButton_1 = require("../../../textIconButton/textIconButton");
const useLanguage_1 = require("../../../utils/localization/useLanguage");
const headerButton_1 = require("../../headerButton");
const headerMenu_1 = require("../../headerMenu");
const headerHelpers_1 = require("../../headerHelpers");
const iconDotVertical_1 = require("../../../icons/iconDotVertical");
const useDrive_1 = require("../../../utils/theme/useDrive");
const handleCustomMenuWrapper_1 = require("./utils/handleCustomMenuWrapper");
const pageHeaderButton_1 = require("../../../pageHeader/pageHeaderButton");
const pageHeaderMenu_1 = require("../../../pageHeader/pageHeaderMenu");
const classNames_1 = require("../../../commonHelpers/classNames/classNames");
exports.HeaderCustomContent = (0, react_1.createContext)({
    content: null,
    setContent: (_1) => { }
});
const HeaderCustomContentProvider = ({ children, content, setContent }) => (0, jsx_runtime_1.jsx)(exports.HeaderCustomContent.Provider, { value: { content: content, setContent }, children: children });
exports.HeaderCustomContentProvider = HeaderCustomContentProvider;
const PageToolbarCollapsedItemsControl = ({ children, className }) => {
    const { translate } = (0, useLanguage_1.useLanguage)();
    const isDrive = (0, useDrive_1.useDrive)();
    const { content } = (0, react_1.useContext)(exports.HeaderCustomContent);
    const adjustmentButton = (0, react_1.useMemo)(() => (0, jsx_runtime_1.jsx)(textIconButton_1.TextIconButton, { "aria-label": translate("More"), className: (0, classNames_1.classNames)(["zen-header-more-button", className || ""]), type: buttonType_1.ButtonType.TertiaryBlack, iconPosition: textIconButton_1.ButtonIconPosition.End, iconSize: isDrive ? svgIconSize_1.SvgIconSize.Huge : svgIconSize_1.SvgIconSize.Large, icon: iconDotVertical_1.IconDotVertical }), [isDrive, translate, className]);
    const items = [];
    const clickHandler = (0, react_1.useCallback)((propsOnClick) => ((_, e) => { propsOnClick(e); }), []);
    react_1.Children.forEach(children, (child) => {
        if ((0, headerHelpers_1.isNamedChildComponent)(child, headerButton_1.HeaderButtonDisplayName) || (0, headerHelpers_1.isNamedChildComponent)(child, pageHeaderButton_1.PageHeaderButtonDisplayName)) {
            const onClick = child.props.onClick;
            const el = (0, react_1.createElement)(menu_1.Menu.Item, {
                id: child.props.id,
                key: child.props.id,
                name: typeof child.props.children === "string"
                    ? child.props.children
                    : child.props.title
                        ? child.props.title
                        : "",
                disabled: child.props.disabled,
                onClick: onClick ? clickHandler((e) => onClick(child.props.id, e)) : undefined,
                link: child.props.link,
                target: child.props.target,
                rel: child.props.rel,
                icon: child.props.icon
            });
            items.push(el);
        }
        if ((0, headerHelpers_1.isNamedChildComponent)(child, headerMenu_1.HeaderMenuDisplayName) || (0, headerHelpers_1.isNamedChildComponent)(child, pageHeaderMenu_1.PageHeaderMenuDisplayName)) {
            const el = (0, react_1.createElement)(menu_1.Menu.Item, {
                id: child.props.id,
                key: child.props.id,
                name: child.props.name,
                disabled: child.props.disabled,
                link: child.props.link,
                target: child.props.target,
                rel: child.props.rel,
                icon: child.props.icon
            }, child.props.children);
            items.push(el);
        }
        if ((0, headerHelpers_1.isNamedChildComponent)(child, headerHelpers_1.HeaderCustomMenuWrapper)) {
            (0, handleCustomMenuWrapper_1.handleCustomMenuWrapper)(child, content, items);
        }
    });
    return items.length ? (0, jsx_runtime_1.jsx)(menu_1.Menu, { title: translate("More"), trigger: adjustmentButton, alignment: "bottom-right", children: items }) : null;
};
exports.PageToolbarCollapsedItemsControl = PageToolbarCollapsedItemsControl;
exports.TRANSLATIONS = [
    "More"
];
