// @geotab/zenith/dist/chart/plugins/linePlugin/linePlugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LinePlugin = void 0;
const formatValue_1 = require("../legend/formatValue");
const renderTooltip_1 = require("../tooltip/renderTooltip");
const getInterpolatedValue_1 = require("./getInterpolatedValue");
const drawLine_1 = require("./drawLine");
const getScrollableParent_1 = require("../../../utils/getScrollableParent");
const LinePlugin = (containerId, options, isDark) => ({
    id: "verticalLine",
    afterInit: (chart) => {
        if (options === false) {
            return;
        }
        const canvas = chart.canvas;
        chart.verticalLine = {
            x: null,
            y: null,
            enabled: false,
            mouseMove: event => {
                const rect = canvas.getBoundingClientRect();
                const x = event.clientX - rect.left;
                const y = event.clientY - rect.top;
                // eslint-disable-next-line no-underscore-dangle, @typescript-eslint/no-explicit-any
                const activeElements = (options === null || options === void 0 ? void 0 : options.tooltipOnPoints) ? chart._active : [];
                if (x >= chart.chartArea.left && x <= chart.chartArea.right && (!(options === null || options === void 0 ? void 0 : options.tooltipOnPoints) || activeElements.length > 0)) {
                    const sortedPoints = activeElements.sort((a, b) => { var _a, _b; return Math.abs((((_a = a.element) === null || _a === void 0 ? void 0 : _a.x) || 0) - x) - Math.abs((((_b = b.element) === null || _b === void 0 ? void 0 : _b.x) || 0) - x); });
                    chart.verticalLine.x = sortedPoints.length > 0 ? sortedPoints[0].element.x : x;
                    chart.verticalLine.y = y;
                    chart.verticalLine.enabled = true;
                    chart.update(); // Trigger redraw
                }
                else {
                    chart.verticalLine.enabled = false;
                    chart.update();
                }
            },
            mouseOut: () => {
                chart.verticalLine.enabled = false;
                chart.update();
            }
        };
        canvas.addEventListener("mousemove", chart.verticalLine.mouseMove);
        canvas.addEventListener("mouseout", chart.verticalLine.mouseOut);
    },
    afterDraw: (chart) => {
        if (options === false) {
            return;
        }
        if (chart.verticalLine.enabled) {
            const ctx = chart.ctx;
            const x = chart.verticalLine.x;
            const y = chart.verticalLine.y;
            const topY = chart.chartArea.top;
            const bottomY = chart.chartArea.bottom;
            (0, drawLine_1.drawLine)(ctx, x, topY, bottomY);
            const datasets = chart.data.datasets;
            let lowestValueY = Infinity;
            const items = datasets.map((dataset, dsIndex) => {
                if (!chart.isDatasetVisible(dsIndex)) {
                    return undefined;
                }
                if (chart.getDatasetMeta(dsIndex).type !== "line") {
                    return undefined;
                }
                const xScale = chart.scales.x;
                const value = (0, getInterpolatedValue_1.getInterpolatedValue)(xScale, chart.data, dsIndex, x);
                const yAxisID = dataset.yAxisID || "y";
                const yScale = chart.scales[yAxisID];
                // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
                if (value && yScale) {
                    const lowestDsY = yScale.getPixelForValue(value);
                    if (lowestDsY < lowestValueY) {
                        lowestValueY = lowestDsY;
                    }
                }
                if (value && dataset.label) {
                    const label = dataset.label;
                    const color = dataset.borderColor;
                    const item = {
                        label: label,
                        value: (0, formatValue_1.formatValue)(value.toFixed(2), dsIndex, options === null || options === void 0 ? void 0 : options.unit),
                        color: typeof color === "string" ? color : ""
                    };
                    return item;
                }
                return undefined;
            }).filter(item => !!item);
            if (items.length === 0) {
                (0, renderTooltip_1.hideTooltip)(containerId);
                return;
            }
            const chartRect = chart.canvas.getBoundingClientRect();
            const top = chartRect.top + (lowestValueY || y);
            const left = chartRect.left + x;
            const scrollableParent = (0, getScrollableParent_1.getScrollableParent)(chart.canvas);
            const scrollTop = scrollableParent ? scrollableParent.scrollTop : 0;
            const scrollLeft = scrollableParent ? scrollableParent.scrollLeft : 0;
            const xValue = chart.scales.x.getValueForPixel(x);
            const xLabel = xValue ? chart.scales.x.getLabelForValue(xValue) : undefined;
            // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
            (0, renderTooltip_1.renderTooltip)(containerId, left + scrollLeft, top + scrollTop, "top", items, options === null || options === void 0 ? void 0 : options.title, options && options.subtitle !== undefined ? options.subtitle : xLabel, (options === null || options === void 0 ? void 0 : options.note) || undefined, isDark);
        }
        else {
            (0, renderTooltip_1.hideTooltip)(containerId);
        }
    },
    beforeDestroy: (chart) => {
        var _a, _b;
        // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
        if (chart.canvas !== null) {
            // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
            ((_a = chart.verticalLine) === null || _a === void 0 ? void 0 : _a.mouseMove) && chart.canvas.removeEventListener("mousemove", chart.verticalLine.mouseMove);
            // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
            ((_b = chart.verticalLine) === null || _b === void 0 ? void 0 : _b.mouseOut) && chart.canvas.removeEventListener("mouseout", chart.verticalLine.mouseOut);
        }
        (0, renderTooltip_1.destroyTooltip)(containerId);
    }
});
exports.LinePlugin = LinePlugin;
