// @geotab/zenith/dist/dateRangeInner/dateRangeInner.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TRANSLATIONS = exports.DateRangeInner = exports.CUSTOM_CALENDAR_ID = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const radioGroup_1 = require("../radioGroup/radioGroup");
const dateInputInner_1 = require("../dateInputInner/dateInputInner");
const react_1 = require("react");
const useLanguage_1 = require("../utils/localization/useLanguage");
const userFormatContext_1 = require("../utils/userFormat/userFormatContext");
const formatDate_1 = require("../utils/formatDate");
const iconChevronRight_1 = require("../icons/iconChevronRight");
const dateRange_1 = require("../dateRange/dateRange");
const normalizeDates_1 = require("../filtersBar/components/filtersBarPeriodPicker/normalizeDates");
const classNames_1 = require("../commonHelpers/classNames/classNames");
exports.CUSTOM_CALENDAR_ID = "Custom";
const DateRangeInner = ({ id, options: optionsArg, selection, range = { from: undefined, to: undefined }, onOptionSelect, onCalendarSelect, timeSelect, disableFutureDates, disablePastDates, disableDatesAfter, disableDatesBefore, showMobileCalendar, setShowMobileCalendar }) => {
    const { translate } = (0, useLanguage_1.useLanguage)();
    const isMobileView = !!setShowMobileCalendar;
    const { dateFormat, weekStartsOnSunday, toLocalDateTime } = (0, react_1.useContext)(userFormatContext_1.userFormatContext);
    const dateFormatter = (d) => (0, formatDate_1.formatDate)(toLocalDateTime(d), dateFormat, translate);
    const dateDeformatter = (d) => (0, formatDate_1.deformatDate)(d, dateFormat, translate) || new Date("");
    const isCalendarVisible = !!onCalendarSelect;
    const calendarLabel = translate(exports.CUSTOM_CALENDAR_ID);
    const normalizedUnknownRange = (0, react_1.useMemo)(() => (0, normalizeDates_1.normalizeDates)(undefined, undefined, disableFutureDates, disablePastDates, disableDatesBefore, disableDatesAfter), [disableDatesAfter, disableDatesBefore, disableFutureDates, disablePastDates]);
    const options = (0, react_1.useMemo)(() => optionsArg.map((opt) => (Object.assign(Object.assign({}, opt), (opt.hasSecondaryLabel ? { secondaryLabel: opt.getRange ? (0, dateRange_1.parseLabel)(Object.assign({ label: opt.label }, opt.getRange()), dateFormat, translate, toLocalDateTime)
            : (0, dateRange_1.parseLabel)({ label: "unknown", from: new Date(normalizedUnknownRange.from), to: new Date(normalizedUnknownRange.to) }, dateFormat, translate, toLocalDateTime) } : {})))), [optionsArg, dateFormat, translate, toLocalDateTime, normalizedUnknownRange.from, normalizedUnknownRange.to]);
    const items = (0, react_1.useMemo)(() => {
        const handleButtonClick = () => {
            setShowMobileCalendar && setShowMobileCalendar(true);
            selection !== exports.CUSTOM_CALENDAR_ID && onOptionSelect(exports.CUSTOM_CALENDAR_ID);
        };
        const mobileCustomLabel = (label, secondaryLabel) => (0, jsx_runtime_1.jsxs)("button", { type: "button", onClick: handleButtonClick, className: (0, classNames_1.classNames)(["zen-date-range-inner__mobile-custom-label", secondaryLabel === undefined ? "zen-date-range-inner__mobile-custom-label--no-secondary" : ""]), children: [(0, jsx_runtime_1.jsxs)("div", { className: "zen-date-range-inner__range-label", children: [(0, jsx_runtime_1.jsx)("div", { children: label }), secondaryLabel ? (0, jsx_runtime_1.jsx)("div", { className: "zen-date-range-inner__label-secondary", children: secondaryLabel }) : null] }), (0, jsx_runtime_1.jsx)("div", { className: "zen-date-range-inner__mobile-custom-icon-wrapper", children: (0, jsx_runtime_1.jsx)(iconChevronRight_1.IconChevronRight, { size: "large" }) })] });
        const getLabel = (mainLabel, secondaryLabel) => (0, jsx_runtime_1.jsxs)("div", { className: "zen-date-range-inner__range-label", children: [(0, jsx_runtime_1.jsx)("div", { children: mainLabel }), secondaryLabel ? (0, jsx_runtime_1.jsx)("div", { className: "zen-date-range-inner__label-secondary", children: secondaryLabel }) : null] });
        const optionsToPass = options.map(option => ({
            id: option.id,
            value: option.id,
            children: isMobileView && option.id === "Custom" ? mobileCustomLabel(calendarLabel, option.secondaryLabel)
                : getLabel(option.label, option.secondaryLabel),
            key: option.id
        }));
        return optionsToPass;
    }, [calendarLabel, isMobileView, onOptionSelect, options, selection, setShowMobileCalendar]);
    const calendarId = (0, react_1.useMemo)(() => `${id}-calendar`, [id]);
    const optionSelectHandler = (0, react_1.useCallback)(e => {
        let selectedValue = e.target.value;
        if (!selectedValue && e.target.classList.contains("zen-date-range-inner__mobile-custom-label")) {
            selectedValue = exports.CUSTOM_CALENDAR_ID;
        }
        onOptionSelect(selectedValue);
    }, [onOptionSelect]);
    const calendarSelectHandler = (0, react_1.useCallback)((selectedRange) => {
        onCalendarSelect && onCalendarSelect(selectedRange);
    }, [onCalendarSelect]);
    const isMobileCalendar = isMobileView && showMobileCalendar;
    const memoizedRadioGroup = (0, react_1.useMemo)(() => {
        if (!isMobileView || !showMobileCalendar) {
            return (0, jsx_runtime_1.jsx)(radioGroup_1.RadioGroup
            // className when there is a custom property in mobile view
            , { 
                // className when there is a custom property in mobile view
                className: isCalendarVisible && isMobileView ? "zen-date-range-inner__radio-group--with-custom" : "", items: items, fullWidth: true, value: selection, name: `daterange__${id}`, direction: "vertical", type: "interactive", onChange: optionSelectHandler, selectItemOnFocus: false, wrappedInLabel: true });
        }
        return null;
    }, [id, isCalendarVisible, isMobileView, items, optionSelectHandler, selection, showMobileCalendar]);
    return (0, jsx_runtime_1.jsxs)("div", { id: id, className: "zen-date-range-inner", children: [memoizedRadioGroup, isCalendarVisible && selection === exports.CUSTOM_CALENDAR_ID && (!isMobileView || showMobileCalendar)
                ? (0, jsx_runtime_1.jsx)(jsx_runtime_1.Fragment, { children: (0, jsx_runtime_1.jsx)("div", { className: "zen-date-range-inner__calendar-wrapper", children: (0, jsx_runtime_1.jsx)(dateInputInner_1.DateInputInner, { id: calendarId, value: range, onChange: calendarSelectHandler, dateFormat: dateFormat, dateFormatter: dateFormatter, dateDeformatter: dateDeformatter, startDayOfWeek: weekStartsOnSunday ? 0 : 1, disableFutureDates: disableFutureDates, disablePastDates: disablePastDates, disableDatesBefore: disableDatesBefore, disableDatesAfter: disableDatesAfter, dateRangeMode: true, selectTime: timeSelect, isMobileView: isMobileCalendar, requireSelection: true }) }) })
                : null] });
};
exports.DateRangeInner = DateRangeInner;
exports.TRANSLATIONS = [
    "Custom Date",
    "Custom Date and Time range",
    "Today"
];
