// @geotab/zenith/dist/dataGrid/withFlexibleColumns/components/columnSettingsList.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ColumnSettingsList = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const groupColumns_1 = require("../../extensions/groupColumns");
const selectList_1 = require("../../../selectList/selectList");
const checkboxState_1 = require("../../../checkbox/checkboxState");
const SubColumnsList = ({ title, columns, settings, firstColumn, onChange, sortColumns }) => {
    const id = (0, react_1.useId)();
    const sortColumnsInt = (0, react_1.useCallback)((a, b) => {
        if (sortColumns) {
            return sortColumns(a.title, b.title);
        }
        return 0;
    }, [sortColumns]);
    const items = (0, react_1.useMemo)(() => columns.sort(sortColumnsInt).map(c => ({
        id: c.id,
        title: c.title,
        disabled: c.name === firstColumn,
        showCheckbox: true,
        children: c.title
    })), [columns, firstColumn, sortColumnsInt]);
    const state = columns.reduce((res, c) => {
        res.set(c.id, settings.get(c.name) !== false ? checkboxState_1.CheckboxState.On : checkboxState_1.CheckboxState.Off);
        return res;
    }, new Map());
    const onChangeItem = (0, react_1.useCallback)((newState) => {
        const newSettings = new Map(settings);
        newState.forEach((stateVal, itemId) => {
            var _a;
            newSettings.set((_a = columns.find(c => c.id === itemId)) === null || _a === void 0 ? void 0 : _a.name, stateVal === checkboxState_1.CheckboxState.On);
        });
        onChange && onChange(newSettings);
    }, [columns, settings, onChange]);
    const list = (0, react_1.useMemo)(() => (0, jsx_runtime_1.jsx)(selectList_1.SelectList, { id: id, items: items, value: state, title: title || "", multiselect: true, showCheckbox: true, onChange: onChangeItem }), [id, items, state, title, onChangeItem]);
    if (!title) {
        return list;
    }
    return (0, jsx_runtime_1.jsxs)("div", { className: "zen-columns-list__sub-section", children: [(0, jsx_runtime_1.jsx)("div", { className: "zen-columns-list__title zen-ellipsis", children: title }), list] });
};
function ColumnSettingsList({ columns, settings, firstColumn, onChange, sortGroups, sortColumns }) {
    const groupped = (0, groupColumns_1.groupColumns)(columns);
    const sortGroupsInt = (0, react_1.useCallback)((a, b) => {
        if (sortGroups) {
            return sortGroups(a[0], b[0]);
        }
        return 0;
    }, [sortGroups]);
    return Array.from(groupped.entries()).sort(sortGroupsInt).map(([title, cs]) => (0, jsx_runtime_1.jsx)(SubColumnsList, { title: title, columns: cs, settings: settings, firstColumn: firstColumn, onChange: onChange, sortColumns: sortColumns }, cs.map(c => c.id).join("-")));
}
exports.ColumnSettingsList = ColumnSettingsList;
