// @geotab/zenith/dist/nav/storyHelpers/storyDecorator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.navContainerDecorator = exports.SearchDecorator = exports.navDecorators = exports.createNavContainerDecorator = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const nav_context_1 = require("../context/nav.context");
const createNavContainerDecorator = (options = {}) => {
    const { collapsed, collapsedWidth = "64px", expandedWidth = "256px", withoutStyles = false } = options;
    // eslint-disable-next-line @typescript-eslint/naming-convention
    const DecoratorStory = (Story, context) => {
        const [isCollapsed, setIsCollapsed] = (0, react_1.useState)(collapsed);
        const [searchTerm, setSearchTerm] = (0, react_1.useState)("");
        const [isSearchOpen, setIsSearchOpen] = (0, react_1.useState)(false);
        const [isMobile, setIsMobile] = (0, react_1.useState)(false);
        const isCollapsedState = !isMobile && isCollapsed && !isSearchOpen;
        const containerWidth = isMobile ? "100%" : isCollapsedState ? collapsedWidth : expandedWidth;
        return ((0, jsx_runtime_1.jsx)("div", { className: withoutStyles ? "" : "zen-shadow-dropdown-default", style: {
                maxWidth: "100%",
                width: containerWidth,
                overflow: withoutStyles ? undefined : "hidden",
                padding: withoutStyles ? undefined : (isMobile ? "12px" : "16px 12px"),
                backgroundColor: withoutStyles ? undefined : "var(--backgrounds-main)",
                display: "flex",
                flexDirection: "column",
                alignItems: withoutStyles ? undefined : (isCollapsedState ? "center" : "normal")
            }, children: (0, jsx_runtime_1.jsx)(nav_context_1.NavContext.Provider, { value: {
                    collapsed: !!isCollapsedState,
                    onCollapseToggle: setIsCollapsed,
                    searchOpen: isSearchOpen,
                    onSearchToggle: setIsSearchOpen,
                    searchTerm,
                    onSearch: setSearchTerm
                }, children: (0, jsx_runtime_1.jsx)(Story, { args: Object.assign(Object.assign({}, context.args), { setIsMobile,
                        setIsCollapsed, collapsed: isCollapsedState }) }) }) }));
    };
    return DecoratorStory;
};
exports.createNavContainerDecorator = createNavContainerDecorator;
/**
 * Pre-configured decorators for common navigation scenarios
 */
exports.navDecorators = {
    /**
     * Uses parameters with collapsed boolean
     * It allows to dynamically change the state from the story
     */
    withParameters: (0, exports.createNavContainerDecorator)(),
    /**
     * Fixed width for expanded navigation (232px)
     */
    expanded: (0, exports.createNavContainerDecorator)({
        collapsed: false
    }),
    /**
     * Fixed width for collapsed navigation (48px)
     */
    collapsed: (0, exports.createNavContainerDecorator)({
        collapsed: true
    }),
    /**
     * Decorator without styles
     */
    withoutStyles: (0, exports.createNavContainerDecorator)({
        withoutStyles: true
    })
};
// eslint-disable-next-line @typescript-eslint/naming-convention
const SearchDecorator = (Story, context) => {
    var _a;
    const loggerRef = (0, react_1.useRef)(null);
    // eslint-disable-next-line @typescript-eslint/naming-convention
    return (0, jsx_runtime_1.jsxs)(jsx_runtime_1.Fragment, { children: [(0, jsx_runtime_1.jsx)(Story, { args: Object.assign(Object.assign({}, context.args), { onSearch: term => {
                        if (loggerRef.current) {
                            loggerRef.current.innerText = term;
                        }
                    } }) }), (0, jsx_runtime_1.jsxs)("div", { style: { padding: "10px 0" }, children: ["Search term: ", (0, jsx_runtime_1.jsx)("span", { style: { fontWeight: "bold" }, ref: loggerRef, children: (_a = loggerRef.current) === null || _a === void 0 ? void 0 : _a.innerText })] })] });
};
exports.SearchDecorator = SearchDecorator;
exports.navContainerDecorator = exports.navDecorators.withParameters;
