// @geotab/zenith/dist/lineChart/lineChart.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LineChart = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const typedCharts_1 = require("../react-chartjs/typedCharts");
const auto_1 = require("chart.js/auto");
require("../react-chartjs/dateAdapter");
const utils_1 = require("../commonHelpers/utils");
const utils_2 = require("./utils");
const htmlLegend_1 = require("./plugins/htmlLegend");
const useDrive_1 = require("../utils/theme/useDrive");
const useMobile_1 = require("../commonHelpers/hooks/useMobile");
const themeContext_1 = require("../utils/theme/themeContext");
auto_1.Chart.register(auto_1.CategoryScale, auto_1.LinearScale, auto_1.PointElement, auto_1.LineElement, auto_1.Title, auto_1.Tooltip, auto_1.Legend, auto_1.Filler, auto_1.TimeScale);
const defaultGridStyle = {
    drawTicks: false,
    color: "#C0CCD8"
};
const defaultGridBorderStyle = {
    display: false,
    dash: [2, 2]
};
const defTicksStyle = {
    padding: 10
};
const getDefaultOptions = (fontSize) => {
    const defTitleStyle = {
        display: true,
        font: {
            size: fontSize,
            weight: "400"
        }
    };
    return {
        responsive: true,
        maintainAspectRatio: false,
        scales: {
            x: {
                grid: defaultGridStyle,
                border: defaultGridBorderStyle,
                ticks: defTicksStyle,
                title: defTitleStyle
            },
            y: {
                grid: defaultGridStyle,
                border: defaultGridBorderStyle,
                ticks: defTicksStyle,
                title: defTitleStyle
            }
        },
        plugins: {
            legend: {
                display: false
            },
            tooltip: {
                enabled: true
            }
        }
    };
};
/**
 * @deprecated - will be removed in the next releases.
 */
const LineChart = (_a) => {
    var { data, options = {}, plugins, className } = _a, rest = __rest(_a, ["data", "options", "plugins", "className"]);
    const isDrive = (0, useDrive_1.useDrive)();
    const isMobile = (0, useMobile_1.useMobile)();
    const memoizedFontSize = (0, react_1.useMemo)(() => isDrive ? (isMobile ? 14 : 16) : 12, [isDrive, isMobile]);
    const { dark } = (0, react_1.useContext)(themeContext_1.themeContext);
    (0, react_1.useEffect)(() => {
        auto_1.Chart.defaults.font = {
            family: "Roboto",
            size: memoizedFontSize
        };
        auto_1.Chart.defaults.color = dark ? "#FFFFFF" : "#1F2833";
    }, [memoizedFontSize, dark]);
    const adaptedData = Object.assign({}, data);
    const memoizedDefaultOptions = (0, react_1.useMemo)(() => getDefaultOptions(memoizedFontSize), [memoizedFontSize]);
    const adaptedOptions = (0, utils_1.deepMerge)(memoizedDefaultOptions, options);
    adaptedData.datasets = adaptedData.datasets.map((ds, i) => (0, utils_2.getDefaultDatasetStyle)(ds, i));
    const memoizedPlagins = (0, react_1.useMemo)(() => plugins === null || plugins === void 0 ? void 0 : plugins.map(el => (el.id === htmlLegend_1.LEGEND_PLUGIN_ID ? Object.assign(Object.assign({}, el), { isDrive: isDrive }) : el)), [isDrive, plugins]);
    return (0, jsx_runtime_1.jsx)(typedCharts_1.Line, Object.assign({ options: adaptedOptions, data: adaptedData, plugins: memoizedPlagins, className: className }, rest));
};
exports.LineChart = LineChart;
