// @geotab/zenith/dist/bulkEditControl/bulkEditControl.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TRANSLATIONS = exports.BulkEditControl = exports.BulkEditItemType = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const classNames_1 = require("../commonHelpers/classNames/classNames");
const select_1 = require("../select/select");
const useLanguage_1 = require("../utils/localization/useLanguage");
var BulkEditItemType;
(function (BulkEditItemType) {
    BulkEditItemType["SelectVisible"] = "selectVisible";
    BulkEditItemType["SelectIndividual"] = "selectIndividual";
    BulkEditItemType["SelectAllInFilter"] = "selectAllInFilter";
    BulkEditItemType["SelectNone"] = "selectNone";
})(BulkEditItemType || (exports.BulkEditItemType = BulkEditItemType = {}));
const selectionItems = [
    {
        id: BulkEditItemType.SelectVisible,
        children: "Select visible"
    },
    {
        id: BulkEditItemType.SelectIndividual,
        children: "Select individual"
    },
    {
        id: BulkEditItemType.SelectAllInFilter,
        children: "Select all"
    },
    {
        id: BulkEditItemType.SelectNone,
        children: "Select none"
    }
];
/**
 * @deprecated - will be removed in future releases
*/
const BulkEditControl = ({ id, value = BulkEditItemType.SelectNone, onChange, disabled, className = "" }) => {
    const { translate } = (0, useLanguage_1.useLanguage)();
    const translatedItems = selectionItems.map(item => (Object.assign(Object.assign({}, item), { children: translate(item.children) })));
    return (0, jsx_runtime_1.jsx)("div", { id: id, className: (0, classNames_1.classNames)(["bulk-edit-container", className || ""]), children: (0, jsx_runtime_1.jsx)(select_1.Select, { items: translatedItems, value: value, onChange: onChange, disabled: disabled, title: translate("Multiple Selection Menu") }) });
};
exports.BulkEditControl = BulkEditControl;
exports.TRANSLATIONS = [
    "Multiple Selection Menu",
    "Select visible",
    "Select individual",
    "Select all",
    "Select none"
];
