// @geotab/zenith/dist/header/headerHelpers.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.findFirstScrollableParent = exports.filterItems = exports.getItemsForRender = exports.isNamedChildComponent = exports.getArrayOfChildren = exports.HeaderCustomMenuWrapper = void 0;
const react_1 = __importStar(require("react"));
const headerTitle_1 = require("../headerTitle/headerTitle");
const tabs_1 = require("../tabs/tabs");
const filtersBar_1 = require("../filtersBar/filtersBar");
const headerButton_1 = require("./headerButton");
const headerMenu_1 = require("./headerMenu");
exports.HeaderCustomMenuWrapper = "HeaderCustomMenuWrapper";
const getArrayOfChildren = (children) => {
    if (!children) {
        return [];
    }
    const singleChildren = (childs) => Array.isArray(childs) ? childs : [childs];
    return (react_1.default.Children.count(children) === 1 ? singleChildren(children) : children);
};
exports.getArrayOfChildren = getArrayOfChildren;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const isNamedChildComponent = (item, componentName) => {
    if (!item) {
        return false;
    }
    if (!item.type) {
        return false;
    }
    const type = item.type;
    return type.displayName === componentName;
};
exports.isNamedChildComponent = isNamedChildComponent;
const getFiltersBar = (item) => {
    const children = (0, exports.getArrayOfChildren)(item.props.children);
    return children.find(child => child.type ? (0, exports.isNamedChildComponent)(child, filtersBar_1.FiltersBarDisplayName) : false);
};
const getItemsForRender = (items, isMobile) => {
    if (!isMobile) {
        return items;
    }
    return items.map((el, ind) => {
        if (el.type !== headerButton_1.HeaderButton) {
            return el;
        }
        const clone = (0, react_1.cloneElement)(el, {
            children: el.props.icon ? null : el.props.children,
            key: `hidden-key-${ind}-${items.length}`,
            title: el.props.title ? el.props.title : (typeof el.props.children === "string" ? el.props.children : "")
        });
        return clone;
    });
};
exports.getItemsForRender = getItemsForRender;
const filterItems = (items, isMobile) => {
    const itemsArr = [];
    react_1.default.Children.forEach(items, (child) => {
        if (react_1.default.isValidElement(child) && child.type === react_1.default.Fragment) {
            if (child.props && child.props.children) {
                const fragmentArr = (0, exports.getArrayOfChildren)(child.props.children);
                itemsArr.push(...fragmentArr.filter((el) => react_1.default.isValidElement(el)));
            }
        }
        else if (react_1.default.isValidElement(child)) {
            itemsArr.push(child);
        }
    });
    const result = itemsArr.reduce((acc, item) => {
        var _a, _b;
        if ((0, exports.isNamedChildComponent)(item, headerTitle_1.HeaderTitleDisplayName)) {
            acc.leftElements.push(item);
        }
        else if ((0, exports.isNamedChildComponent)(item, tabs_1.TabsDisplayName)) {
            acc.tabs = item;
        }
        else if ((0, exports.isNamedChildComponent)(item, filtersBar_1.FiltersBarDisplayName) || getFiltersBar(item)) {
            acc.filtersBar = item;
        }
        else if ((0, exports.isNamedChildComponent)(item, headerButton_1.HeaderButtonDisplayName) ||
            (0, exports.isNamedChildComponent)(item, headerMenu_1.HeaderMenuDisplayName) ||
            (0, exports.isNamedChildComponent)(item, exports.HeaderCustomMenuWrapper)) {
            const targetElements = ((_a = item.props) === null || _a === void 0 ? void 0 : _a.important) && !acc.rightElements.length ? acc.rightElements : acc.resizableElements;
            if (((_b = item.props) === null || _b === void 0 ? void 0 : _b.important) && item.type === headerButton_1.HeaderButton && isMobile) {
                const clone = (0, react_1.cloneElement)(item, {
                    children: item.props.icon ? null : item.props.children,
                    key: item.key || `key-${item.props.id}`,
                    title: item.props.title ? item.props.title : (typeof item.props.children === "string" ? item.props.children : "")
                });
                targetElements.push(clone);
            }
            else {
                const element = item.key || item.props.id ? (0, react_1.cloneElement)(item, {
                    key: item.key || `key-${item.props.id}`
                }) : item;
                targetElements.push(element);
            }
        }
        return acc;
    }, { leftElements: [], resizableElements: [], rightElements: [], tabs: null, filtersBar: null });
    let isHasPrimary = result.rightElements.length > 0;
    const preparedResizableElements = result.resizableElements.map((el) => {
        if (el.props.important || el.props.type === "primary") {
            const elClone = (0, react_1.cloneElement)(el, {
                key: el.key,
                important: false,
                type: el.props.type === "primary" && isHasPrimary ? "secondary" : el.props.type
            });
            isHasPrimary = el.props.type === "primary" ? true : isHasPrimary;
            return elClone;
        }
        return el;
    });
    return Object.assign(Object.assign({}, result), { resizableElements: [...preparedResizableElements.reverse()] });
};
exports.filterItems = filterItems;
const findFirstScrollableParent = (element) => {
    let parent = element.parentElement;
    while (parent) {
        const style = window.getComputedStyle(parent);
        if (style.overflowY === "auto" || style.overflowY === "scroll") {
            return parent;
        }
        parent = parent.parentElement;
    }
    return document.body;
};
exports.findFirstScrollableParent = findFirstScrollableParent;
