// @geotab/zenith/dist/header/header.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TRANSLATIONS = exports.HeaderDisplayName = exports.Header = exports.HeaderType = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = __importStar(require("react"));
const headerHelpers_1 = require("./headerHelpers");
const headerBack_1 = require("./headerBack");
const mobileSearchControl_1 = require("./components/mobileSearchControl/mobileSearchControl");
const mobileFilterControl_1 = require("./components/mobileFilterControl/mobileFilterControl");
const collapsedItemsControl_1 = require("./components/collapsedItemsControl/collapsedItemsControl");
const classNames_1 = require("../commonHelpers/classNames/classNames");
const basicToolbar_1 = require("./components/basicToolbar/basicToolbar");
const headerButton_1 = require("./headerButton");
const headerMenu_1 = require("./headerMenu");
const useNonFittingElements_1 = require("./hooks/useNonFittingElements");
const useResize_1 = require("../commonHelpers/hooks/useResize");
const useThrottle_1 = require("../commonHelpers/hooks/useThrottle");
const headerTitle_1 = require("../headerTitle/headerTitle");
const bookmark_1 = require("../bookmark/bookmark");
const useMobile_1 = require("../commonHelpers/hooks/useMobile");
const useResizeObserver_1 = require("../commonHelpers/hooks/useResizeObserver");
const headerContext_1 = require("./headerContext");
const filtersBarProvider_1 = require("../filtersBar/filtersBarProvider");
const filtersBarProviderSearch_1 = require("../filtersBar/filtersBarProviderSearch");
const useDriveClassName_1 = require("../utils/theme/useDriveClassName");
const filtersBarProviderTrigger_1 = require("../filtersBar/filtersBarProviderTrigger");
const headerButtonsProvider_1 = require("../utils/headerButtons/headerButtonsProvider");
var HeaderType;
(function (HeaderType) {
    HeaderType["Desktop"] = "desktop";
    HeaderType["Mobile"] = "mobile";
})(HeaderType || (exports.HeaderType = HeaderType = {}));
// eslint-disable-next-line @typescript-eslint/naming-convention, react/prop-types
const Header = ({ children, className, onFiltersBarOpen }) => {
    var _a, _b;
    const toolbarRef = (0, react_1.useRef)(null);
    const leftElementsRef = (0, react_1.useRef)(null);
    const rightElementsRef = (0, react_1.useRef)(null);
    const mobileFilterControlRef = (0, react_1.useRef)(null);
    const mobileFilterNewLineControlRef = (0, react_1.useRef)(null);
    const headerRef = (0, react_1.useRef)(null);
    const [isFilterHidden, setIsFilterHidden] = (0, react_1.useState)(false);
    const [middleWidth, setMiddleWidth] = (0, react_1.useState)(0);
    const [content, setContent] = (0, react_1.useState)(null);
    const previousScrollPosition = (0, react_1.useRef)(0);
    const [searchComponent, setSearchComponent] = (0, react_1.useState)(null);
    const [mobileFilterTrigger, setMobileFilterTrigger] = (0, react_1.useState)();
    const handleSearchChange = (comp) => {
        setSearchComponent(comp);
    };
    const [filtersBarToggleAllFiltersFn, setFiltersBarToggleAllFiltersFn] = (0, react_1.useState)(null);
    const { result } = (0, useResizeObserver_1.useResizeObserver)({ target: leftElementsRef.current });
    const moreButtonRef = (0, react_1.useRef)(null);
    const scrollableParentEl = (0, react_1.useMemo)(() => headerRef.current ? (0, headerHelpers_1.findFirstScrollableParent)(headerRef.current) : undefined, [headerRef]);
    const [searchExpandedState, setSearchExpandedState] = (0, react_1.useState)(false);
    const isMobile = (0, useMobile_1.useMobile)();
    const { buttons: additionalButtons, priority } = (0, headerButtonsProvider_1.useHeaderButtons)();
    const driveClassname = (0, useDriveClassName_1.useDriveClassName)("zen-main-header");
    const driveClassnameTab = (0, useDriveClassName_1.useDriveClassName)("zen-main-header__tabs");
    const getPreviousScrollPosition = (0, react_1.useCallback)(() => previousScrollPosition.current, []);
    const scrollCallback = (0, react_1.useCallback)((e) => {
        if (!e.target) {
            return;
        }
        const scrollThreshold = 56;
        const delta = 50;
        const scrollTop = e.target instanceof Document ? e.target.documentElement.scrollTop : e.target.scrollTop || 0;
        const previousScrollPositionValue = getPreviousScrollPosition();
        const isDownDirection = scrollTop > previousScrollPositionValue;
        if (headerRef.current && scrollTop < ((headerRef.current.getBoundingClientRect().height || scrollThreshold) + delta)) {
            previousScrollPosition.current = scrollTop;
            setIsFilterHidden(false);
            return;
        }
        if (Math.abs(scrollTop - previousScrollPositionValue) < delta) {
            previousScrollPosition.current = scrollTop;
            return;
        }
        previousScrollPosition.current = scrollTop;
        setIsFilterHidden(isDownDirection);
    }, [getPreviousScrollPosition, previousScrollPosition, headerRef]);
    const handleScroll = (0, useThrottle_1.useThrottle)((e) => {
        var _a;
        const isKeyboardLikelyOpen = ((_a = document.activeElement) === null || _a === void 0 ? void 0 : _a.matches("input, textarea, [contenteditable]")) || false;
        const isMobileFilterOpen = (mobileFilterTrigger === null || mobileFilterTrigger === void 0 ? void 0 : mobileFilterTrigger.current) && mobileFilterTrigger.current.getAttribute("aria-expanded") === "true" || false;
        if (isMobileFilterOpen || isKeyboardLikelyOpen) {
            return;
        }
        scrollCallback(e);
    }, 200);
    (0, react_1.useEffect)(() => {
        if (scrollableParentEl && isMobile) {
            scrollableParentEl.addEventListener("scroll", handleScroll);
        }
        window.addEventListener("scroll", handleScroll);
        return () => {
            if (scrollableParentEl) {
                scrollableParentEl.removeEventListener("scroll", handleScroll);
            }
            window.removeEventListener("scroll", handleScroll);
        };
    }, [isMobile, handleScroll, scrollableParentEl]);
    (0, react_1.useEffect)(() => {
        !isMobile && window.removeEventListener("scroll", handleScroll);
    }, [handleScroll, isMobile]);
    const headerItems = (0, headerHelpers_1.getArrayOfChildren)(children);
    const { leftElements, resizableElements, rightElements, tabs, filtersBar } = (0, react_1.useMemo)(() => (0, headerHelpers_1.filterItems)(priority ? [...headerItems, ...additionalButtons] : [...additionalButtons, ...headerItems], isMobile), [additionalButtons, headerItems, isMobile, priority]);
    const [numberOfChangedFilters, setNumberOfChangedFilters] = (0, react_1.useState)(undefined);
    const isOnlySearch = (0, react_1.useMemo)(() => filtersBar && searchComponent && react_1.default.Children.count(filtersBar.props.children) === 1, [filtersBar, searchComponent]);
    const memoizedFiltersBar = (0, react_1.useMemo)(() => filtersBar, [filtersBar]);
    const filtersBarToggler = (0, react_1.useCallback)(() => {
        if (onFiltersBarOpen || filtersBarToggleAllFiltersFn) {
            filtersBarToggleAllFiltersFn && filtersBarToggleAllFiltersFn.fn(true);
            !filtersBarToggleAllFiltersFn && (onFiltersBarOpen === null || onFiltersBarOpen === void 0 ? void 0 : onFiltersBarOpen());
        }
    }, [filtersBarToggleAllFiltersFn, onFiltersBarOpen]);
    const hiddenItemForRender = (0, react_1.useMemo)(() => (0, headerHelpers_1.getItemsForRender)(resizableElements, isMobile), [isMobile, resizableElements]);
    const widthDelta = ((_a = moreButtonRef.current) === null || _a === void 0 ? void 0 : _a.offsetWidth) || 0;
    const maxWidth = middleWidth || ((_b = toolbarRef.current) === null || _b === void 0 ? void 0 : _b.getBoundingClientRect().width) || 0;
    const actionCount = (0, react_1.useMemo)(() => {
        if (!isMobile) {
            return 4;
        }
        if (memoizedFiltersBar && !searchComponent) {
            return 1;
        }
        return 2;
    }, [memoizedFiltersBar, isMobile, searchComponent]);
    const hiddenElementsCountToRender = rightElements.length <= actionCount ? actionCount - rightElements.length : 0;
    const { visibleElements, nonFittingElements, hiddenContainerRef } = (0, useNonFittingElements_1.useNonFittingElements)({ maxWidth, data: resizableElements, hiddenElementsCountToRender, widthDelta, reverseOrderOfVisible: true });
    const calculateMiddleWidth = (0, react_1.useCallback)(() => {
        var _a, _b, _c;
        const headerRefWidth = ((_a = headerRef.current) === null || _a === void 0 ? void 0 : _a.getBoundingClientRect().width) || 0;
        const leftElementsRefWidth = ((_b = leftElementsRef.current) === null || _b === void 0 ? void 0 : _b.getBoundingClientRect().width) || 0;
        const rightelementsRefWidth = ((_c = rightElementsRef.current) === null || _c === void 0 ? void 0 : _c.getBoundingClientRect().width) || 0;
        const paddings = 48;
        const gaps = 8 + (rightelementsRefWidth ? 8 : 0);
        const calculatedMiddleWidth = headerRefWidth - gaps - paddings - rightelementsRefWidth - leftElementsRefWidth;
        setMiddleWidth(calculatedMiddleWidth);
    }, []);
    const resizeCallBack = (0, useThrottle_1.useThrottle)(() => {
        calculateMiddleWidth();
        if (!isMobile) {
            setSearchExpandedState(false);
        }
    }, 500);
    (0, useResize_1.useResize)(resizeCallBack, true);
    (0, react_1.useEffect)(() => {
        (result === null || result === void 0 ? void 0 : result.contentRect.width) && calculateMiddleWidth();
    }, [result === null || result === void 0 ? void 0 : result.contentRect.width, calculateMiddleWidth]);
    (0, react_1.useEffect)(() => {
        calculateMiddleWidth();
    }, [calculateMiddleWidth]);
    const moreButton = (0, react_1.useMemo)(() => (0, jsx_runtime_1.jsx)(collapsedItemsControl_1.PageToolbarCollapsedItemsControl, { children: nonFittingElements }, "pageCollapsedToolbar"), [nonFittingElements]);
    const mobileSearch = (0, react_1.useMemo)(() => searchComponent ?
        (0, jsx_runtime_1.jsx)(mobileSearchControl_1.MobileSearchControl, { state: searchExpandedState, searchInput: searchComponent, onClick: (state) => setSearchExpandedState(state) }) : null, [searchComponent, searchExpandedState]);
    const mobileFilterButton = (0, react_1.useMemo)(() => isMobile && memoizedFiltersBar ? (0, jsx_runtime_1.jsx)(mobileFilterControl_1.MobileFilterControl, { count: numberOfChangedFilters, onClick: filtersBarToggler, ref: mobileFilterControlRef }) : null, [filtersBarToggler, isMobile, memoizedFiltersBar, numberOfChangedFilters]);
    const mobileFilterNewlineButton = (0, react_1.useMemo)(() => isMobile && memoizedFiltersBar ? (0, jsx_runtime_1.jsx)(mobileFilterControl_1.MobileFilterControl, { count: numberOfChangedFilters, onClick: filtersBarToggler, ref: mobileFilterNewLineControlRef }) : null, [filtersBarToggler, isMobile, memoizedFiltersBar, numberOfChangedFilters]);
    const mobileNewLineContainer = (0, react_1.useMemo)(() => !searchComponent || isFilterHidden ? null : (0, jsx_runtime_1.jsxs)("div", { className: "zen-main-header__mobile-container-new-line", children: [(0, react_1.cloneElement)(searchComponent, { key: "searchComponentKey", className: `${searchComponent.props.className || ""} zen-main-header__mobile-search-input` }), !isOnlySearch ? mobileFilterNewlineButton : null] }), [searchComponent, isFilterHidden, mobileFilterNewlineButton, isOnlySearch]);
    const isFilterInNewLine = (0, react_1.useMemo)(() => !!searchComponent && !isOnlySearch && resizableElements.length > 0 && !searchExpandedState, [searchComponent, isOnlySearch, resizableElements.length, searchExpandedState]);
    const collapsedElements = (0, react_1.useMemo)(() => nonFittingElements.length ? [moreButton] : [], [moreButton, nonFittingElements.length]);
    const mobileSearchCondition = !resizableElements.length && mobileSearch !== null;
    const mobileFilterCondition = (0, react_1.useMemo)(() => memoizedFiltersBar && !isOnlySearch && (!resizableElements.length || resizableElements.length && !mobileSearch), [memoizedFiltersBar, isOnlySearch, resizableElements.length, mobileSearch]);
    const mobileContainer = (0, react_1.useMemo)(() => (0, jsx_runtime_1.jsxs)("div", { ref: toolbarRef, className: "zen-main-header__mobile-container", children: [collapsedElements.length ? collapsedElements : null, visibleElements.length ? (0, headerHelpers_1.getItemsForRender)(visibleElements, isMobile) : null, mobileSearchCondition ? mobileSearch : null, mobileFilterCondition && isMobile ? mobileFilterButton : null] }), [collapsedElements, isMobile, mobileFilterButton, mobileFilterCondition, mobileSearch, mobileSearchCondition, visibleElements]);
    const middleElements = (0, react_1.useMemo)(() => isMobile ? mobileContainer : (0, jsx_runtime_1.jsxs)("div", { className: "zen-main-header__adaptive-section", ref: toolbarRef, children: [collapsedElements, visibleElements] }), [collapsedElements, isMobile, mobileContainer, visibleElements]);
    const titleSize = (0, react_1.useMemo)(() => {
        let count = 0;
        if (collapsedElements.length) {
            count++;
        }
        if (visibleElements.length) {
            count = count + visibleElements.length;
        }
        if (rightElements.length) {
            count = count + rightElements.length;
        }
        if (mobileSearchCondition) {
            count++;
        }
        if (mobileFilterCondition) {
            count++;
        }
        if (count === 0 || count > 3) {
            return undefined;
        }
        const prefixesForTitleSize = ["large", "medium", "small"];
        return prefixesForTitleSize[count - 1];
    }, [collapsedElements.length, mobileFilterCondition, mobileSearchCondition, rightElements.length, visibleElements.length]);
    const titleShrinkCondition = visibleElements.length === 0 && rightElements.length || mobileSearchCondition && mobileFilterCondition;
    const fullClassName = (0, classNames_1.classNames)([
        "zen-main-header zen-main-header--background-color",
        driveClassname ? driveClassname : "",
        (memoizedFiltersBar && !isMobile) || tabs ? "" : "zen-main-header--border",
        isMobile ? "zen-main-header--mobile" : "",
        titleShrinkCondition ? "zen-main-header--shrink-title" : "",
        isMobile && titleSize ? `zen-main-header--mobile-title-${titleSize}` : "",
        className !== null && className !== void 0 ? className : ""
    ]);
    const mobileExpandedSearchWrapper = (0, react_1.useMemo)(() => (0, jsx_runtime_1.jsx)("div", { className: "zen-main-header__mobile-search", ref: toolbarRef, children: mobileSearch }), [toolbarRef, mobileSearch]);
    const mainLineContainer = (0, react_1.useMemo)(() => {
        if (searchExpandedState && !resizableElements.length) {
            return mobileExpandedSearchWrapper;
        }
        return (0, jsx_runtime_1.jsx)(basicToolbar_1.BasicToolbar, { className: "zen-main-header__section zen-main-header__section--padding", leftElements: leftElements, middleElements: middleElements, rightElements: rightElements.length ? rightElements : undefined, isMobile: isMobile, leftElementsRef: leftElementsRef, rightElementsRef: rightElementsRef });
    }, [isMobile, leftElements, middleElements, mobileExpandedSearchWrapper, resizableElements.length, rightElements, searchExpandedState]);
    const newLineContainer = (0, react_1.useMemo)(() => {
        if (!isMobile || !resizableElements.length || !mobileNewLineContainer) {
            return null;
        }
        if (searchExpandedState) {
            return mobileExpandedSearchWrapper;
        }
        return (0, jsx_runtime_1.jsx)(basicToolbar_1.BasicToolbar, { className: "zen-main-header__section zen-main-header__section--secondary-padding", leftElements: undefined, middleElements: mobileNewLineContainer, rightElements: undefined, isMobile: true });
    }, [isMobile, mobileExpandedSearchWrapper, mobileNewLineContainer, resizableElements.length, searchExpandedState]);
    const isMobileFilterButtonExists = (0, react_1.useMemo)(() => isMobile && memoizedFiltersBar && mobileFilterButton && !searchExpandedState && !isFilterHidden, [isFilterHidden, isMobile, memoizedFiltersBar, mobileFilterButton, searchExpandedState]);
    (0, react_1.useEffect)(() => {
        if (!isMobileFilterButtonExists && !mobileFilterTrigger) {
            return;
        }
        if (!isMobileFilterButtonExists) {
            setMobileFilterTrigger(undefined);
            return;
        }
        if (mobileFilterControlRef.current && !isFilterInNewLine) {
            setMobileFilterTrigger(mobileFilterControlRef);
            return;
        }
        if (mobileFilterNewLineControlRef.current) {
            setMobileFilterTrigger(mobileFilterNewLineControlRef);
            return;
        }
    }, [mobileFilterControlRef, mobileFilterNewLineControlRef, isMobileFilterButtonExists, mobileFilterTrigger, isFilterInNewLine]);
    return (0, jsx_runtime_1.jsx)(headerContext_1.HeaderMenuProvider, { alignment: "left-top", children: (0, jsx_runtime_1.jsx)(collapsedItemsControl_1.HeaderCustomContent.Provider, { value: { content, setContent }, children: (0, jsx_runtime_1.jsx)(filtersBarProvider_1.FiltersBarContextProvider, { value: { setFiltersBarToggleAllFiltersFn, onUpdateNumberOfChangedFiltersFn: setNumberOfChangedFilters }, children: (0, jsx_runtime_1.jsx)(filtersBarProviderSearch_1.SearchContextProvider, { value: { setSearchComponent: handleSearchChange }, children: (0, jsx_runtime_1.jsx)(filtersBarProviderTrigger_1.TriggerContextProvider, { value: { triggerComponent: mobileFilterTrigger }, children: (0, jsx_runtime_1.jsxs)("div", { className: fullClassName, ref: headerRef, children: [mainLineContainer, newLineContainer, tabs ? (0, jsx_runtime_1.jsxs)("div", { className: (0, classNames_1.classNames)(["zen-main-header__tabs", "zen-main-header--border", isMobile ? "zen-main-header__tabs--mobile" : "", driveClassnameTab || ""]), children: [" ", tabs, " "] }) : null, memoizedFiltersBar ? (0, jsx_runtime_1.jsx)("div", { className: "zen-main-header__filtersBar", children: memoizedFiltersBar }) : null, (0, jsx_runtime_1.jsx)("div", { className: "zen-main-header__section zen-main-header__section--hidden", ref: hiddenContainerRef, children: hiddenItemForRender }), (0, jsx_runtime_1.jsx)("div", { className: "zen-main-header__section zen-main-header__section--hidden", children: (0, jsx_runtime_1.jsx)("div", { className: "zen-main-header__section--inline", ref: moreButtonRef, children: moreButton }) })] }) }) }) }) }) });
};
exports.Header = Header;
exports.HeaderDisplayName = "Header";
exports.Header.displayName = exports.HeaderDisplayName;
exports.Header.Button = headerButton_1.HeaderButton;
exports.Header.Menu = headerMenu_1.HeaderMenu;
exports.Header.Title = headerTitle_1.HeaderTitle;
exports.Header.Bookmark = bookmark_1.Bookmark;
exports.Header.Back = headerBack_1.HeaderBack;
exports.TRANSLATIONS = [
    ...mobileSearchControl_1.TRANSLATIONS,
    ...mobileFilterControl_1.TRANSLATIONS,
    ...headerBack_1.TRANSLATIONS
];
