// @geotab/zenith/dist/dataGrid/withSortableColumns/columns/sortableColumnWrapper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SortableColumnWrapper = exports.SortableHeaderCell = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const classNames_1 = require("../../../commonHelpers/classNames/classNames");
const textIconButton_1 = require("../../../textIconButton/textIconButton");
const buttonType_1 = require("../../../button/buttonType");
const columnSortDirection_1 = require("./columnSortDirection");
const useLanguage_1 = require("../../../utils/localization/useLanguage");
const iconFilter3Desc_1 = require("../../../icons/iconFilter3Desc");
const iconFilter3Asc_1 = require("../../../icons/iconFilter3Asc");
const getSortButtonTitle = (sortDirection, title, translate) => {
    const titleTemplate = sortDirection === columnSortDirection_1.ColumnSortDirection.Descending ? "Sort by {COLUMN_NAME} descending" : "Sort by {COLUMN_NAME} ascending";
    return translate(titleTemplate).replace("{COLUMN_NAME}", title);
};
const SortableHeaderCell = ({ columnName, isSortedBy, sortDirection, title, tooltip, onSortChange, isMobile = false, initialSortDirection = columnSortDirection_1.ColumnSortDirection.Ascending }) => {
    const { translate } = (0, useLanguage_1.useLanguage)();
    const [isHovered, setIsHovered] = (0, react_1.useState)(false);
    const [isFocused, setIsFocused] = (0, react_1.useState)(false);
    const isActive = isHovered || isFocused;
    const iconTitleText = (0, react_1.useMemo)(() => {
        switch (sortDirection) {
            case columnSortDirection_1.ColumnSortDirection.Ascending: {
                return columnSortDirection_1.ColumnSortDirection.Descending;
            }
            case columnSortDirection_1.ColumnSortDirection.Descending: {
                return columnSortDirection_1.ColumnSortDirection.Ascending;
            }
            case undefined:
            default:
                return initialSortDirection;
        }
    }, [sortDirection, initialSortDirection]);
    const iconTitle = getSortButtonTitle(iconTitleText, tooltip, translate);
    const icon = (0, react_1.useMemo)(() => {
        switch (sortDirection) {
            case columnSortDirection_1.ColumnSortDirection.Ascending: {
                return isActive ? iconFilter3Desc_1.IconFilter3Desc : iconFilter3Asc_1.IconFilter3Asc;
            }
            case columnSortDirection_1.ColumnSortDirection.Descending:
                return isActive ? iconFilter3Asc_1.IconFilter3Asc : iconFilter3Desc_1.IconFilter3Desc;
            case undefined:
            default: {
                return initialSortDirection === columnSortDirection_1.ColumnSortDirection.Ascending ? iconFilter3Asc_1.IconFilter3Asc : iconFilter3Desc_1.IconFilter3Desc;
            }
        }
    }, [sortDirection, isActive, initialSortDirection]);
    const iconButton = ((0, jsx_runtime_1.jsx)(textIconButton_1.TextIconButton, { className: "zen-sortable-column-button__dir-icon zen-caption__post-icon", icon: icon, iconPosition: isMobile ? textIconButton_1.ButtonIconPosition.Start : textIconButton_1.ButtonIconPosition.End, type: buttonType_1.ButtonType.TertiaryBlack, title: iconTitle }));
    return ((0, jsx_runtime_1.jsx)("div", { className: (0, classNames_1.classNames)([
            "zen-sortable-column",
            isSortedBy ? `zen-sortable-column--active` : "",
            isMobile ? `zen-sortable-column--mobile` : ""
        ]), onMouseEnter: () => setIsHovered(true), onMouseLeave: () => setIsHovered(false), onFocus: () => setIsFocused(true), onBlur: () => setIsFocused(false), children: (0, jsx_runtime_1.jsx)("div", { className: (0, classNames_1.classNames)([
                "zen-caption",
                "zen-sortable-column__button",
                "zen-sortable-column-button",
                isSortedBy ? `zen-sortable-column-button--active` : "",
                sortDirection === columnSortDirection_1.ColumnSortDirection.Ascending
                    ? "zen-sortable-column-button--dir-asc"
                    : "zen-sortable-column-button--dir-desc"
            ]), title: iconTitle, "aria-label": iconTitle, onClick: () => {
                onSortChange({
                    sortColumn: columnName,
                    sortDirection: sortDirection === columnSortDirection_1.ColumnSortDirection.None ? initialSortDirection :
                        sortDirection === columnSortDirection_1.ColumnSortDirection.Ascending
                            ? columnSortDirection_1.ColumnSortDirection.Descending
                            : columnSortDirection_1.ColumnSortDirection.Ascending
                });
            }, children: (0, jsx_runtime_1.jsxs)(jsx_runtime_1.Fragment, { children: [isMobile ? iconButton : null, (0, jsx_runtime_1.jsx)("div", { className: "zen-sortable-column-button__content zen-caption__content", children: title }), isMobile ? null : iconButton] }) }) }));
};
exports.SortableHeaderCell = SortableHeaderCell;
class SortableColumnWrapper {
    constructor(columnName, options, columnTitle, onSort) {
        Object.defineProperty(this, "columnName", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: columnName
        });
        Object.defineProperty(this, "options", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: options
        });
        Object.defineProperty(this, "columnTitle", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: columnTitle
        });
        Object.defineProperty(this, "onSort", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: onSort
        });
    }
    renderHeader(title) {
        return (0, jsx_runtime_1.jsx)(exports.SortableHeaderCell, { columnName: this.columnName, isSortedBy: this.options.isSortedBy === undefined ? false : this.options.isSortedBy, sortDirection: this.options.sortDirection || columnSortDirection_1.ColumnSortDirection.None, title: title, tooltip: this.columnTitle, onSortChange: this.onSort, initialSortDirection: this.options.initialSortDirection });
    }
    render(value, _) {
        return value;
    }
}
exports.SortableColumnWrapper = SortableColumnWrapper;
Object.defineProperty(SortableColumnWrapper, "translations", {
    enumerable: true,
    configurable: true,
    writable: true,
    value: [
        "Sort by {COLUMN_NAME} ascending",
        "Sort by {COLUMN_NAME} descending"
    ]
});
