// @geotab/zenith/dist/dataGrid/withFlexibleColumns/flexibleColumnsStorage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mergeColumnsSettings = exports.ColumnsSettingsStorage = void 0;
const STORAGE_PREFIX = "columnsSettings_";
const createFakeSettings = () => {
    const settings = new Map();
    return {
        getItem: (key) => settings.get(key) || null,
        setItem: (key, s) => settings.set(key, s),
        removeItem: (key) => settings.delete(key)
    };
};
class ColumnsSettingsStorage {
    constructor(pageName, storage, defaultSettings) {
        Object.defineProperty(this, "storeKey", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "storage", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.storage = storage || createFakeSettings();
        this.storeKey = `${STORAGE_PREFIX}${pageName}`;
        defaultSettings && this.setItem(defaultSettings);
    }
    getItem() {
        const settings = this.storage.getItem(this.storeKey);
        try {
            return (settings ? JSON.parse(settings) : []);
        }
        catch (e) {
            this.storage.setItem(this.storeKey, "[]");
            return [];
        }
    }
    setItem(s) {
        this.storage.setItem(this.storeKey, JSON.stringify(s));
    }
    clear() {
        this.storage.removeItem(this.storeKey);
    }
}
exports.ColumnsSettingsStorage = ColumnsSettingsStorage;
const createColumnSettings = (c) => ({
    name: c.name,
    visible: c.visible,
    width: c.meta.defaultWidth
});
const findPrevColumnIndex = (columns, settings, index) => {
    if (index === 0) {
        return 0;
    }
    const name = columns[index - 1].name;
    return settings.findIndex(s => s.name === name) + 1;
};
const mergeColumnsSettings = (columns, savedSettings) => {
    const columnSettingsMap = new Set(savedSettings.map(s => s.name));
    let settings = savedSettings;
    if (settings.length === 0) {
        return columns.map(createColumnSettings);
    }
    // we should add all new columns that we don't have in settings
    for (let i = 0; i < columns.length; ++i) {
        const column = columns[i];
        if (!columnSettingsMap.has(column.name)) {
            const index = findPrevColumnIndex(columns, settings, i);
            settings = [
                ...settings.slice(0, index),
                createColumnSettings(column),
                ...settings.slice(index)
            ];
        }
    }
    return settings;
};
exports.mergeColumnsSettings = mergeColumnsSettings;
