// @geotab/zenith/dist/dataGrid/withFlexibleColumns/flexibleColumnsResizer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FlexibleColumnResizer = void 0;
const simpleEventHandler_1 = require("../../commonHelpers/simpleEventHandler");
const dataGrid_1 = require("../dataGrid");
const MIN_COLUMN_WIDTH = 100;
class FlexibleColumnResizer extends simpleEventHandler_1.SimpleEventHandler {
    createResizer() {
        if (this.resizer === undefined) {
            const resizer = document.createElement("DIV");
            resizer.classList.add("zen-resizer", "zen-resizer--active", "zen-resizer--vertical");
            this.resizer = resizer;
        }
        this.listRef.append(this.resizer);
        return this.resizer;
    }
    constructor(listRef) {
        super();
        Object.defineProperty(this, "listRef", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: listRef
        });
        Object.defineProperty(this, "currentWidth", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: 0
        });
        Object.defineProperty(this, "pageX", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: 0
        });
        Object.defineProperty(this, "offsetLeft", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: 0
        });
        Object.defineProperty(this, "resizer", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "columnIndex", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: 0
        });
        Object.defineProperty(this, "resizers", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: []
        });
        Object.defineProperty(this, "columns", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: []
        });
        Object.defineProperty(this, "preventScroll", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: (e) => {
                e.preventDefault();
                return false;
            }
        });
        Object.defineProperty(this, "stopResizing", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: () => {
                this.pageX = 0;
                this.currentWidth = 0;
                this.columnIndex = 0;
                document.removeEventListener("pointermove", this.resizeColumn);
                document.removeEventListener("pointerup", this.endColumn);
                document.removeEventListener("pointercancel", this.stopResizing);
                this.listRef.classList.remove(dataGrid_1.NOT_INTERACTIVE_MODIFIER);
                this.hideResizer();
            }
        });
        Object.defineProperty(this, "resizeColumn", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: (e) => {
                const shift = e.pageX - this.pageX;
                const width = this.currentWidth + shift;
                if (width < MIN_COLUMN_WIDTH || this.resizer === undefined) {
                    return;
                }
                this.resizer.style.left = `${this.offsetLeft + shift}px`;
            }
        });
        Object.defineProperty(this, "endColumn", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: (e) => {
                const shift = e.pageX - this.pageX;
                this.currentWidth = Math.max(this.currentWidth + shift, MIN_COLUMN_WIDTH);
                this.fire("resize", [this.columnIndex, this.currentWidth]);
                this.stopResizing();
            }
        });
        Object.defineProperty(this, "startColumn", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: (e) => {
                var _a;
                if (e.button !== 0) {
                    return;
                }
                this.pageX = e.pageX;
                const th = e.target.closest("th");
                if (th === null) {
                    return;
                }
                this.offsetLeft = th.offsetLeft;
                const height = this.listRef.clientHeight;
                this.columnIndex = Math.max(th.cellIndex - 1, 0);
                const columnId = this.columns[this.columnIndex].id;
                const headerEl = (_a = th.parentElement) === null || _a === void 0 ? void 0 : _a.querySelector(`[data-column-id="${columnId}"]`);
                if (!headerEl) {
                    return;
                }
                this.currentWidth = headerEl.offsetWidth || 0;
                document.addEventListener("pointermove", this.resizeColumn);
                document.addEventListener("pointerup", this.endColumn);
                document.addEventListener("pointercancel", this.stopResizing);
                this.listRef.classList.add(dataGrid_1.NOT_INTERACTIVE_MODIFIER);
                const scrollTop = this.listRef.scrollTop;
                const resizer = this.createResizer();
                resizer.style.left = `${this.offsetLeft}px`;
                resizer.style.top = `${scrollTop}px`;
                resizer.style.height = `${height}px`;
            }
        });
    }
    showResizerForColumn(columnIndex) {
        const header = this.listRef.querySelector("thead > tr");
        const th = header.children[columnIndex + 1];
        if (th === undefined) {
            return;
        }
        this.offsetLeft = th.offsetLeft;
        const height = this.listRef.clientHeight;
        const scrollTop = this.listRef.scrollTop;
        const resizer = this.createResizer();
        resizer.style.left = `${this.offsetLeft}px`;
        resizer.style.top = `${scrollTop}px`;
        resizer.style.height = `${height}px`;
    }
    hideResizer() {
        this.resizer && this.resizer.remove();
    }
    add(columns) {
        this.remove();
        this.columns = columns;
        const header = this.listRef.querySelector("thead");
        if (header === null) {
            return;
        }
        this.resizers = Array.from(header.getElementsByClassName("zen-flexible-column__draggable"));
        this.resizers.forEach(elt => {
            elt.addEventListener("pointerdown", this.startColumn);
            elt.addEventListener("touchstart", this.preventScroll);
        });
    }
    remove() {
        this.resizers.forEach(elt => {
            elt.removeEventListener("pointerdown", this.startColumn);
            elt.removeEventListener("touchstart", this.preventScroll);
        });
        this.resizers.length = 0;
        this.hideResizer();
    }
}
exports.FlexibleColumnResizer = FlexibleColumnResizer;
