// @geotab/zenith/dist/dataGrid/withFlexibleColumns/flexibleColumnsReorder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FlexibleColumnReorder = void 0;
const simpleEventHandler_1 = require("../../commonHelpers/simpleEventHandler");
const dataGrid_1 = require("../dataGrid");
class FlexibleColumnReorder extends simpleEventHandler_1.SimpleEventHandler {
    createColumnToDrag() {
        if (this.dragElt === undefined) {
            this.dragElt = document.createElement("table");
            this.dragElt.classList.add("zen-table-column");
        }
        this.list.append(this.dragElt);
        return this.dragElt;
    }
    cloneCell(columnCell) {
        let cell = columnCell;
        const cellPart = columnCell.querySelector("[data-column]");
        if (cellPart !== null) {
            cell = document.createElement("td");
            cell.classList.add("zen-data-grid__row-cell");
            cell.append(cellPart.cloneNode(true));
            return cell;
        }
        return cell.cloneNode(true);
    }
    cloneCellContent(cell, parent) {
        const clone = this.cloneCell(cell);
        // we should prevent id duplication
        const eltsWithIds = clone.querySelectorAll("[id]");
        eltsWithIds.forEach(elt => {
            elt.id = "";
        });
        const p = document.createElement("tr");
        p.classList.add("zen-table-column__row");
        p.append(clone);
        parent.append(p);
    }
    getFirstDraggable() {
        for (let i = 0; i < this.columnWidths.length; i++) {
            if (this.isDraggable(i)) {
                return i;
            }
        }
        return -1;
    }
    getLastDraggable() {
        for (let i = this.columnWidths.length - 1; i >= 0; --i) {
            if (this.isDraggable(i)) {
                return i;
            }
        }
        return -1;
    }
    calculateClosestColumn(e) {
        const shiftX = e.pageX - this.pageX;
        let distanceFromTableStart = this.offsetLeft + shiftX + this.pointerOffset;
        const firstDraggable = this.getFirstDraggable();
        const lastDraggable = this.getLastDraggable();
        let closest = lastDraggable;
        for (let i = 0; i < this.columnWidths.length; ++i) {
            const w = this.columnWidths[i];
            if (distanceFromTableStart < w) {
                if (i < firstDraggable) {
                    closest = firstDraggable;
                    break;
                }
                if (i > lastDraggable) {
                    closest = lastDraggable + 1;
                    break;
                }
                const half = w / 2;
                if (distanceFromTableStart < half) {
                    closest = i;
                }
                else {
                    closest = i + 1;
                }
                break;
            }
            distanceFromTableStart -= w;
        }
        return closest;
    }
    constructor(list, resizer, isDraggable) {
        super();
        Object.defineProperty(this, "list", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: list
        });
        Object.defineProperty(this, "resizer", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: resizer
        });
        Object.defineProperty(this, "isDraggable", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: isDraggable
        });
        Object.defineProperty(this, "pageX", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: 0
        });
        Object.defineProperty(this, "pageY", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: 0
        });
        Object.defineProperty(this, "offsetLeft", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: 0
        });
        Object.defineProperty(this, "offsetTop", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: 0
        });
        Object.defineProperty(this, "pointerOffset", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: 0
        });
        Object.defineProperty(this, "dragElt", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "targetCellElt", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "closest", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: -1
        });
        Object.defineProperty(this, "columnWidths", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: []
        });
        Object.defineProperty(this, "columnIndex", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: 0
        });
        Object.defineProperty(this, "draggables", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: []
        });
        Object.defineProperty(this, "preventScroll", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: (e) => {
                e.preventDefault();
                return false;
            }
        });
        Object.defineProperty(this, "stopMoving", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: () => {
                this.pageX = 0;
                this.pageY = 0;
                this.columnIndex = 0;
                this.pointerOffset = 0;
                this.closest = -1;
                document.removeEventListener("pointermove", this.moveColumn);
                document.removeEventListener("pointerup", this.endColumn);
                document.removeEventListener("pointercancel", this.stopMoving);
                this.list.classList.remove(dataGrid_1.NOT_INTERACTIVE_MODIFIER);
                this.list.classList.remove(dataGrid_1.DRAGGING_MODIFIER);
                if (this.dragElt === undefined) {
                    return;
                }
                this.dragElt.remove();
                this.dragElt.innerHTML = "";
                this.dragElt = undefined;
                this.targetCellElt = undefined;
                this.resizer.hideResizer();
            }
        });
        Object.defineProperty(this, "moveColumn", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: (e) => {
                const shiftX = e.pageX - this.pageX;
                const shiftY = Math.max(0, e.pageY - this.pageY);
                const isThresholdReached = Math.abs(shiftX) >= 10 || Math.abs(shiftY) >= 10;
                if (!isThresholdReached && !this.dragElt) {
                    return;
                }
                const dragElt = this.dragElt || this.createDragElement();
                if (!dragElt) {
                    return;
                }
                dragElt.style.left = `${this.offsetLeft + shiftX}px`;
                dragElt.style.top = `${this.offsetTop + shiftY}px`;
                const closest = this.calculateClosestColumn(e);
                if (closest === this.closest) {
                    return;
                }
                this.closest = closest;
                this.resizer.showResizerForColumn(closest - 1);
            }
        });
        Object.defineProperty(this, "endColumn", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: () => {
                if (this.columnIndex !== this.closest && this.closest >= 0) {
                    this.fire("reorder", [this.columnIndex, this.closest]);
                }
                this.stopMoving();
            }
        });
        Object.defineProperty(this, "createDragElement", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: () => {
                var _a;
                if (!this.targetCellElt) {
                    return undefined;
                }
                const dragElt = this.createColumnToDrag();
                dragElt.style.left = `${this.offsetLeft}px`;
                dragElt.style.top = `${this.offsetTop}px`;
                dragElt.style.width = `${this.targetCellElt.offsetWidth}px`;
                this.resizer.showResizerForColumn(this.closest - 1);
                this.cloneCellContent(this.targetCellElt, dragElt);
                this.list.classList.add(dataGrid_1.NOT_INTERACTIVE_MODIFIER);
                this.list.classList.add(dataGrid_1.DRAGGING_MODIFIER);
                const columns = ((_a = this.targetCellElt.parentElement) === null || _a === void 0 ? void 0 : _a.querySelectorAll("th")) || [];
                this.columnWidths = [...columns].map(c => c.offsetWidth);
                return dragElt;
            }
        });
        Object.defineProperty(this, "startColumn", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: (e) => {
                var _a;
                if (e.button !== 0) {
                    return;
                }
                this.pageX = e.pageX;
                this.pageY = e.pageY;
                (_a = window.getSelection()) === null || _a === void 0 ? void 0 : _a.removeAllRanges();
                const th = e.target.closest("th");
                if (th === null) {
                    return;
                }
                this.targetCellElt = th;
                this.offsetLeft = th.offsetLeft;
                this.offsetTop = th.offsetTop;
                this.columnIndex = Math.max(th.cellIndex, 0);
                this.pointerOffset = (e.offsetX || (th.clientWidth / 2));
                document.addEventListener("pointermove", this.moveColumn);
                document.addEventListener("pointerup", this.endColumn);
                document.addEventListener("pointercancel", this.stopMoving);
                this.closest = this.columnIndex;
            }
        });
    }
    add() {
        this.remove();
        this.draggables = Array.from(this.list.querySelectorAll(".zen-data-grid__header .zen-flexible-column--draggable"));
        this.draggables.forEach(elt => {
            elt.addEventListener("pointerdown", this.startColumn);
            elt.addEventListener("touchstart", this.preventScroll);
        });
    }
    remove() {
        this.stopMoving();
        this.draggables.forEach(elt => {
            elt.removeEventListener("pointerdown", this.startColumn);
            elt.removeEventListener("touchstart", this.preventScroll);
        });
        this.draggables.length = 0;
    }
}
exports.FlexibleColumnReorder = FlexibleColumnReorder;
