// @geotab/zenith/dist/dataGrid/columns/actionsColumn/actionsButton.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActionsButton = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const actionsColumnCell_1 = require("./actionsColumnCell");
const controlledMenu_1 = require("../../../menu/controlledMenu");
function createSubAction(action, entity) {
    const href = action.createLink === undefined ? undefined : action.createLink(entity);
    const title = action.createTitle(entity);
    const handler = action.handler;
    return (0, react_1.createElement)(controlledMenu_1.ControlledMenu.Item, {
        id: action.id,
        key: action.id,
        name: title,
        link: href,
        icon: action.icon,
        onClick: handler
            ? (_1, e) => {
                handler(entity, e.target);
            }
            : undefined
    });
}
function ActionsButton({ action, entity }) {
    const [isMenuOpen, setIsMenuOpen] = (0, react_1.useState)(false);
    const buttonRef = (0, react_1.useRef)(null);
    const entityId = (entity === null || entity === void 0 ? void 0 : entity.id) || "";
    if (!entity) {
        return null;
    }
    const items = (action.actions || []).filter(a => a.condition ? a.condition(entity) : true).map(a => createSubAction(a, entity));
    const isTriggerVisible = action.condition ? action.condition(entity) : true;
    if (!isTriggerVisible) {
        return null;
    }
    return (0, jsx_runtime_1.jsxs)(jsx_runtime_1.Fragment, { children: [(0, jsx_runtime_1.jsx)(actionsColumnCell_1.ActionColumnCell, { ref: buttonRef, tag: action.createLink !== undefined ? "a" : "button", hash: action.id, title: action.createTitle(entity), href: action.createLink !== undefined ? action.createLink(entity) : undefined, icon: action.icon, onClick: (e) => {
                    setIsMenuOpen(!isMenuOpen);
                    action.handler && action.handler(entity, e.target);
                } }, action.id), items.length
                ? (0, jsx_runtime_1.jsx)(controlledMenu_1.ControlledMenu, { triggerRef: buttonRef, alignment: "bottom-right", isOpen: isMenuOpen, setIsOpen: setIsMenuOpen, title: action.createTitle(entity), children: items }, entityId + "-" + action.id + "-menu")
                : null] });
}
exports.ActionsButton = ActionsButton;
