// @geotab/zenith/dist/dataGrid/columns/actionsColumn/actionsColumn.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActionsColumn = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const basicColumn_1 = require("../basicColumn");
const generateId_1 = require("../../../commonHelpers/generateId");
const actionsButton_1 = require("./actionsButton");
const actionsMenu_1 = require("../../../table/actions/actionsMenu");
class ActionsColumn extends basicColumn_1.BasicColumn {
    renderIconActions(entity, actions, isNested) {
        const entityActions = typeof actions === "function"
            ? entity ? actions(entity, isNested || false).filter(action => action.condition === undefined || action.condition(entity)) : []
            : actions.filter(action => action.condition === undefined || action.condition(entity));
        const entityId = (entity === null || entity === void 0 ? void 0 : entity.id) || "";
        return entityActions.length > 0 ? (0, jsx_runtime_1.jsx)("div", { className: "zen-actions-column", children: entityActions.map(action => (0, jsx_runtime_1.jsx)(actionsButton_1.ActionsButton, { action: action, entity: entity }, action.id)) }, [entityId, ...entityActions.map(a => a.id)].join("-")) : null;
    }
    updateActions(options) {
        this.actions = options.actions ? options.actions.map(action => { var _a; return (Object.assign(Object.assign({}, action), { id: action.id || (0, generateId_1.generateId)(), actions: (_a = action.actions) === null || _a === void 0 ? void 0 : _a.map(a => (Object.assign(Object.assign({}, a), { id: (0, generateId_1.generateId)() }))) })); }) : undefined;
        this.renderFn = options.render;
        this.renderAsyncFn = options.renderAsync;
        this.headerActions = (options.headerActions || []).map(action => { var _a; return (Object.assign(Object.assign({}, action), { id: action.id || (0, generateId_1.generateId)(), actions: (_a = action.actions) === null || _a === void 0 ? void 0 : _a.map(a => (Object.assign(Object.assign({}, a), { id: (0, generateId_1.generateId)() }))) })); });
        if (typeof options.primaryActions === "function") {
            this.primaryActions = options.primaryActions;
        }
        else {
            this.primaryActions = (options.primaryActions || []).map(action => { var _a; return (Object.assign(Object.assign({}, action), { id: action.id || (0, generateId_1.generateId)(), actions: (_a = action.actions) === null || _a === void 0 ? void 0 : _a.map(a => (Object.assign(Object.assign({}, a), { id: (0, generateId_1.generateId)() }))) })); });
        }
    }
    renderCell(_, entity, isNested) {
        if (this.options.render || this.options.renderAsync) {
            return (0, jsx_runtime_1.jsxs)("div", { className: "zen-actions-column", children: [this.primaryActions && (typeof this.primaryActions === "function" || this.primaryActions.length > 0) && this.renderIconActions(entity, this.primaryActions, isNested || false), this.options.render && (0, jsx_runtime_1.jsx)(actionsMenu_1.ActionsMenu, { entity: entity, isNested: isNested || false, render: this.options.render }), this.options.renderAsync && (0, jsx_runtime_1.jsx)(actionsMenu_1.ActionsMenu, { entity: entity, isNested: isNested || false, renderAsync: this.options.renderAsync })] });
        }
        return this.renderIconActions(entity, (this.actions || []));
    }
    renderHeaderCell(_) {
        return this.renderIconActions(undefined, this.headerActions);
    }
    constructor(actionsOptions) {
        super({ render: () => "" });
        Object.defineProperty(this, "actionsOptions", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: actionsOptions
        });
        Object.defineProperty(this, "actions", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "headerActions", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: []
        });
        Object.defineProperty(this, "primaryActions", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "renderFn", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "renderAsyncFn", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.updateActions(actionsOptions);
    }
    get options() {
        return this.actionsOptions;
    }
    set options(actionsOptions) {
        this.actionsOptions = actionsOptions;
        this.updateActions(actionsOptions);
    }
    get columnActions() {
        return (this.actions ? this.actions[0] : undefined) || this.renderFn;
    }
    get columnActionsAsync() {
        return this.renderAsyncFn || undefined;
    }
    get columnPrimaryActions() {
        return this.primaryActions || undefined;
    }
}
exports.ActionsColumn = ActionsColumn;
