// @geotab/zenith/dist/utils/compareElementsUtils/isElementsEqual.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isElementsEqual = exports.isPropertiesEqual = exports.isObjectsNotEqualShallow = exports.isPrimitivesNotEqual = exports.isArraysNotEqual = exports.isBothObjectsShallow = exports.isBothFunctions = exports.isBothArrays = exports.isBothObjects = void 0;
const isArraysEqualDeeply_1 = require("./isArraysEqualDeeply");
const isArraysEqualShallow_1 = require("./isArraysEqualShallow");
const isBothArraysShallow_1 = require("./isBothArraysShallow");
const isObjectNested_1 = require("./isObjectNested");
const isObjectsEqualShallow_1 = require("./isObjectsEqualShallow");
const isPrimitive_1 = require("./isPrimitive");
const isBothObjects = (elem1, elem2) => typeof elem1 === "object" && typeof elem2 === "object" && elem1 !== null && elem2 !== null;
exports.isBothObjects = isBothObjects;
const isBothArrays = (elem1, elem2) => Array.isArray(elem1) && Array.isArray(elem2);
exports.isBothArrays = isBothArrays;
const isBothFunctions = (elem1, elem2) => typeof elem1 === "function" && typeof elem2 === "function";
exports.isBothFunctions = isBothFunctions;
const isBothObjectsShallow = (elem1, elem2) => !(0, isObjectNested_1.isObjectNested)(elem1) && !(0, isObjectNested_1.isObjectNested)(elem2);
exports.isBothObjectsShallow = isBothObjectsShallow;
const isArraysNotEqual = (val1, val2) => {
    if (!(0, exports.isBothArrays)(val1, val2)) {
        return false;
    }
    if ((0, isBothArraysShallow_1.isBothArraysShallow)(val1, val2) && !(0, isArraysEqualShallow_1.isArraysEqualShallow)(val1, val2)) {
        return true;
    }
    if (!(0, isArraysEqualDeeply_1.isArraysEqualDeeply)(val1, val2)) {
        return true;
    }
    return false;
};
exports.isArraysNotEqual = isArraysNotEqual;
const isPrimitivesNotEqual = (item1, item2) => (0, isPrimitive_1.isPrimitive)(item1) && (0, isPrimitive_1.isPrimitive)(item2) && item1 !== item2;
exports.isPrimitivesNotEqual = isPrimitivesNotEqual;
const isObjectsNotEqualShallow = (val1, val2) => {
    if (!(0, exports.isBothObjects)(val1, val2)) {
        return false;
    }
    if ((0, exports.isBothObjectsShallow)(val1, val2) && !(0, isObjectsEqualShallow_1.isObjectsEqualShallow)(val1, val2)) {
        return true;
    }
    return false;
};
exports.isObjectsNotEqualShallow = isObjectsNotEqualShallow;
const isPropertiesEqual = (elem1, elem2) => {
    const elem1Keys = Object.keys(elem1);
    for (const key of elem1Keys) {
        if ((0, exports.isPrimitivesNotEqual)(elem1[key], elem2[key])) {
            return false;
        }
        if ((0, exports.isArraysNotEqual)(elem1[key], elem2[key])) {
            return false;
        }
        if ((0, exports.isObjectsNotEqualShallow)(elem1[key], elem2[key])) {
            return false;
        }
        // eslint-disable-next-line @typescript-eslint/no-use-before-define
        if (!isElementsEqual(elem1[key], elem2[key])) {
            return false;
        }
    }
    return true;
};
exports.isPropertiesEqual = isPropertiesEqual;
function isElementsEqual(...args) {
    const elem1 = args[0];
    const elem2 = args[1];
    // compare dates
    if (elem1 instanceof Date && elem2 instanceof Date) {
        return elem1.getTime() === elem2.getTime();
    }
    // compare primitives
    if ((0, isPrimitive_1.isPrimitive)(elem1) && (0, isPrimitive_1.isPrimitive)(elem2)) {
        return elem1 === elem2;
    }
    // is the same function
    if ((0, exports.isBothFunctions)(elem1, elem2)) {
        return elem1.toString() === elem2.toString();
    }
    // check {} and [] case
    if (Array.isArray(elem1) !== Array.isArray(elem2)) {
        return false;
    }
    // check 'function' and 'object' case
    if (typeof elem1 !== typeof elem2) {
        return false;
    }
    // is the same object
    if (elem1 === elem2) {
        return true;
    }
    if ((0, isBothArraysShallow_1.isBothArraysShallow)(elem1, elem2)) {
        return (0, isArraysEqualShallow_1.isArraysEqualShallow)(elem1, elem2);
    }
    if ((0, exports.isBothObjectsShallow)(elem1, elem2)) {
        return (0, isObjectsEqualShallow_1.isObjectsEqualShallow)(elem1, elem2);
    }
    return (0, exports.isPropertiesEqual)(elem1, elem2);
}
exports.isElementsEqual = isElementsEqual;
