// @geotab/zenith/dist/react-chartjs/dateAdapter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addDate = exports.getEndOf = exports.getStartOf = exports.getDiff = exports.formatDate = exports.parseDate = exports.dateFormats = exports.FORMATS = void 0;
const auto_1 = require("chart.js/auto");
const dateFns_1 = require("./dateFns/dateFns");
exports.FORMATS = {
    datetime: "MMM d, yyyy, h:mm:ss aaaa",
    millisecond: "h:mm:ss.SSS aaaa",
    second: "h:mm:ss aaaa",
    minute: "h:mm aaaa",
    hour: "ha",
    day: "MMM d",
    week: "PP",
    month: "MMM yyyy",
    quarter: "qqq - yyyy",
    year: "yyyy"
};
const dateFormats = () => exports.FORMATS;
exports.dateFormats = dateFormats;
const parseDate = (value) => value;
exports.parseDate = parseDate;
const formatDate = (time, fmt) => (0, dateFns_1.format)(time, fmt);
exports.formatDate = formatDate;
const getDiff = () => 0;
exports.getDiff = getDiff;
const getStartOf = (time) => time;
exports.getStartOf = getStartOf;
const getEndOf = (time) => time;
exports.getEndOf = getEndOf;
const addDate = (time, amount, unit) => {
    switch (unit) {
        case "millisecond": return (0, dateFns_1.addMilliseconds)(time, amount).valueOf();
        case "second": return (0, dateFns_1.addSeconds)(time, amount).valueOf();
        case "minute": return (0, dateFns_1.addMinutes)(time, amount).valueOf();
        case "hour": return (0, dateFns_1.addHours)(time, amount).valueOf();
        case "day": return (0, dateFns_1.addDays)(time, amount).valueOf();
        case "week": return (0, dateFns_1.addWeeks)(time, amount).valueOf();
        case "month": return (0, dateFns_1.addMonths)(time, amount).valueOf();
        case "quarter": return (0, dateFns_1.addQuarters)(time, amount).valueOf();
        case "year": return (0, dateFns_1.addYears)(time, amount).valueOf();
        default: return new Date(time).valueOf();
    }
};
exports.addDate = addDate;
// eslint-disable-next-line no-underscore-dangle
auto_1._adapters._date.override({
    formats: exports.dateFormats,
    parse: exports.parseDate,
    //todo third arg cab be this.options
    format: exports.formatDate,
    add: exports.addDate,
    diff: exports.getDiff,
    startOf: exports.getStartOf,
    endOf: exports.getEndOf
});
