// @geotab/zenith/dist/react-chartjs/typedCharts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Scatter = exports.Pie = exports.Bubble = exports.PolarArea = exports.Doughnut = exports.Radar = exports.Bar = exports.Line = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const auto_1 = require("chart.js/auto");
const chart_1 = require("./chart");
function createTypedChart(type, registerables) {
    auto_1.Chart.register(registerables);
    const TypedChart = (0, react_1.forwardRef)((props, ref) => (0, jsx_runtime_1.jsx)(chart_1.Chart, Object.assign({}, props, { ref: ref, type: type })));
    TypedChart.displayName = `TypedChart(${type})`;
    return TypedChart;
}
exports.Line = createTypedChart("line", auto_1.LineController);
exports.Bar = createTypedChart("bar", auto_1.BarController);
exports.Radar = createTypedChart("radar", auto_1.RadarController);
exports.Doughnut = createTypedChart("doughnut", auto_1.DoughnutController);
exports.PolarArea = createTypedChart("polarArea", auto_1.PolarAreaController);
exports.Bubble = createTypedChart("bubble", auto_1.BubbleController);
exports.Pie = createTypedChart("pie", auto_1.PieController);
exports.Scatter = createTypedChart("scatter", auto_1.ScatterController);
