// @geotab/zenith/dist/react-chartjs/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getElementsAtEvent = exports.getElementAtEvent = exports.getDatasetAtEvent = exports.cloneData = exports.setDatasets = exports.setLabels = exports.setOptions = exports.reforwardRef = void 0;
const defaultDatasetIdKey = "label";
function reforwardRef(ref, value) {
    if (typeof ref === "function") {
        ref(value);
    }
    else if (ref) {
        ref.current = value;
    }
}
exports.reforwardRef = reforwardRef;
function setOptions(chart, nextOptions) {
    const options = chart.options;
    if (options && nextOptions) {
        Object.assign(options, nextOptions);
    }
}
exports.setOptions = setOptions;
function setLabels(currentData, nextLabels) {
    currentData.labels = nextLabels;
}
exports.setLabels = setLabels;
function setDatasets(currentData, nextDatasets, datasetIdKey = defaultDatasetIdKey) {
    const addedDatasets = [];
    currentData.datasets = nextDatasets.map((nextDataset) => {
        // given the new set, find it's current match
        const currentDataset = currentData.datasets.find((dataset) => dataset[datasetIdKey] === nextDataset[datasetIdKey]);
        // There is no original to update, so simply add new one
        if (!currentDataset ||
            !nextDataset.data ||
            addedDatasets.includes(currentDataset)) {
            return Object.assign({}, nextDataset);
        }
        addedDatasets.push(currentDataset);
        Object.assign(currentDataset, nextDataset);
        return currentDataset;
    });
}
exports.setDatasets = setDatasets;
function cloneData(data, datasetIdKey = defaultDatasetIdKey) {
    const nextData = {
        labels: [],
        datasets: []
    };
    setLabels(nextData, data.labels);
    setDatasets(nextData, data.datasets, datasetIdKey);
    return nextData;
}
exports.cloneData = cloneData;
/**
 * Get dataset from mouse click event
 * @param chart - Chart.js instance
 * @param event - Mouse click event
 * @returns Dataset
 */
function getDatasetAtEvent(chart, event) {
    return chart.getElementsAtEventForMode(event.nativeEvent, "dataset", { intersect: true }, false);
}
exports.getDatasetAtEvent = getDatasetAtEvent;
/**
 * Get single dataset element from mouse click event
 * @param chart - Chart.js instance
 * @param event - Mouse click event
 * @returns Dataset
 */
function getElementAtEvent(chart, event) {
    return chart.getElementsAtEventForMode(event.nativeEvent, "nearest", { intersect: true }, false);
}
exports.getElementAtEvent = getElementAtEvent;
/**
 * Get all dataset elements from mouse click event
 * @param chart - Chart.js instance
 * @param event - Mouse click event
 * @returns Dataset
 */
function getElementsAtEvent(chart, event) {
    return chart.getElementsAtEventForMode(event.nativeEvent, "index", { intersect: true }, false);
}
exports.getElementsAtEvent = getElementsAtEvent;
