// @geotab/zenith/dist/filters/components/filtersSelect.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FiltersSelect = exports.getAllOption = exports.getFiltersSearchTypeAllOption = exports.FiltersSelectDisplayName = exports.ALL_SELECT_OPTION_ID = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const classNames_1 = require("../../commonHelpers/classNames/classNames");
const useLanguage_1 = require("../../utils/localization/useLanguage");
const useUniqueId_1 = require("../../commonHelpers/useUniqueId");
const filtersSelectListItem_1 = require("./filtersSelectListItem");
exports.ALL_SELECT_OPTION_ID = "All";
exports.FiltersSelectDisplayName = "FiltersSelect";
const getFiltersSearchTypeAllOption = () => ({
    [exports.ALL_SELECT_OPTION_ID]: undefined
});
exports.getFiltersSearchTypeAllOption = getFiltersSearchTypeAllOption;
const getAllOption = (translate) => ({
    id: exports.ALL_SELECT_OPTION_ID,
    name: translate("All")
});
exports.getAllOption = getAllOption;
const FiltersSelect = ({ value, name, onChange, id, items, className }) => {
    const { translate } = (0, useLanguage_1.useLanguage)();
    const [activeId, setActiveId] = (0, react_1.useState)("");
    const filtersSelectListId = (0, useUniqueId_1.useUniqueId)();
    const filteredItems = (0, react_1.useMemo)(() => (items === null || items === void 0 ? void 0 : items.filter(el => el.id !== exports.ALL_SELECT_OPTION_ID)) || [], [items]);
    const preparedItems = (0, react_1.useMemo)(() => [
        (0, exports.getAllOption)(translate),
        ...filteredItems
    ], [translate, filteredItems]);
    const handleExpandableItemClick = (0, react_1.useCallback)((clickedId) => () => {
        if (activeId === clickedId) {
            setActiveId("");
            return;
        }
        setActiveId(clickedId);
        onChange({ [clickedId]: undefined });
    }, [activeId, onChange]);
    const handleItemClick = (0, react_1.useCallback)((clickedId, parentId) => () => {
        if (parentId) {
            onChange({ [parentId]: clickedId });
            return;
        }
        setActiveId("");
        onChange(Object.keys(value).includes(clickedId) ? (0, exports.getFiltersSearchTypeAllOption)() : { [clickedId]: undefined });
    }, [onChange, value]);
    const selectListComponent = (0, react_1.useMemo)(() => preparedItems.map(el => el.subItems ?
        (0, jsx_runtime_1.jsx)(filtersSelectListItem_1.FiltersSelectListExpandableItem, { id: `${filtersSelectListId}-${el.id}`, name: el.name, isActive: el.id === activeId, className: (0, classNames_1.classNames)([value.hasOwnProperty(el.id) ? "zen-filters-select__list-item--active" : ""]), onClick: handleExpandableItemClick(el.id), content: el.id === activeId ? el.subItems.map(subEl => (0, jsx_runtime_1.jsx)(filtersSelectListItem_1.FiltersSelectListItem, { className: (0, classNames_1.classNames)(["zen-filters-select__sub-item"]), id: `${filtersSelectListId}-${el.id}-${subEl.id}`, name: subEl.name, isActive: value[el.id] === subEl.id, onClick: handleItemClick(subEl.id, el.id) }, `${el.id}-${subEl.id}`)) : (0, jsx_runtime_1.jsx)(jsx_runtime_1.Fragment, {}) }, el.id) : (0, jsx_runtime_1.jsx)(filtersSelectListItem_1.FiltersSelectListItem, { id: `${filtersSelectListId}-${el.id}`, name: el.name, isActive: Object.keys(value).includes(el.id), onClick: handleItemClick(el.id) }, el.id)), [activeId, filtersSelectListId, handleExpandableItemClick, handleItemClick, preparedItems, value]);
    return (0, jsx_runtime_1.jsxs)("div", { className: "zen-filters-select", children: [id && name ? (0, jsx_runtime_1.jsx)("div", { className: "zen-filters-select__label", id: id, children: name }) : null, (0, jsx_runtime_1.jsx)("div", { className: (0, classNames_1.classNames)(["zen-filters-select__list", className || ""]), children: selectListComponent })] });
};
exports.FiltersSelect = FiltersSelect;
exports.FiltersSelect.displayName = exports.FiltersSelectDisplayName;
