// @geotab/zenith/dist/dataGrid/columns/basicColumn.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BasicColumn = exports.isPromiseLike = void 0;
const isPromiseLike = (value) => !!value && typeof value.then === "function" && typeof value.catch === "function";
exports.isPromiseLike = isPromiseLike;
class BasicColumn {
    getValue(entity, isNested) {
        if (this.column.render !== undefined) {
            // eslint-disable-next-line @typescript-eslint/no-unsafe-argument, @typescript-eslint/no-explicit-any
            return this.column.render(entity, isNested || false);
        }
        if (this.column.name) {
            return entity[this.column.name];
        }
        return this.column.defaultValue;
    }
    getHeaderTitle(title) {
        return this.column.renderHeader !== undefined
            ? this.column.renderHeader(title)
            : title;
    }
    renderCellAsync(entity, isNested, render) {
        const result = this.getValue(entity, isNested);
        if ((0, exports.isPromiseLike)(result)) {
            return result.then(r => r === undefined ? r : render(r, entity, isNested));
        }
        return render(result, entity, isNested);
    }
    renderHeaderCell(title) {
        return title;
    }
    renderFeedLabel(title) {
        return this.renderHeaderCell(title);
    }
    renderFeedValue(value, entity, isNested) {
        return this.renderCell(value, entity, isNested || false);
    }
    constructor(column) {
        Object.defineProperty(this, "column", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: column
        });
    }
    render(entity, isNested) {
        return this.renderCellAsync(entity, isNested || false, (value, e) => this.renderCell(value, e, isNested || false));
    }
    renderHeader(title) {
        return this.renderHeaderCell(this.getHeaderTitle(title));
    }
    renderFeed(title, entity, isNested) {
        return [
            this.renderFeedLabel(this.getHeaderTitle(title)),
            this.renderCellAsync(entity, isNested || false, (value, e) => this.renderFeedValue(value, e, isNested))
        ];
    }
}
exports.BasicColumn = BasicColumn;
