// @geotab/zenith/dist/header/components/collapsedItemsControl/utils/handleCustomMenuWrapper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleCustomMenuWrapper = void 0;
const headerHelpers_1 = require("../../../headerHelpers");
const controlledMenu_1 = require("../../../../menu/controlledMenu");
const menu_1 = require("../../../../menu/menu");
const react_1 = require("react");
const headerButton_1 = require("../../../headerButton");
// eslint-disable-next-line complexity
const handleCustomMenuWrapper = (reactNode, content, menuItems) => {
    if ((0, headerHelpers_1.isNamedChildComponent)(reactNode, headerHelpers_1.HeaderCustomMenuWrapper)) {
        const customContent = content === null || content === void 0 ? void 0 : content[reactNode.props.id];
        if (!customContent) {
            return;
        }
        let menu;
        let isOnlyButton = false;
        if (customContent.type === menu_1.Menu || customContent.type === controlledMenu_1.ControlledMenu || customContent.type === headerButton_1.HeaderButton) {
            menu = customContent;
            isOnlyButton = customContent.type === headerButton_1.HeaderButton;
        }
        else {
            react_1.Children.forEach(customContent.props.children, ch => {
                if (ch.type === menu_1.Menu || ch.type === controlledMenu_1.ControlledMenu) {
                    menu = ch;
                }
                if (ch.type === headerButton_1.HeaderButton) {
                    isOnlyButton = true;
                    menu = ch;
                }
            });
        }
        if (!menu) {
            return;
        }
        // TODO: which props could be taken from Menu itself? id, name?
        const el = (0, react_1.createElement)(menu_1.Menu.Item, {
            id: reactNode.props.id || menu.props.id,
            key: reactNode.props.id || menu.props.id,
            name: reactNode.props.name || menu.props.name,
            disabled: reactNode.props.disabled || menu.props.disabled,
            link: reactNode.props.link || menu.props.link,
            target: reactNode.props.target || menu.props.target,
            rel: reactNode.props.rel || menu.props.rel,
            icon: reactNode.props.icon || menu.props.icon
        }, isOnlyButton ? [] : menu.props.children);
        menuItems.push(el);
    }
};
exports.handleCustomMenuWrapper = handleCustomMenuWrapper;
