// @geotab/zenith/dist/groupsFilter/stateReducer/stateReducerHelper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPreparedState = exports.getAdvancedFilterNamedState = exports.isAdvancedFilterInitialState = exports.getCurrentlyDeselectedMenu = exports.getCurrentlySelectedMenu = exports.getSearchResultFromObject = exports.getGroupsTreeItem = exports.getPathToRoot = exports.getSelectedChildren = exports.getMenu = exports.getCurrentlySelectedCount = exports.getFrozenCount = exports.isStateChanged = exports.getCurrentState = void 0;
const groupsFilterHelper_1 = require("../groupsFilterHelper");
const groupsHelper_1 = require("../groupsHelper");
const getCurrentState = (stateObj) => ({
    advancedFilterState: stateObj.advancedFilterState,
    simpleFilterState: stateObj.selectedIds,
    relation: stateObj.relation
});
exports.getCurrentState = getCurrentState;
const isStateChanged = (stateObj) => !(stateObj.relation === stateObj.frozenState.relation
    && stateObj.sideWideState === stateObj.frozenSideWideState
    && stateObj.selectedIds.sort().join(",") === (stateObj.frozenState.simpleFilterState || []).sort().join(",")
    && (0, groupsHelper_1.compareFilters)(stateObj.advancedFilterState, stateObj.frozenState.advancedFilterState || (0, groupsHelper_1.getDefaultFilterState)()));
exports.isStateChanged = isStateChanged;
const getFrozenCount = (stateObj) => {
    var _a;
    return ((0, groupsHelper_1.compareFilters)(stateObj.frozenState.advancedFilterState || (0, groupsHelper_1.getDefaultFilterState)(), (0, groupsHelper_1.getDefaultFilterState)()) ? 0 : 1)
        + (((_a = stateObj.frozenState.simpleFilterState) === null || _a === void 0 ? void 0 : _a.filter((0, groupsFilterHelper_1.doesGroupExist)(stateObj.groupsMap)).length) || 0);
};
exports.getFrozenCount = getFrozenCount;
const getCurrentlySelectedCount = (stateObj) => ((0, groupsHelper_1.compareFilters)(stateObj.advancedFilterState, (0, groupsHelper_1.getDefaultFilterState)()) ? 0 : 1)
    + (stateObj.selectedIds.filter((0, groupsFilterHelper_1.doesGroupExist)(stateObj.groupsMap)).length || 0);
exports.getCurrentlySelectedCount = getCurrentlySelectedCount;
const getMenu = (stateObj, currentId = groupsHelper_1.ENTIRE_ORGANIZATION_GROUP_ID) => (0, groupsFilterHelper_1.prepareMenu)(stateObj.groupsMap[currentId]);
exports.getMenu = getMenu;
const getSelectedChildren = (stateObj, currentId) => {
    const currentEl = stateObj.groupsMap[currentId];
    let selectedChild = [];
    const currentChild = (0, groupsFilterHelper_1.getChildList)(currentEl.children);
    currentChild.forEach(el => {
        if (stateObj.selectedIds.indexOf(el.id) > -1) {
            selectedChild = [...selectedChild, el.id];
        }
    });
    return selectedChild;
};
exports.getSelectedChildren = getSelectedChildren;
const getPathToRoot = (stateObj, itemId) => {
    if (!itemId) {
        return "";
    }
    const pathToRoot = (0, groupsFilterHelper_1.getParentsNameArr)(stateObj.translate, stateObj.groupsMap, itemId, "root");
    return pathToRoot.length > 0 ? pathToRoot.join("/") : pathToRoot[0] || "";
};
exports.getPathToRoot = getPathToRoot;
const getGroupsTreeItem = (stateObj, itemId) => {
    const item = stateObj.groupsMap[itemId];
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
    return item ? Object.assign({}, item) : undefined;
};
exports.getGroupsTreeItem = getGroupsTreeItem;
const getSearchResultFromObject = (stateObj, searchStr) => {
    if (!searchStr) {
        return undefined;
    }
    const searchResult = Object.keys(stateObj.groupsMap).filter(key => key !== groupsHelper_1.ENTIRE_ORGANIZATION_GROUP_ID).map(key => stateObj.groupsMap[key]).filter(el => { var _a; return ((_a = el.name) === null || _a === void 0 ? void 0 : _a.toLowerCase().includes(searchStr.toLowerCase())) || stateObj.translate(el.id).toLowerCase().includes(searchStr.toLowerCase()); });
    const searchResultItems = (0, groupsFilterHelper_1.prepareNameSelectedItems)(stateObj.translate, stateObj.groupsMap, searchResult.map(el => el.id));
    const searchResultMenu = (0, groupsFilterHelper_1.prepareSelectedMenu)(searchResultItems);
    return searchResultMenu;
};
exports.getSearchResultFromObject = getSearchResultFromObject;
const getCurrentlySelectedMenu = (stateObj) => {
    const frozenItems = stateObj.frozenState.simpleFilterState || [];
    const allItems = stateObj.selectedIds.concat(frozenItems.filter(el => stateObj.selectedIds.indexOf(el) < 0));
    const allNamedItems = (0, groupsFilterHelper_1.prepareNameSelectedItems)(stateObj.translate, stateObj.groupsMap, allItems);
    const selectedItem = (0, groupsFilterHelper_1.prepareSelectedMenu)(allNamedItems);
    const advancedGroups = (0, groupsHelper_1.compareFilters)(stateObj.advancedFilterState, (0, groupsHelper_1.getDefaultFilterState)()) ? stateObj.frozenState.advancedFilterState || (0, groupsHelper_1.getDefaultFilterState)()
        : stateObj.advancedFilterState;
    return Object.assign(Object.assign({}, selectedItem), { advancedGroups });
};
exports.getCurrentlySelectedMenu = getCurrentlySelectedMenu;
const getCurrentlyDeselectedMenu = (stateObj, deselectedItems) => {
    const frozenItems = stateObj.frozenState.simpleFilterState || [];
    const allNamedItems = (0, groupsFilterHelper_1.prepareNameSelectedItems)(stateObj.translate, stateObj.groupsMap, deselectedItems.filter(el => !frozenItems.includes(el)));
    const deselectedMenu = (0, groupsFilterHelper_1.prepareSelectedMenu)(allNamedItems);
    return Object.assign({}, deselectedMenu);
};
exports.getCurrentlyDeselectedMenu = getCurrentlyDeselectedMenu;
const isAdvancedFilterInitialState = (stateObj, state) => (0, groupsHelper_1.compareFilters)(state || stateObj.advancedFilterState, (0, groupsHelper_1.getDefaultFilterState)());
exports.isAdvancedFilterInitialState = isAdvancedFilterInitialState;
const getAdvancedFilterNamedState = (stateObj, advState) => {
    const advancedState = advState || stateObj.advancedFilterState;
    const itemsForFill = advancedState.items;
    const getFillNamedItems = (items) => {
        const filledItems = items.map(item => {
            if (item.id) {
                const groupMapItem = stateObj.groupsMap[item.id];
                // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
                return Object.assign(Object.assign({}, item), { name: groupMapItem && (0, groupsHelper_1.getGroupDescription)(groupMapItem, stateObj.translate) });
            }
            return Object.assign(Object.assign({}, item), { items: getFillNamedItems(item.items) });
        });
        return [...filledItems];
    };
    return Object.assign(Object.assign({}, advancedState), { items: getFillNamedItems(itemsForFill) });
};
exports.getAdvancedFilterNamedState = getAdvancedFilterNamedState;
const getPreparedState = (stateObj) => ({
    groups: (0, groupsFilterHelper_1.convertPreparedFilters)(stateObj.frozenState),
    sideWide: stateObj.frozenSideWideState
});
exports.getPreparedState = getPreparedState;
