// @geotab/zenith/dist/groupsFilter/stateReducer/stateReducer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.stateReducer = exports.getInitialState = exports.defaultPreparedFilterState = void 0;
const groupsFilterHelper_1 = require("../groupsFilterHelper");
const groupsHelper_1 = require("../groupsHelper");
const stateActionType_1 = require("./stateActionType");
exports.defaultPreparedFilterState = {
    advancedFilterState: (0, groupsHelper_1.getDefaultFilterState)(),
    simpleFilterState: [],
    relation: groupsHelper_1.DEFAULT_RELATION_OPERATOR
};
const getInitialState = (translate, initState) => {
    const frozenInitialSideWideState = initState ? initState.sideWide : undefined;
    let initialOptions = {};
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
    const preparedInitialState = (initState === null || initState === void 0 ? void 0 : initState.groups) && initState.groups.items ? (0, groupsFilterHelper_1.prepareFilters)(initState.groups) : undefined;
    if (preparedInitialState) {
        initialOptions = {
            frozenState: {
                advancedFilterState: preparedInitialState.advancedFilterState || (0, groupsHelper_1.getDefaultFilterState)(),
                simpleFilterState: preparedInitialState.simpleFilterState || [],
                relation: preparedInitialState.relation
            },
            selectedIds: preparedInitialState.simpleFilterState || [],
            advancedFilterState: preparedInitialState.advancedFilterState ? preparedInitialState.advancedFilterState : (0, groupsHelper_1.getDefaultFilterState)(),
            relation: preparedInitialState.relation,
            frozenSideWideState: initState.sideWide,
            sideWideState: initState.sideWide
        };
    }
    return Object.assign({ translate: translate, groupsMap: {}, frozenSideWideState: frozenInitialSideWideState || false, sideWideState: false, globalChanges: false, relation: groupsHelper_1.DEFAULT_RELATION_OPERATOR, selectedIds: [], advancedFilterState: (0, groupsHelper_1.getDefaultFilterState)(), frozenState: exports.defaultPreparedFilterState }, initialOptions);
};
exports.getInitialState = getInitialState;
// eslint-disable-next-line complexity
function stateReducer(state, action) {
    var _a, _b;
    const { type, payload } = action;
    switch (type) {
        case stateActionType_1.StateActionType.CreateGroupsMap:
            return Object.assign(Object.assign({}, state), { groupsMap: (0, groupsFilterHelper_1.createGroupsMap)(payload, groupsHelper_1.ENTIRE_ORGANIZATION_GROUP_ID) });
        case stateActionType_1.StateActionType.CreateInitialState:
            return Object.assign(Object.assign({}, state), { frozenState: {
                    advancedFilterState: payload.value.advancedFilterState || (0, groupsHelper_1.getDefaultFilterState)(),
                    simpleFilterState: payload.value.simpleFilterState || [],
                    relation: payload.value.relation
                }, selectedIds: payload.value.simpleFilterState || [], advancedFilterState: payload.value.advancedFilterState || (0, groupsHelper_1.getDefaultFilterState)(), relation: payload.value.relation, frozenSideWideState: payload.global, sideWideState: payload.global });
        case stateActionType_1.StateActionType.FreezeSideWide:
            return Object.assign(Object.assign({}, state), { frozenSideWideState: payload, globalChanges: true });
        case stateActionType_1.StateActionType.FreezeState:
            return Object.assign(Object.assign({}, state), { frozenState: {
                    advancedFilterState: payload.advancedFilterState || (0, groupsHelper_1.getDefaultFilterState)(),
                    simpleFilterState: payload.simpleFilterState || [],
                    relation: payload.relation
                }, selectedIds: payload.simpleFilterState || [], advancedFilterState: payload.advancedFilterState || (0, groupsHelper_1.getDefaultFilterState)(), relation: payload.relation, globalChanges: true });
        case stateActionType_1.StateActionType.TurnOfGlobalChanges:
            return Object.assign(Object.assign({}, state), { globalChanges: false });
        case stateActionType_1.StateActionType.SideWide:
            return Object.assign(Object.assign({}, state), { sideWideState: payload });
        case stateActionType_1.StateActionType.SetState:
            return Object.assign(Object.assign({}, state), { selectedIds: payload.simpleFilterState || [], advancedFilterState: payload.advancedFilterState ? payload.advancedFilterState : state.advancedFilterState, relation: payload.relation });
        case stateActionType_1.StateActionType.SetAdvancedFilterState:
            return Object.assign(Object.assign({}, state), { advancedFilterState: payload });
        case stateActionType_1.StateActionType.Relation:
            return Object.assign(Object.assign({}, state), { relation: payload });
        case stateActionType_1.StateActionType.ChangeSelection: {
            const currentSelectedIds = new Set(state.selectedIds);
            if (currentSelectedIds.has(payload.itemId)) {
                currentSelectedIds.delete(payload.itemId);
            }
            else {
                currentSelectedIds.add(payload.itemId);
            }
            if (payload.value) {
                // deselect parent
                const parentForDeselect = [];
                let parent = (_a = state.groupsMap[payload.itemId].parent) === null || _a === void 0 ? void 0 : _a.id;
                while (parent && parent !== "root") {
                    parentForDeselect.push(parent);
                    parent = (_b = state.groupsMap[parent].parent) === null || _b === void 0 ? void 0 : _b.id;
                }
                parentForDeselect.forEach(parentId => {
                    currentSelectedIds.delete(parentId);
                });
                // deselect children
                const childrenForDeselect = [];
                const currentChild = (0, groupsFilterHelper_1.getChildList)(state.groupsMap[payload.itemId].children);
                currentChild.forEach(el => {
                    childrenForDeselect.push(el.id);
                });
                childrenForDeselect.forEach(childId => currentSelectedIds.delete(childId));
            }
            return Object.assign(Object.assign({}, state), { selectedIds: Array.from(currentSelectedIds) });
        }
        case stateActionType_1.StateActionType.ResetSelection:
            return Object.assign(Object.assign({}, state), { selectedIds: [], advancedFilterState: (0, groupsHelper_1.getDefaultFilterState)(), relation: groupsHelper_1.DEFAULT_RELATION_OPERATOR, sideWideState: false });
        case stateActionType_1.StateActionType.SelectAllChildren: {
            const currentEl = state.groupsMap[payload];
            const newSelected = new Set(state.selectedIds);
            (currentEl.children || []).forEach(el => {
                var _a, _b;
                if (!newSelected.has(el.id)) {
                    newSelected.add(el.id);
                    // deselect parent
                    const parentForDeselect = [];
                    let parent = (_a = state.groupsMap[el.id].parent) === null || _a === void 0 ? void 0 : _a.id;
                    while (parent && parent !== "root") {
                        parentForDeselect.push(parent);
                        parent = (_b = state.groupsMap[parent].parent) === null || _b === void 0 ? void 0 : _b.id;
                    }
                    parentForDeselect.forEach(parentId => {
                        newSelected.delete(parentId);
                    });
                    //deselect children
                    const childrenForDeselect = [];
                    const currentChild = (0, groupsFilterHelper_1.getChildList)(state.groupsMap[el.id].children);
                    currentChild.forEach(elt => {
                        childrenForDeselect.push(elt.id);
                    });
                    childrenForDeselect.forEach(childId => newSelected.delete(childId));
                }
            });
            return Object.assign(Object.assign({}, state), { selectedIds: Array.from(newSelected) });
        }
        case stateActionType_1.StateActionType.DeselectAllChildren: {
            const childrenForDeselect = [];
            const currentChild = (0, groupsFilterHelper_1.getChildList)(state.groupsMap[payload].children);
            currentChild.forEach(el => {
                childrenForDeselect.push(el.id);
            });
            const newSelected = new Set(state.selectedIds);
            childrenForDeselect.forEach(childId => newSelected.delete(childId));
            return Object.assign(Object.assign({}, state), { selectedIds: Array.from(newSelected) });
        }
        default: {
            throw Error("Unknown action: " + type);
        }
    }
}
exports.stateReducer = stateReducer;
