// @geotab/zenith/dist/groupsFilter/stateReducer/stateActionType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StateActionType = void 0;
var StateActionType;
(function (StateActionType) {
    StateActionType["CreateGroupsMap"] = "CREATE_GROUPS_MAP";
    StateActionType["CreateInitialState"] = "CREATE_INITIAL_FILTER_STATE";
    StateActionType["FreezeSideWide"] = "FREEZE_SIDE_WIDE";
    StateActionType["FreezeState"] = "FREEZE_STATE";
    StateActionType["SideWide"] = "SIDE_WIDE";
    StateActionType["SetState"] = "SET_STATE";
    StateActionType["SetAdvancedFilterState"] = "SET_ADVANCED_FILTER_STATE";
    StateActionType["Relation"] = "RELATION";
    StateActionType["ChangeSelection"] = "CHANGE_SELECTION";
    StateActionType["ResetSelection"] = "RESET_SELECTION";
    StateActionType["DeselectAllChildren"] = "DESELECT_ALL_CHILDREN";
    StateActionType["SelectAllChildren"] = "SELECT_ALL_CHILDREN";
    StateActionType["TurnOfGlobalChanges"] = "TURN_OF_GLOBAL_CHANGES";
})(StateActionType || (exports.StateActionType = StateActionType = {}));
