// @geotab/zenith/dist/advancedGroupsFilter/advancedGroupsFilterFormHelper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.prepareToFilterState = exports.prepareCurrentFilterState = void 0;
const groupsHelper_1 = require("../groupsFilter/groupsHelper");
const groupsFilterInterfaces_1 = require("../groupsFilter/groupsFilterInterfaces");
const advancedGroupsFilterHelper_1 = require("./advancedGroupsFilterHelper");
const prepareCurrentFilterState = (state) => {
    if (!state) {
        return [{ sectionState: (0, groupsHelper_1.getDefaultFilterState)(), interSectionRelation: groupsHelper_1.DEFAULT_RELATION_OPERATOR }];
    }
    const sectionItems = [];
    const addToSelections = (item) => sectionItems.push(item);
    const setItemState = (parentItem, prevRelation = groupsHelper_1.DEFAULT_RELATION_OPERATOR) => {
        if (!(0, advancedGroupsFilterHelper_1.isFilterStateLeafItem)(parentItem)) {
            return parentItem.items.map(item => setItemState(item, parentItem.relation));
        }
        return addToSelections({
            interSectionRelation: prevRelation,
            sectionState: Object.assign({}, parentItem)
        });
    };
    setItemState(state);
    return sectionItems;
};
exports.prepareCurrentFilterState = prepareCurrentFilterState;
const prepareToFilterState = (state) => {
    if (state.length === 1) {
        return state[0].sectionState;
    }
    const getState = (sect) => {
        const getId = entity => entity.id;
        return Object.assign(Object.assign({}, sect), { sectionState: Object.assign(Object.assign({}, sect.sectionState), { items: sect.sectionState.items.map(getId).sort() }) });
    };
    const preparedState = state.reduce((prevState, section) => {
        const itemState = getState(section);
        if (state.length > 1 && (0, groupsHelper_1.isEntireOrganization)(itemState.sectionState) && itemState.sectionState.relation === groupsFilterInterfaces_1.RelationOperator.OR) {
            return prevState;
        }
        if (!Object.keys(prevState).length) {
            return itemState.sectionState;
        }
        return {
            relation: itemState.interSectionRelation,
            items: [
                prevState,
                itemState.sectionState
            ]
        };
    }, {});
    return Object.keys(preparedState).length ? (0, groupsHelper_1.toFilterState)((0, groupsHelper_1.simplifyFilterState)(preparedState)) : (0, groupsHelper_1.getDefaultFilterState)();
};
exports.prepareToFilterState = prepareToFilterState;
