// @geotab/zenith/dist/dataGrid/withLoading/withLoading.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.withLoading = exports.DEFAULT_LOADING_ROWS = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const useFetch_1 = require("../../commonHelpers/hooks/useFetch");
const generateId_1 = require("../../commonHelpers/generateId");
const loadingRowCell_1 = require("./components/loadingRowCell");
const classNames_1 = require("../../commonHelpers/classNames/classNames");
const defaultColumnComponent_1 = require("../extensions/defaultColumnComponent");
exports.DEFAULT_LOADING_ROWS = 20;
const getDefaultEntitiesLoadingColumn = (columnId, originalColumnComponent) => {
    const renderFeed = (title, entity, isNested) => originalColumnComponent.renderFeed
        ? originalColumnComponent.renderFeed(title, entity, isNested)
        : [
            originalColumnComponent.renderHeader ? originalColumnComponent.renderHeader(title) : title,
            originalColumnComponent.render(entity, isNested)
        ];
    return {
        render: (entity, isNested) => entity.isLoading ? (0, jsx_runtime_1.jsx)(loadingRowCell_1.LoadingCell, { children: entity[columnId] }) : originalColumnComponent.render(entity, isNested),
        renderHeader: (title) => originalColumnComponent.renderHeader ? originalColumnComponent.renderHeader(title) : title,
        renderFeed: (title, entity, isNested) => entity.isLoading
            ? [(0, jsx_runtime_1.jsx)(loadingRowCell_1.LoadingCell, { children: title }, entity.id), (0, jsx_runtime_1.jsx)(loadingRowCell_1.LoadingCell, { children: entity[columnId] }, entity.id)]
            : renderFeed(title, entity, isNested)
    };
};
const getEntitiesLoadingColumn = (column) => column.loadingEntitiesColumnComponent || getDefaultEntitiesLoadingColumn(column.id, column.columnComponent || (0, defaultColumnComponent_1.defaultColumnComponent)(column.id));
// T - type of entity
// P - props of wrapped component
// eslint-disable-next-line @typescript-eslint/naming-convention
const withLoading = (DataGridComponent) => (0, react_1.forwardRef)(
// eslint-disable-next-line @typescript-eslint/naming-convention
function InnerWithLoadingList(props, ref) {
    const { entitiesPromise, loadingRows, columns, options } = props;
    const [data, setData] = (0, react_1.useState)([]);
    const state = (0, useFetch_1.useFetch)({}, entitiesPromise);
    const sampleEntity = (0, react_1.useMemo)(() => columns.map((column) => ({ [column.id]: column.title })).reduce((acc, el) => (Object.assign(Object.assign({}, acc), el))), [columns]);
    const generateLoadingEntities = (0, react_1.useCallback)(() => (new Array(loadingRows || exports.DEFAULT_LOADING_ROWS).fill(undefined)).map(() => (Object.assign(Object.assign({}, sampleEntity), { isLoading: true, id: (0, generateId_1.generateId)() }))), [loadingRows, sampleEntity]);
    (0, react_1.useEffect)(() => {
        if (state.data) {
            setData(prevData => {
                const newData = state.data || [];
                return (options === null || options === void 0 ? void 0 : options.addToEnd) ? [...prevData, ...newData] : newData;
            });
        }
    }, [state.data, options === null || options === void 0 ? void 0 : options.addToEnd]);
    const selectMode = (0, react_1.useMemo)(() => {
        if (state.isLoading) {
            return false;
        }
        return props.selectMode;
    }, [state.isLoading, props.selectMode]);
    return (0, jsx_runtime_1.jsx)(DataGridComponent, Object.assign({}, props, { selectMode: selectMode, className: (0, classNames_1.classNames)([
            props.className || "",
            state.isLoading ? "zen-data-grid--loading" : ""
        ]), entities: (state.isLoading)
            ? (options === null || options === void 0 ? void 0 : options.addToEnd) ? [...data, ...generateLoadingEntities()] : generateLoadingEntities()
            : data, columns: columns.map(column => (Object.assign(Object.assign({}, column), { meta: Object.assign({ renderPlaceholder: (entity) => (0, jsx_runtime_1.jsx)(loadingRowCell_1.LoadingCell, { children: entity[column.id] }), renderFeedPlaceholder: (entity) => [
                    (0, jsx_runtime_1.jsx)(loadingRowCell_1.LoadingCell, { children: column.title }, `${column.id}_label`),
                    (0, jsx_runtime_1.jsx)(loadingRowCell_1.LoadingCell, { children: entity[column.id] }, `${column.id}_value`)
                ] }, column.meta), columnComponent: state.isLoading ? getEntitiesLoadingColumn(column) : column.columnComponent || (0, defaultColumnComponent_1.defaultColumnComponent)(column.id) }))), ref: ref }), "grid");
});
exports.withLoading = withLoading;
