// @geotab/zenith/dist/table/selectable/selectableHeader.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TRANSLATIONS = exports.SelectableHeader = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const useLanguage_1 = require("../../utils/localization/useLanguage");
const textIconButton_1 = require("../../textIconButton/textIconButton");
const iconChevronDownSmall_1 = require("../../icons/iconChevronDownSmall");
const menu_1 = require("../../menu/menu");
const checkbox_1 = require("../../checkbox/checkbox");
const interfaces_1 = require("./interfaces");
const selectionOptions = [{
        id: "all",
        value: interfaces_1.SelectionState.All,
        element: "all",
        label: "Select all"
    }, {
        id: "page",
        value: interfaces_1.SelectionState.CurrentPage,
        element: "page",
        label: "Select current page"
    }, {
        id: "none",
        value: interfaces_1.SelectionState.None,
        element: "none",
        label: "Deselect all"
    }];
const SelectableHeader = ({ label, title, checked, indeterminate, disabled, onChange, calculateSelectionState, onSelect, checkboxInHeader, header, turnOffSelectAll }) => {
    const { translate } = (0, useLanguage_1.useLanguage)();
    const listItems = (0, react_1.useMemo)(() => selectionOptions
        .filter(option => !turnOffSelectAll || option.value !== interfaces_1.SelectionState.All)
        .map(option => ({
        id: option.id,
        name: translate(option.label),
        onClick: () => onSelect && onSelect(option.id)
    }))
        .filter(option => option.id !== calculateSelectionState()), [calculateSelectionState, onSelect, translate, turnOffSelectAll]);
    const actionsMenu = (0, react_1.useMemo)(() => (0, jsx_runtime_1.jsx)(menu_1.Menu, { title: translate("Select menu"), className: "zen-selectable-column__menu", trigger: (0, jsx_runtime_1.jsx)(textIconButton_1.TextIconButton, { className: "zen-selectable-column__popup", iconClasses: "zen-icon--no-events", icon: iconChevronDownSmall_1.IconChevronDownSmall, iconPosition: textIconButton_1.ButtonIconPosition.Start, iconSize: "large", title: translate("Multiple row selection menu"), "aria-label": translate("Multiple row selection menu"), type: "tertiary-black" }), children: listItems.map(item => (0, jsx_runtime_1.jsx)(menu_1.Menu.Item, { id: item.id, name: item.name, onClick: item.onClick }, item.id)) }), [listItems, translate]);
    return (0, jsx_runtime_1.jsxs)("div", { className: "zen-selectable-column zen-selectable-column--with-actions", children: [checkboxInHeader !== false
                ? (0, jsx_runtime_1.jsxs)("div", { className: "zen-selectable-column__checkbox", children: [(0, jsx_runtime_1.jsx)("div", { className: "zen-selectable-column__checkbox-input", children: (0, jsx_runtime_1.jsx)(checkbox_1.Checkbox, { checked: checked, indeterminate: indeterminate, disabled: disabled, title: label, "aria-label": label, onChange: e => onChange && onChange(e.target.checked) }) }), actionsMenu] })
                : null, header !== false
                ? (0, jsx_runtime_1.jsx)("div", { className: "zen-selectable-column__content", children: title })
                : null] });
};
exports.SelectableHeader = SelectableHeader;
exports.TRANSLATIONS = [
    "Select menu",
    "Select all",
    "Select current page",
    "Deselect all",
    "Select menu",
    "Multiple row selection menu"
];
