// @geotab/zenith/dist/dataGrid/columns/checkboxColumn/checkboxColumn.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CheckboxColumn = void 0;
const react_1 = require("react");
const checkboxColumnWrapper_1 = require("./checkboxColumnWrapper");
const simpleEventHandler_1 = require("../../../commonHelpers/simpleEventHandler");
class CheckboxColumn extends simpleEventHandler_1.SimpleEventHandler {
    constructor(selections, options) {
        super();
        Object.defineProperty(this, "wrapper", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.wrapper = new checkboxColumnWrapper_1.CheckboxColumnWrapper(selections, Object.assign(Object.assign({}, options), { createTitle: options.createTitle || (() => "") }), (0, react_1.createRef)());
        this.wrapper.attach("change", e => this.fire("change", e));
    }
    render(entity) {
        return this.wrapper.render("", entity);
    }
    renderHeader(title) {
        return title;
    }
    renderFeed(title, entity) {
        return [this.renderHeader(title), this.render(entity)];
    }
    get selectionState() {
        return this.wrapper.selectionState;
    }
    set selectionState(state) {
        this.wrapper.selectionState = state;
    }
}
exports.CheckboxColumn = CheckboxColumn;
