// @geotab/zenith/dist/dataGrid/columns/checkboxColumn/checkboxColumnWrapper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CheckboxColumnWrapper = exports.SelectionState = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const simpleEventHandler_1 = require("../../../commonHelpers/simpleEventHandler");
const checkboxCell_1 = require("./checkboxCell");
const checkboxHeaderCell_1 = require("./checkboxHeaderCell");
var SelectionState;
(function (SelectionState) {
    SelectionState[SelectionState["All"] = 0] = "All";
    SelectionState[SelectionState["CurrentPage"] = 1] = "CurrentPage";
    SelectionState[SelectionState["Partial"] = 2] = "Partial";
    SelectionState[SelectionState["None"] = 3] = "None";
})(SelectionState || (exports.SelectionState = SelectionState = {}));
class CheckboxColumnWrapper extends simpleEventHandler_1.SimpleEventHandler {
    constructor(selections, options, menuTriggerRef) {
        super();
        Object.defineProperty(this, "selections", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: selections
        });
        Object.defineProperty(this, "options", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: options
        });
        Object.defineProperty(this, "menuTriggerRef", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: menuTriggerRef
        });
        Object.defineProperty(this, "condition", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "disabled", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "state", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: SelectionState.None
        });
        Object.defineProperty(this, "actionsClickHandler", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: e => {
                this.fire("toggle", [e, e.target]);
            }
        });
        Object.defineProperty(this, "changeHandler", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: (id, checked) => {
                this.fire("change", [id, checked]);
            }
        });
        Object.defineProperty(this, "changeHandlerAll", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: (checked) => {
                this.selectionState = checked ? SelectionState.All : SelectionState.None;
                this.fire("change-all", [checked]);
            }
        });
        this.condition = options.condition || (() => true);
        this.disabled = options.disabled;
    }
    get selectionState() {
        return this.state;
    }
    set selectionState(state) {
        this.state = state;
    }
    // cloning nodes is twice as fast as creating a new one
    render(value, entity) {
        if (!entity) {
            return null;
        }
        const allowed = this.condition(entity);
        const isDisable = this.disabled ? this.disabled(entity) : this.state === SelectionState.All;
        const cell = (0, jsx_runtime_1.jsx)(checkboxCell_1.CheckboxCell, { id: entity.id, className: allowed ? "" : "zen-checkbox-column--no-checkbox", checked: this.state === SelectionState.All || this.selections.has(entity.id), indeterminate: false, disabled: isDisable, label: this.options.createTitle(entity), onChange: this.changeHandler, children: (0, jsx_runtime_1.jsx)(checkboxCell_1.CheckboxCell.Content, { children: value }) });
        return cell;
    }
    renderHeader(title) {
        if (this.options.header === false) {
            return title;
        }
        const cell = (0, jsx_runtime_1.jsxs)(checkboxHeaderCell_1.CheckboxHeaderCell, { label: this.options.headerTitle || "", actions: true, checked: this.state === SelectionState.CurrentPage || this.state === SelectionState.All, indeterminate: this.state === SelectionState.Partial, onChange: this.changeHandlerAll, children: [(0, jsx_runtime_1.jsx)(checkboxHeaderCell_1.CheckboxHeaderCell.Actions, { ref: this.menuTriggerRef, onClick: this.actionsClickHandler }), (0, jsx_runtime_1.jsx)(checkboxHeaderCell_1.CheckboxHeaderCell.Content, { children: title })] });
        return cell;
    }
}
exports.CheckboxColumnWrapper = CheckboxColumnWrapper;
Object.defineProperty(CheckboxColumnWrapper, "translations", {
    enumerable: true,
    configurable: true,
    writable: true,
    value: [
        "Multiple row selection menu"
    ]
});
