// @geotab/zenith/dist/chart/lineChart/useSummary.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useSummary = void 0;
const react_1 = require("react");
const mapChartSummaryToSummary = (chartSummary) => {
    var _a;
    const mappedSummary = {
        value: (_a = chartSummary.value) !== null && _a !== void 0 ? _a : 0,
        color: chartSummary.color,
        title: chartSummary.label,
        unit: chartSummary.unit,
        description: chartSummary.description,
        more: chartSummary.more,
        onToggle: chartSummary.onClick,
        isEnabled: chartSummary.isHidden
    };
    return mappedSummary;
};
const useSummary = (chartSummary, data) => {
    const summary = (0, react_1.useMemo)(() => chartSummary ? Object.assign(Object.assign({}, chartSummary), mapChartSummaryToSummary(chartSummary)) : false, [chartSummary]);
    const value = (0, react_1.useMemo)(() => {
        if (summary && summary.value) {
            return summary.value;
        }
        if (data.datasets.length === 0) {
            return "-";
        }
        const dataset = data.datasets[0];
        if (dataset.data.length === 0) {
            return "-";
        }
        for (let i = dataset.data.length - 1; i >= 0; i--) {
            const point = dataset.data[i];
            // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
            if (point === null) {
                continue;
            }
            if (typeof point === "number") {
                return point;
            }
            // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
            if (point.y !== undefined && point.y !== null) {
                return point.y;
            }
        }
        return "-";
    }, [data, summary]);
    const color = (0, react_1.useMemo)(() => {
        if (summary && summary.color) {
            return summary.color;
        }
        return data.datasets[0] && typeof data.datasets[0].borderColor === "string"
            ? data.datasets[0].borderColor
            : (summary && summary.title) ? "#2853BD" : undefined;
    }, [data.datasets, summary]);
    const title = (0, react_1.useMemo)(() => {
        if (summary && summary.title) {
            return summary.title;
        }
        return data.datasets[0] ? data.datasets[0].label : undefined;
    }, [summary, data.datasets]);
    return Object.assign(Object.assign({}, summary), { title,
        value,
        color });
};
exports.useSummary = useSummary;
