// @geotab/zenith/dist/calendar/calendarUtils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getCalendarRange = exports.setFocusTo = exports.getButtonLabel = exports.isDateInRange = exports.isRightBorderVisible = exports.isLeftBorderVisible = void 0;
const dateUtils_1 = require("./dateUtils");
const isLeftBorderVisible = (date, hoverDate, startDate, endDate) => {
    if (!startDate) {
        return false;
    }
    if (endDate) {
        return date > startDate && date <= endDate;
    }
    if (hoverDate) {
        if (hoverDate < startDate) {
            return date > hoverDate && date <= startDate;
        }
        return date > startDate && date <= hoverDate;
    }
    return false;
};
exports.isLeftBorderVisible = isLeftBorderVisible;
const isRightBorderVisible = (date, hoverDate, startDate, endDate) => {
    if (!startDate) {
        return false;
    }
    const dateLike = new Date(date);
    dateLike.setHours(0, 0, 0, 0);
    if (endDate) {
        const endDateLike = new Date(endDate);
        endDateLike.setHours(0, 0, 0, 0);
        if (date >= startDate && date < endDate) {
            return true;
        }
        return false;
    }
    if (hoverDate) {
        const startDateLike = new Date(startDate);
        startDateLike.setHours(0, 0, 0, 0);
        const hoverDateLike = new Date(hoverDate);
        hoverDateLike.setHours(0, 0, 0, 0);
        if (hoverDate < startDate) {
            return date >= hoverDate && date < startDate;
        }
        return date >= startDate && date < hoverDate;
    }
    return false;
};
exports.isRightBorderVisible = isRightBorderVisible;
const isDateInRange = (date, hoverDate, startDate, endDate) => {
    if (!startDate) {
        return false;
    }
    if (endDate) {
        return date >= startDate && date <= endDate;
    }
    if (hoverDate) {
        if (hoverDate < startDate) {
            return date >= hoverDate && date <= startDate;
        }
        return date >= startDate && date <= hoverDate;
    }
    return false;
};
exports.isDateInRange = isDateInRange;
const getButtonLabel = (translate, date, isRangeMode, startDate, endDate) => {
    if (isRangeMode) {
        if (startDate && endDate) {
            if (date < startDate) {
                return translate("Change daterange start date, {date}").replace("{date}", startDate);
            }
            return translate("Change daterange end date, {date}").replace("{date}", endDate);
        }
        if (startDate) {
            if (date < startDate) {
                return translate("Change daterange start date, {date}").replace("{date}", startDate);
            }
            return translate("Choose daterange end date");
        }
        return translate("Choose daterange start date");
    }
    return startDate
        ? translate("Change date, {date}").replace("{date}", date)
        : translate("Choose date");
};
exports.getButtonLabel = getButtonLabel;
const setFocusTo = (e, date, selectionHandler) => {
    e.preventDefault();
    selectionHandler(date);
};
exports.setFocusTo = setFocusTo;
// eslint-disable-next-line complexity
const getCalendarRange = (selectedDate, currentActiveField, isWrongPeriodFn, dateFrom, dateTo) => {
    const selectedDateObj = new Date(selectedDate);
    const dateFromObj = dateFrom && (0, dateUtils_1.isValidDate)(dateFrom) ? new Date(dateFrom) : undefined;
    const dateToObj = dateTo && (0, dateUtils_1.isValidDate)(dateTo) ? new Date(dateTo) : undefined;
    if (currentActiveField === "start") {
        // Cursor on Start Date
        if (dateToObj && selectedDateObj.getTime() < dateToObj.getTime()) {
            const isWrongPeriod = isWrongPeriodFn(selectedDateObj, dateToObj);
            return isWrongPeriod ? { from: selectedDate, to: undefined } : { from: selectedDate, to: dateTo };
        }
        // Overlapping dates: selectedDate later than dateTo
        return { from: selectedDate, to: undefined };
    }
    // Cursor on End Date
    if (dateFromObj && dateFromObj.getTime() < selectedDateObj.getTime()) {
        const isWrongPeriod = isWrongPeriodFn(dateFromObj, selectedDateObj);
        return isWrongPeriod ? { from: undefined, to: selectedDate } : { from: dateFrom, to: selectedDate };
    }
    // Overlapping dates: selectedDate earlier than dateFrom
    return { from: undefined, to: selectedDate };
};
exports.getCalendarRange = getCalendarRange;
