// @geotab/zenith/dist/summaryTile/summaryTile.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SummaryTile = exports.SummaryTileDisplayName = exports.SummaryTileSize = exports.SummaryTileType = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const React = __importStar(require("react"));
const classNames_1 = require("../commonHelpers/classNames/classNames");
const tooltip_1 = require("../tooltip/tooltip");
const summaryTileTrigger_1 = require("./summaryTileTrigger");
const getTypeClassName_1 = require("./utils/getTypeClassName");
const iconInfoCircle_1 = require("../icons/iconInfoCircle");
const iconWarning_1 = require("../icons/iconWarning");
const iconCheckRadio_1 = require("../icons/iconCheckRadio");
const iconException_1 = require("../icons/iconException");
var SummaryTileType;
(function (SummaryTileType) {
    SummaryTileType["Error"] = "error";
    SummaryTileType["Success"] = "success";
    SummaryTileType["Warning"] = "warning";
    SummaryTileType["Active"] = "active";
    SummaryTileType["Default"] = "default";
    /**
     * @deprecated - will be removed in the next releases.
     * Please use {@link SummaryTileType.Active} instead.
    */
    SummaryTileType["Info"] = "info";
})(SummaryTileType || (exports.SummaryTileType = SummaryTileType = {}));
var SummaryTileSize;
(function (SummaryTileSize) {
    SummaryTileSize["Small"] = "small";
    SummaryTileSize["Medium"] = "medium";
})(SummaryTileSize || (exports.SummaryTileSize = SummaryTileSize = {}));
exports.SummaryTileDisplayName = "SummaryTile";
const SummaryTile = ({ className, children, size = SummaryTileSize.Small, title, tileType = SummaryTileType.Default, id, onClick, tooltipText, tooltipAlignment = "top", tooltipSize }) => {
    const generatedId = React.useId();
    const eltId = id || generatedId;
    const memoizedGetIconId = React.useMemo(() => {
        const iconIds = {
            [SummaryTileType.Warning]: iconException_1.IconException,
            [SummaryTileType.Success]: iconCheckRadio_1.IconCheckRadio,
            [SummaryTileType.Error]: iconWarning_1.IconWarning,
            [SummaryTileType.Active]: iconInfoCircle_1.IconInfoCircle,
            [SummaryTileType.Default]: iconInfoCircle_1.IconInfoCircle,
            [SummaryTileType.Info]: iconInfoCircle_1.IconInfoCircle
        };
        return iconIds[tileType];
    }, [tileType]);
    const componentProps = {
        className: (0, classNames_1.classNames)([
            "zen-summary-tile",
            (0, getTypeClassName_1.getTypeClassName)(tileType),
            `zen-summary-tile--${size}`,
            onClick ? "zen-summary-tile--action" : "",
            className || ""
        ]),
        "data-summary-tile-id": eltId
    };
    if (onClick) {
        componentProps.onClick = () => {
            onClick(eltId);
        };
    }
    const Tag = onClick ? "button" : "div";
    return (0, jsx_runtime_1.jsxs)(Tag, Object.assign({}, componentProps, { children: [(0, jsx_runtime_1.jsxs)("div", { className: "zen-summary-tile__title-block", children: [(0, jsx_runtime_1.jsx)("div", { className: "zen-summary-tile__title-text zen-ellipsis heading-05", children: title }), tooltipText ? (0, jsx_runtime_1.jsx)(tooltip_1.Tooltip, { alignment: tooltipAlignment, size: tooltipSize, trigger: (0, jsx_runtime_1.jsx)(summaryTileTrigger_1.SummaryTileTrigger, { icon: memoizedGetIconId }), children: tooltipText }) : null] }), (0, jsx_runtime_1.jsx)("div", { className: "zen-summary-tile__content", children: children })] }));
};
exports.SummaryTile = SummaryTile;
exports.SummaryTile.displayName = exports.SummaryTileDisplayName;
