// @geotab/zenith/dist/sidePanel/sidePanel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SidePanel = exports.CUSTOM_POPUP_COMPONENT_CLASSNAME = exports.SidePanelCloseReason = exports.formatWidth = exports.isChildPopup = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const classNames_1 = require("../commonHelpers/classNames/classNames");
const react_1 = require("react");
const dialogHelpers_1 = require("../dialog/dialogHelpers");
const focusableSelector_1 = require("../utils/focusableSelector");
const useEscape_1 = require("../commonHelpers/hooks/useEscape");
const useTrapFocus_1 = require("../commonHelpers/hooks/useTrapFocus");
const isChildOf_1 = require("../utils/isChildOf");
const getParentByClass_1 = require("../utils/getParentByClass");
const sidePanelCell_1 = require("./sidePanelCell/sidePanelCell");
const useUniqueId_1 = require("../commonHelpers/useUniqueId");
const react_dom_1 = require("react-dom");
const useDriveClassName_1 = require("../utils/theme/useDriveClassName");
const themeContext_1 = require("../utils/theme/themeContext");
const useFadeComponent_1 = require("../utils/useFadeComponent");
const useMobile_1 = require("../commonHelpers/hooks/useMobile");
/* eslint-enable @typescript-eslint/naming-convention */
const isChildPopup = (target, stopElement) => {
    const parentPopup = (0, getParentByClass_1.getParentByClass)(target, "zen-absolute");
    const parentSidePanel = (0, getParentByClass_1.getParentByClass)(target, "zen-side-panel");
    if (!parentPopup && !parentSidePanel) {
        return false;
    }
    if (parentPopup === stopElement || parentSidePanel === stopElement) {
        return true;
    }
    const popupTrigger = parentPopup
        ? document.querySelector(`[data-popup-id='${parentPopup.id}']`)
        : undefined;
    if (!popupTrigger) {
        return false;
    }
    return (0, exports.isChildPopup)(popupTrigger, stopElement);
};
exports.isChildPopup = isChildPopup;
const formatWidth = (width) => {
    if (typeof width === "number") {
        return `${width}px`;
    }
    if (typeof width === "string" && /^\d+$/.test(width)) {
        return `${width}px`;
    }
    return width;
};
exports.formatWidth = formatWidth;
var SidePanelCloseReason;
(function (SidePanelCloseReason) {
    SidePanelCloseReason["Escape"] = "Escape";
    SidePanelCloseReason["ClickOutside"] = "ClickOutside";
})(SidePanelCloseReason || (exports.SidePanelCloseReason = SidePanelCloseReason = {}));
exports.CUSTOM_POPUP_COMPONENT_CLASSNAME = "zen-custom-component-popup";
const SidePanel = ({ label, className, id, isOpen, onHidePanel, children, triggerRef, panelPosition = "right", useTrapFocusWithTrigger, width, onTransitionEnd }) => {
    const { renderComponent, showContent, readyForFocus, setIsOpen, handleTransitionEnd } = (0, useFadeComponent_1.useFadeComponent)(isOpen, onTransitionEnd);
    const { dark } = (0, react_1.useContext)(themeContext_1.themeContext);
    const isMobile = (0, useMobile_1.useMobile)();
    const sidePanelRef = (0, react_1.useRef)(null);
    const driveClassName = (0, useDriveClassName_1.useDriveClassName)("zen-side-panel");
    const autoId = (0, useUniqueId_1.useUniqueId)();
    const popupId = id || autoId;
    const prevReadyForFocus = (0, react_1.useRef)(false);
    (0, react_1.useEffect)(() => {
        if (triggerRef.current) {
            triggerRef.current.setAttribute("aria-haspopup", "true");
            triggerRef.current.setAttribute("aria-expanded", renderComponent ? "true" : "false");
            triggerRef.current.setAttribute("aria-controls", popupId);
            triggerRef.current.setAttribute("data-popup-id", popupId);
        }
    }, [triggerRef, renderComponent, popupId]);
    function handleEscape() {
        onHidePanel(SidePanelCloseReason.Escape);
        if (!triggerRef.current) {
            return;
        }
        if ((0, dialogHelpers_1.isFocusable)(triggerRef.current)) {
            triggerRef.current.focus();
            return;
        }
        const focusable = Array.from(triggerRef.current.querySelectorAll(focusableSelector_1.FOCUSABLE_SELECTOR));
        (focusable[0] || triggerRef.current).focus();
    }
    (0, useTrapFocus_1.useTrapFocus)(sidePanelRef, useTrapFocusWithTrigger ? triggerRef : undefined, showContent);
    (0, useEscape_1.useEscape)(sidePanelRef, handleEscape, renderComponent);
    (0, react_1.useEffect)(() => {
        var _a, _b;
        // eslint-disable-next-line complexity
        const closeOnClickOutside = (e) => {
            if (!sidePanelRef.current || !e.target) {
                return;
            }
            const dataShieldId = e.target.getAttribute("data-shield-id") || "";
            if (dataShieldId) {
                onHidePanel(SidePanelCloseReason.ClickOutside, dataShieldId);
                return;
            }
            const isGroupsFilterBackButton = [
                "zen-groups-filter-menu__currently-selected-button",
                "zen-groups-filter-menu__action-button-back"
            ]
                .some((groupsFilterButton) => e.target.classList.contains(groupsFilterButton));
            const isOpenAdvancedDialog = document.querySelector(".zen-advanced-groups-filter") !== null;
            const isOpenDropdownMobile = document.querySelector(".zen-dropdown--mobile");
            const isOpenPeriodPicker = document.querySelector(".zen-filters-bar-period-picker__popup");
            const isOpenTimePicker = document.querySelector(".zen-time-picker__popup");
            const isOpenYearPicker = document.querySelector(".zen-parallel-selection-popup");
            const openedCustomComponents = document.querySelectorAll(`.${exports.CUSTOM_POPUP_COMPONENT_CLASSNAME}`);
            const isClickInChildPopup = (0, exports.isChildPopup)(e.target, sidePanelRef.current)
                || isOpenDropdownMobile && (0, isChildOf_1.isChildOf)(e.target, isOpenDropdownMobile)
                || isOpenPeriodPicker && (0, isChildOf_1.isChildOf)(e.target, isOpenPeriodPicker)
                || isOpenPeriodPicker && isOpenTimePicker && (0, isChildOf_1.isChildOf)(e.target, isOpenTimePicker)
                || isOpenPeriodPicker && isOpenYearPicker && (0, isChildOf_1.isChildOf)(e.target, isOpenYearPicker)
                || openedCustomComponents.length && Array.from(openedCustomComponents).some(el => (0, isChildOf_1.isChildOf)(e.target, el));
            const isClickInPopup = !isClickInChildPopup
                ? false
                : isClickInChildPopup;
            const isClickOnTrigger = triggerRef.current ? e.target === triggerRef.current || (0, isChildOf_1.isChildOf)(e.target, triggerRef.current) : false;
            if (isClickInPopup || isClickOnTrigger || isGroupsFilterBackButton || isOpenAdvancedDialog) {
                return;
            }
            onHidePanel(SidePanelCloseReason.ClickOutside);
        };
        if (renderComponent && readyForFocus && !prevReadyForFocus.current) {
            const firstFocusable = (_a = sidePanelRef.current) === null || _a === void 0 ? void 0 : _a.querySelector(focusableSelector_1.FOCUSABLE_SELECTOR);
            prevReadyForFocus.current = readyForFocus;
            if (!useTrapFocusWithTrigger) {
                firstFocusable ? firstFocusable.focus() : (_b = sidePanelRef.current) === null || _b === void 0 ? void 0 : _b.focus();
            }
        }
        if (renderComponent) {
            document.body.addEventListener("click", closeOnClickOutside, true);
        }
        else {
            prevReadyForFocus.current = false;
        }
        return () => document.body.removeEventListener("click", closeOnClickOutside, true);
    }, [renderComponent, readyForFocus, onHidePanel, triggerRef, useTrapFocusWithTrigger]);
    (0, react_1.useEffect)(() => {
        setIsOpen(isOpen);
    }, [isOpen, setIsOpen]);
    const positioningClass = panelPosition === "left"
        ? "zen-side-panel--left-positioned"
        : panelPosition === "right" ? "zen-side-panel--right-positioned" : "zen-side-panel--bottom-positioned";
    const memoizedChildForRender = (0, react_1.useMemo)(() => isOpen || renderComponent ? children : null, [children, isOpen, renderComponent]);
    const memoizedStyles = (0, react_1.useMemo)(() => {
        if (width && !isMobile && panelPosition !== "bottom") {
            return { style: { width: (0, exports.formatWidth)(width) } };
        }
        return {};
    }, [width, isMobile, panelPosition]);
    if (!renderComponent && !isOpen) {
        return null;
    }
    return (0, react_dom_1.createPortal)((0, jsx_runtime_1.jsx)("div", Object.assign({ ref: sidePanelRef, id: popupId }, memoizedStyles, { role: "dialog", "aria-label": label, onTransitionEnd: handleTransitionEnd, className: (0, classNames_1.classNames)([
            "zen-side-panel",
            driveClassName || "",
            dark ? "zen-dark" : "",
            showContent ? "zen-side-panel--shown" : "",
            positioningClass,
            className !== null && className !== void 0 ? className : ""
        ]), children: memoizedChildForRender })), document.fullscreenElement || document.body);
};
exports.SidePanel = SidePanel;
exports.SidePanel.Cell = sidePanelCell_1.SidePanelCell;
