// @geotab/zenith/dist/progressBar/progressBarUtils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getProgressBarAttr = exports.getProgressBarValue = exports.isValidProggressBar = void 0;
const progressBar_1 = require("./progressBar");
const PROGRESS_BAR_PRECISION = 2;
const isValidProggressBar = (min, max, now) => !(min > max || now < min);
exports.isValidProggressBar = isValidProggressBar;
const getProgressBarValue = (min, max, now) => {
    if (!(0, exports.isValidProggressBar)(min, max, now)) {
        return null;
    }
    const result = (now - min) / (max - min) * 100;
    return (result > 0 ? result : 0).toFixed(PROGRESS_BAR_PRECISION);
};
exports.getProgressBarValue = getProgressBarValue;
const getProgressBarAttr = (type, min, max, now) => {
    const progressBarAttr = {
        className: type === progressBar_1.ProgressBarType.Average ? "zen-progress__bar" : `zen-progress__bar zen-progress__bar--${type}`,
        role: "progressbar",
        "aria-valuemin": min,
        "aria-valuemax": max
    };
    if (now) {
        progressBarAttr["aria-valuenow"] = now;
        progressBarAttr["style"] = { width: `${now}%` };
    }
    return progressBarAttr;
};
exports.getProgressBarAttr = getProgressBarAttr;
