// @geotab/zenith/dist/progressBar/progressBar.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProgressBar = exports.ProgreesBarSize = exports.ProgressBarType = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const progressBarUtils_1 = require("./progressBarUtils");
var ProgressBarType;
(function (ProgressBarType) {
    ProgressBarType["Low"] = "low";
    ProgressBarType["Average"] = "average";
    ProgressBarType["High"] = "high";
})(ProgressBarType || (exports.ProgressBarType = ProgressBarType = {}));
exports.ProgreesBarSize = {
    Small: "small",
    Medium: "medium",
    Large: "large"
};
const ProgressBar = ({ type = ProgressBarType.Average, min = 0, max = 100, now = 0, className = "", size = exports.ProgreesBarSize.Large }) => ((0, jsx_runtime_1.jsx)("div", { className: `zen-progress ${className} ${"zen-progress--" + size}`, children: (0, jsx_runtime_1.jsx)("div", Object.assign({}, (0, progressBarUtils_1.getProgressBarAttr)(type, min, max, (0, progressBarUtils_1.getProgressBarValue)(min, max, now)))) }));
exports.ProgressBar = ProgressBar;
