// @geotab/zenith/dist/gridLayout/utils/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MOBILE_START_UNSET_CLASS = exports.TABLET_START_UNSET_CLASS = exports.DESKTOP_START_UNSET_CLASS = exports.MOBILE_UNSET_CLASS = exports.TABLET_UNSET_CLASS = exports.DESKTOP_UNSET_CLASS = exports.MOBILE_COLS_NUMBER = exports.TABLET_COLS_NUMBER = exports.DESKTOP_COLS_NUMBER = exports.MOBILE_PATTERN = exports.MOBILE_START_PATTERN = exports.TABLET_START_PATTERN = exports.TABLET_PATTERN = exports.DESKTOP_START_PATTERN = exports.DESKTOP_PATTERN = exports.GRID_COL_START_PATTERN = exports.GRID_COL_PATTERN = exports.GRID_COL_START_PREFIX = exports.GRID_COL_PREFIX = exports.MOBILE_START_PREFIX = exports.TABLET_START_PREFIX = exports.DESKTOP_START_PREFIX = exports.MOBILE_PREFIX = exports.TABLET_PREFIX = exports.DESKTOP_PREFIX = void 0;
exports.DESKTOP_PREFIX = "zen-lg-";
exports.TABLET_PREFIX = "zen-md-";
exports.MOBILE_PREFIX = "zen-sm-";
exports.DESKTOP_START_PREFIX = `${exports.DESKTOP_PREFIX}start-`;
exports.TABLET_START_PREFIX = `${exports.TABLET_PREFIX}start-`;
exports.MOBILE_START_PREFIX = `${exports.MOBILE_PREFIX}start-`;
exports.GRID_COL_PREFIX = "zen-col-";
exports.GRID_COL_START_PREFIX = `${exports.GRID_COL_PREFIX}start-`;
exports.GRID_COL_PATTERN = new RegExp(`^${exports.GRID_COL_PREFIX}\\d+$`, "g");
exports.GRID_COL_START_PATTERN = new RegExp(`^${exports.GRID_COL_START_PREFIX}\\d+$`, "g");
exports.DESKTOP_PATTERN = new RegExp(`^${exports.DESKTOP_PREFIX}\\d+$`, "g");
exports.DESKTOP_START_PATTERN = new RegExp(`^${exports.DESKTOP_START_PREFIX}\\d+$`, "g");
exports.TABLET_PATTERN = new RegExp(`^${exports.TABLET_PREFIX}\\d+$`, "g");
exports.TABLET_START_PATTERN = new RegExp(`^${exports.TABLET_START_PREFIX}\\d+$`, "g");
exports.MOBILE_START_PATTERN = new RegExp(`^${exports.MOBILE_START_PREFIX}\\d+$`, "g");
exports.MOBILE_PATTERN = new RegExp(`^${exports.MOBILE_PREFIX}\\d+$`, "g");
exports.DESKTOP_COLS_NUMBER = 12;
exports.TABLET_COLS_NUMBER = 8;
exports.MOBILE_COLS_NUMBER = 4;
exports.DESKTOP_UNSET_CLASS = `${exports.DESKTOP_PREFIX}unset`;
exports.TABLET_UNSET_CLASS = `${exports.TABLET_PREFIX}unset`;
exports.MOBILE_UNSET_CLASS = `${exports.MOBILE_PREFIX}unset`;
exports.DESKTOP_START_UNSET_CLASS = `${exports.DESKTOP_START_PREFIX}unset`;
exports.TABLET_START_UNSET_CLASS = `${exports.TABLET_START_PREFIX}unset`;
exports.MOBILE_START_UNSET_CLASS = `${exports.MOBILE_START_PREFIX}unset`;
