// @geotab/zenith/dist/react-chartjs/dateFns/dateFns.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.format = exports.addYears = exports.addQuarters = exports.addWeeks = exports.addHours = exports.addMinutes = exports.addSeconds = exports.addMilliseconds = exports.addMonths = exports.addDays = void 0;
function addDays(date, amount) {
    let result;
    if (date instanceof Date) {
        result = new Date(date.getTime());
    }
    else {
        result = new Date(date);
    }
    result.setDate(result.getDate() + amount);
    return result;
}
exports.addDays = addDays;
function addMonths(date, amount) {
    let result;
    if (date instanceof Date) {
        result = new Date(date.getTime());
    }
    else {
        result = new Date(date);
    }
    result.setMonth(result.getMonth() + amount);
    return result;
}
exports.addMonths = addMonths;
function addMilliseconds(date, amount) {
    let result;
    if (date instanceof Date) {
        result = new Date(date.getTime());
    }
    else {
        result = new Date(date);
    }
    result.setTime(result.getTime() + amount);
    return result;
}
exports.addMilliseconds = addMilliseconds;
function addSeconds(date, amount) {
    let result;
    if (date instanceof Date) {
        result = new Date(date.getTime());
    }
    else {
        result = new Date(date);
    }
    result.setSeconds(result.getSeconds() + amount);
    return result;
}
exports.addSeconds = addSeconds;
function addMinutes(date, amount) {
    let result;
    if (date instanceof Date) {
        result = new Date(date.getTime());
    }
    else {
        result = new Date(date);
    }
    result.setMinutes(result.getMinutes() + amount);
    return result;
}
exports.addMinutes = addMinutes;
function addHours(date, amount) {
    let result;
    if (date instanceof Date) {
        result = new Date(date.getTime());
    }
    else {
        result = new Date(date);
    }
    result.setHours(result.getHours() + amount);
    return result;
}
exports.addHours = addHours;
function addWeeks(date, amount) {
    let result;
    if (date instanceof Date) {
        result = new Date(date.getTime());
    }
    else {
        result = new Date(date);
    }
    result.setDate(result.getDate() + 7 * amount);
    return result;
}
exports.addWeeks = addWeeks;
function addQuarters(date, amount) {
    let result;
    if (date instanceof Date) {
        result = new Date(date.getTime());
    }
    else {
        result = new Date(date);
    }
    result.setMonth(result.getMonth() + 3 * amount);
    return result;
}
exports.addQuarters = addQuarters;
function addYears(date, amount) {
    let result;
    if (date instanceof Date) {
        result = new Date(date.getTime());
    }
    else {
        result = new Date(date);
    }
    result.setFullYear(result.getFullYear() + amount);
    return result;
}
exports.addYears = addYears;
const monthNames = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"];
const quarterNames = ["Q1", "Q2", "Q3", "Q4"];
const formatHours = (date) => date.getHours() > 12 ? date.getHours() - 12 : date.getHours();
function getWeekRange(date) {
    const startDate = new Date(date);
    const endDate = new Date(date);
    startDate.setDate(date.getDate() - 3);
    endDate.setDate(date.getDate() + 3);
    const formattedStartDate = monthNames[startDate.getMonth()] + " " + startDate.getDate() + ", " + startDate.getFullYear();
    const formattedEndDate = monthNames[endDate.getMonth()] + " " + endDate.getDate() + ", " + endDate.getFullYear();
    return `${formattedStartDate} – ${formattedEndDate}`;
}
function format(date, formatString) {
    let result;
    if (date instanceof Date) {
        result = new Date(date.getTime());
    }
    else {
        result = new Date(date);
    }
    const options = {
        "yyyy": result.getFullYear(),
        "MM": String(result.getMonth() + 1).padStart(2, "0"),
        "d": result.getDate(),
        "h": formatHours(result),
        "ha": formatHours(result) + " " + (result.getHours() >= 12 ? "PM" : "AM"),
        "mm": String(result.getMinutes()).padStart(2, "0"),
        "ss": String(result.getSeconds()).padStart(2, "0"),
        "SSS": String(result.getMilliseconds()).padStart(3, "0"),
        "aaaa": result.getHours() >= 12 ? "PM" : "AM",
        "MMM": monthNames[result.getMonth()],
        "PP": getWeekRange(result),
        "qqq": quarterNames[Math.floor(result.getMonth() / 3)],
        "qqq - yyyy": quarterNames[Math.floor(result.getMonth() / 3)] + " - " + result.getFullYear()
    };
    let formattedDate = formatString;
    const keys = Object.keys(options).sort((a, b) => b.length - a.length);
    keys.forEach((key) => {
        if (options.hasOwnProperty(key)) {
            formattedDate = formattedDate.replace(key, options[key]);
        }
    });
    return formattedDate;
}
exports.format = format;
