// @geotab/zenith/dist/alertRaw/alertRaw.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TRANSLATIONS = exports.AlertRaw = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const alertAnimation_1 = __importDefault(require("./components/alertAnimation"));
const classNames_1 = require("../commonHelpers/classNames/classNames");
const buttonType_1 = require("../button/buttonType");
const iconInfoCircle_1 = require("../icons/iconInfoCircle");
const iconCheckRadio_1 = require("../icons/iconCheckRadio");
const useLanguage_1 = require("../utils/localization/useLanguage");
const useDrive_1 = require("../utils/theme/useDrive");
const useMobile_1 = require("../commonHelpers/hooks/useMobile");
const iconClose_1 = require("../icons/iconClose");
const iconWarning_1 = require("../icons/iconWarning");
const iconException_1 = require("../icons/iconException");
const themeContext_1 = require("../utils/theme/themeContext");
const textIconButton_1 = require("../textIconButton/textIconButton");
const button_1 = require("../button/button");
const typedIcons = {
    error: iconException_1.IconException,
    success: iconCheckRadio_1.IconCheckRadio,
    warning: iconWarning_1.IconWarning,
    info: iconInfoCircle_1.IconInfoCircle
};
const getIcon = (t) => typedIcons[t];
exports.AlertRaw = 
// eslint-disable-next-line @typescript-eslint/naming-convention, complexity
(0, react_1.forwardRef)(function InnerAlertRaw({ id, type, header, text, buttons, className, onClose, animation, isOpen }, ref) {
    const { translate } = (0, useLanguage_1.useLanguage)();
    const isDrive = (0, useDrive_1.useDrive)();
    const isMobile = (0, useMobile_1.useMobile)();
    const { dark } = (0, react_1.useContext)(themeContext_1.themeContext);
    const alertClasses = (0, classNames_1.classNames)([
        "zen-alert",
        `zen-alert--${type || "info"}`,
        isDrive ? "zen-alert--drive" : "",
        isMobile ? "zen-alert--mobile" : "",
        dark ? "zen-dark" : "",
        header ? "zen-alert--with-header" : "",
        className || ""
    ]);
    const onFooterPrimaryClick = (0, react_1.useCallback)(() => {
        var _a, _b;
        (_b = (_a = buttons.primary).onClick) === null || _b === void 0 ? void 0 : _b.call(_a);
        onClose();
    }, [buttons, onClose]);
    const onFooterTertiaryClick = (0, react_1.useCallback)(() => {
        var _a, _b;
        (_b = (_a = buttons.tertiary) === null || _a === void 0 ? void 0 : _a.onClick) === null || _b === void 0 ? void 0 : _b.call(_a);
        onClose();
    }, [buttons, onClose]);
    const primaryButton = (0, react_1.useMemo)(() => {
        if (buttons.primary.icon) {
            return (0, jsx_runtime_1.jsx)(textIconButton_1.TextIconButton, { iconPosition: buttons.primary.iconPosition || textIconButton_1.ButtonIconPosition.Start, icon: buttons.primary.icon, onClick: onFooterPrimaryClick, link: buttons.primary.link, className: "zen-alert__button zen-alert__primary-button zen-alert__primary-button--" + (type || "info"), title: buttons.primary.title, type: buttonType_1.ButtonType.Secondary, children: buttons.primary.text });
        }
        return (0, jsx_runtime_1.jsx)(button_1.Button, { onClick: onFooterPrimaryClick, link: buttons.primary.link, className: "zen-alert__button zen-alert__primary-button zen-alert__primary-button--" + (type || "info"), title: buttons.primary.title, type: buttonType_1.ButtonType.Secondary, children: buttons.primary.text });
    }, [buttons.primary.icon, buttons.primary.iconPosition, buttons.primary.link, buttons.primary.text, buttons.primary.title, onFooterPrimaryClick, type]);
    const tertiaryButton = (0, react_1.useMemo)(() => {
        if (!buttons.tertiary) {
            return null;
        }
        if (buttons.tertiary.icon) {
            return (0, jsx_runtime_1.jsx)(textIconButton_1.TextIconButton, { iconPosition: buttons.tertiary.iconPosition || textIconButton_1.ButtonIconPosition.Start, icon: buttons.tertiary.icon, title: buttons.tertiary.title, type: buttonType_1.ButtonType.Tertiary, className: `zen-alert__button zen-alert__tertiary-button zen-alert__tertiary-button--${type || "info"}`, onClick: onFooterTertiaryClick, link: buttons.tertiary.link, children: buttons.tertiary.text });
        }
        return (0, jsx_runtime_1.jsx)(button_1.Button, { title: buttons.tertiary.title, type: buttonType_1.ButtonType.Tertiary, className: `zen-alert__button zen-alert__tertiary-button zen-alert__tertiary-button--${type || "info"}`, onClick: onFooterTertiaryClick, link: buttons.tertiary.link, children: buttons.tertiary.text });
    }, [buttons.tertiary, onFooterTertiaryClick, type]);
    return (0, jsx_runtime_1.jsx)(alertAnimation_1.default, { id: id, isOpen: isOpen, animation: animation, children: (0, jsx_runtime_1.jsx)("div", { role: "status", "aria-atomic": "true", "aria-live": "polite", ref: ref, className: alertClasses, id: id, children: (0, jsx_runtime_1.jsxs)("div", { className: "zen-alert__body-container", children: [(0, jsx_runtime_1.jsx)("div", { className: "zen-alert__content", children: (0, jsx_runtime_1.jsx)("div", { className: "zen-alert__header-container", children: (0, jsx_runtime_1.jsxs)("div", { className: "zen-alert__header", children: [(0, jsx_runtime_1.jsx)("div", { className: "zen-alert__icon", children: (0, react_1.createElement)(getIcon(type || "info"), {
                                            size: isDrive ? "huge" : "large",
                                            className: `zen-alert__icon--${type || "info"}`
                                        }) }), (0, jsx_runtime_1.jsxs)("div", { className: "zen-alert__body", children: [header && (0, jsx_runtime_1.jsx)("div", { className: `zen-alert__title zen-alert__title--${type || "info"}`, children: header }), (0, jsx_runtime_1.jsx)("div", { className: "zen-alert__text", children: text })] }), (0, jsx_runtime_1.jsx)("div", { className: "zen-alert__actions", children: (0, jsx_runtime_1.jsx)("button", { type: "button", className: (0, classNames_1.classNames)(["zen-alert__close-button", `zen-alert__close-button--${type || "info"}`]), onClick: onClose, title: translate("Close"), "aria-label": translate("Close"), children: (0, jsx_runtime_1.jsx)(iconClose_1.IconClose, { size: isDrive ? "huge" : "large" }) }) })] }) }) }), (0, jsx_runtime_1.jsxs)("div", { className: "zen-alert__body-actions", children: [tertiaryButton, primaryButton] })] }) }) });
});
exports.TRANSLATIONS = [
    "Close"
];
