// @geotab/zenith/dist/groupsFilter/groupsFilterHelper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.comparePreparedFilters = exports.isFilterHasSelection = exports.doesGroupExist = exports.getAdvancedFilterString = exports.createGroupsMap = exports.getChildList = exports.prepareMenu = exports.prepareSelectedMenu = exports.prepareNameSelectedItems = exports.getParentsNameArr = exports.prepareAdvancedOptions = exports.prepareSelectedOptions = exports.prepareOptions = exports.convertPreparedFilters = exports.prepareFilters = void 0;
const utils_1 = require("../commonHelpers/utils");
const groupsFilterInterfaces_1 = require("./groupsFilterInterfaces");
const groupsHelper_1 = require("./groupsHelper");
const isAdvancedFilterItem = (item) => item && item.relation !== undefined;
function prepareFilters(state) {
    const canBeSimplify = (item, relation) => item.items.length === 1 && item.items[0].id && item.relation === relation;
    const advancedFilterItems = [];
    const simpleListItems = [];
    state.items.forEach(el => {
        if (!isAdvancedFilterItem(el)) {
            el.id !== groupsHelper_1.ENTIRE_ORGANIZATION_GROUP_ID && simpleListItems.push(el.id);
        }
        else {
            (canBeSimplify(el, state.relation) ? el.items[0].id !== groupsHelper_1.ENTIRE_ORGANIZATION_GROUP_ID && simpleListItems.push(el.items[0].id)
                : advancedFilterItems.push(el));
        }
    });
    return {
        advancedFilterState: advancedFilterItems.length ? { relation: state.relation, items: [...advancedFilterItems] } : undefined,
        simpleFilterState: simpleListItems.length ? Array.from(new Set(simpleListItems)) : undefined,
        relation: state.relation
    };
}
exports.prepareFilters = prepareFilters;
function convertPreparedFilters(state) {
    const convertedItems = [];
    if (state.advancedFilterState && state.advancedFilterState.items[0].id !== groupsHelper_1.ENTIRE_ORGANIZATION_GROUP_ID) {
        convertedItems.push(state.advancedFilterState);
    }
    if (state.simpleFilterState && state.simpleFilterState.length > 0) {
        state.simpleFilterState.forEach(el => {
            convertedItems.push({ id: el });
        });
    }
    return { relation: state.relation, items: convertedItems.length ? [...convertedItems] : [{ id: groupsHelper_1.ENTIRE_ORGANIZATION_GROUP_ID }] };
}
exports.convertPreparedFilters = convertPreparedFilters;
const findGroupsChildren = (groupsMap, id, parentGroupId) => {
    var _a;
    const group = groupsMap.get(id);
    if (!group) {
        return undefined;
    }
    const foundItem = Object.assign(Object.assign({}, group), { parent: parentGroupId ? Object.assign({}, groupsMap.get(parentGroupId)) : undefined });
    if (foundItem.children && ((_a = foundItem.children) === null || _a === void 0 ? void 0 : _a.length) > 0) {
        foundItem.children = foundItem.children
            .map(child => findGroupsChildren(groupsMap, child.id, foundItem.id))
            .filter(item => !!item);
    }
    return foundItem;
};
function prepareOptions(translate, options, getSelectedFn, getSelectedChildFn, isWithColor, getTitle) {
    const selectedIds = new Set(getSelectedFn());
    const preparedOptions = options && options.length ? options.map(option => {
        const selectedChild = getSelectedChildFn(option.id);
        return Object.assign(Object.assign({ label: (0, groupsHelper_1.getGroupDescription)(option, translate), property: option.id, checked: selectedIds.has(option.id), isWithAction: option.children && option.children.length > 0 || false, partialChecked: selectedChild.length > 0, countSelectedChild: selectedChild.length, blocked: false }, (getTitle ? { title: getTitle(option.id) } : {})), (isWithColor ? { color: option.color } : {}));
    }) : [];
    return preparedOptions;
}
exports.prepareOptions = prepareOptions;
function prepareSelectedOptions(translate, options, getSelectedFn, isWithColor, getTitle) {
    const selectedIds = new Set(getSelectedFn());
    const preparedOptions = options && options.length ? options.map(option => (Object.assign(Object.assign({ label: (0, groupsHelper_1.getGroupDescription)(option, translate), property: option.id, checked: selectedIds.has(option.id), isWithAction: false, partialChecked: false, blocked: false }, (getTitle ? { title: getTitle(option.id) } : {})), (isWithColor ? { color: option.color } : {})))) : [];
    return preparedOptions;
}
exports.prepareSelectedOptions = prepareSelectedOptions;
function prepareAdvancedOptions(options, currentAdvancedOption) {
    const preparedOptions = options && options.length ? options.map(option => ({
        label: option,
        property: option,
        checked: option === currentAdvancedOption,
        isWithAction: false,
        partialChecked: false,
        blocked: false
    })) : [];
    return preparedOptions;
}
exports.prepareAdvancedOptions = prepareAdvancedOptions;
function getParentsNameArr(translate, groupsMap, itemId, rootId) {
    const item = groupsMap[itemId];
    if (!item) {
        return [];
    }
    const name = [(0, groupsHelper_1.getGroupDescription)(item, translate)];
    let selectedParent = item.parent ? groupsMap[item.parent.id] : undefined;
    while (selectedParent && selectedParent.id !== rootId) {
        name.push((0, groupsHelper_1.getGroupDescription)(selectedParent, translate));
        selectedParent = selectedParent.parent ? groupsMap[selectedParent.parent.id] : undefined;
    }
    const preparedNameArr = name.length > 1 ? name.reverse() : name;
    return preparedNameArr;
}
exports.getParentsNameArr = getParentsNameArr;
function prepareNameSelectedItems(translate, groupsMap, selectedIds) {
    const selectedItems = selectedIds.map(id => (groupsMap[id] ? Object.assign({}, groupsMap[id]) : undefined)).filter(utils_1.echo);
    const namedItems = selectedItems.map(el => {
        const preparedNameArr = getParentsNameArr(translate, groupsMap, el.id, groupsHelper_1.ENTIRE_ORGANIZATION_GROUP_ID);
        return Object.assign(Object.assign({}, el), { name: ["...", ...(preparedNameArr.length > 1 ? preparedNameArr.slice(-2) : preparedNameArr)].join("/") });
    });
    return namedItems;
}
exports.prepareNameSelectedItems = prepareNameSelectedItems;
function prepareSelectedMenu(namedItems) {
    const systemGroup = [];
    const userGroup = [];
    namedItems.forEach(el => (0, groupsHelper_1.isSystemGroup)(el.id)
        ? systemGroup.push(el) : userGroup.push(el));
    return { builtInGroups: [...systemGroup], userGroups: [...userGroup], currentId: undefined };
}
exports.prepareSelectedMenu = prepareSelectedMenu;
function prepareMenu(item) {
    const systemGroup = [];
    const userGroup = [];
    if (!item) {
        return { builtInGroups: [], userGroups: [], currentId: undefined };
    }
    (item.children || []).forEach(el => (0, groupsHelper_1.isSystemGroup)(el.id)
        ? systemGroup.push(el) : userGroup.push(el));
    const result = { builtInGroups: [...systemGroup], userGroups: [...userGroup], currentId: item.id };
    return result;
}
exports.prepareMenu = prepareMenu;
function createGroupsTree(groupsMap, id) {
    let groupsTree = findGroupsChildren(groupsMap, id);
    if (!groupsTree) {
        const tree = (0, groupsHelper_1.createLinkedTree)(Array.from(groupsMap.values())).map((el) => (Object.assign(Object.assign({}, el), { parent: undefined })));
        groupsTree = { id: id, name: "", children: tree.map(el => (Object.assign(Object.assign({}, el), { parent: { id: id, name: "", parent: undefined } }))), parent: undefined };
    }
    groupsTree.parent = { id: "root", name: "", children: [Object.assign({}, groupsTree)], parent: undefined };
    return groupsTree;
}
function getChildList(childArr) {
    const list = [];
    if (!childArr) {
        return [];
    }
    (childArr || []).forEach((el) => {
        el && list.push([el]);
        if ((el === null || el === void 0 ? void 0 : el.children) && el.children.length > 0) {
            const childList = getChildList(el.children);
            childList && list.push(childList);
        }
    });
    return (0, utils_1.flattenArrays)(list);
}
exports.getChildList = getChildList;
function createGroupsMap(groups, rootChildId) {
    var _a;
    const groupsMap = new Map();
    const groupTreeMap = {};
    groups.forEach(group => {
        groupsMap.set(group.id, Object.assign({}, group));
    });
    const groupsTree = createGroupsTree(groupsMap, rootChildId);
    if (!groupsTree) {
        return {};
    }
    groupTreeMap[rootChildId] = Object.assign({}, groupsTree);
    (_a = groupsTree.children) === null || _a === void 0 ? void 0 : _a.forEach(child => {
        groupTreeMap[child.id] = Object.assign({}, child);
        const childList = getChildList(child.children || []);
        childList.forEach(el => {
            if (el) {
                groupTreeMap[el.id] = Object.assign({}, el);
            }
        });
    });
    return groupTreeMap;
}
exports.createGroupsMap = createGroupsMap;
function getAdvancedFilterString(translate, state) {
    const getRelationStr = (relation) => relation === groupsFilterInterfaces_1.RelationOperator.OR ? translate("OperatorOr") : translate("OperatorAnd");
    const getItemState = (itemState, prev = "") => {
        const items = itemState.items;
        if (items && items.length && isAdvancedFilterItem(items[0])) {
            return items.reduce((res, item) => {
                const prevAndRelationStr = `${res
                    ? `${res} ${getRelationStr(itemState.relation)} `
                    : ""}`;
                return `${prevAndRelationStr}${getItemState(item, res)}`;
            }, prev);
        }
        const str = items.map(item => `${item.name}`)
            .join(` ${getRelationStr(itemState.relation)} `);
        return `${items && items.length > 1 ? `(${str})` : str}`;
    };
    return `${getItemState(state)}`;
}
exports.getAdvancedFilterString = getAdvancedFilterString;
function doesGroupExist(groupMap) {
    return (id) => !!groupMap[id];
}
exports.doesGroupExist = doesGroupExist;
const isFilterHasSelection = (filter) => {
    var _a;
    if ((_a = filter.simpleFilterState) === null || _a === void 0 ? void 0 : _a.length) {
        return true;
    }
    if (!(0, groupsHelper_1.compareFilters)(filter.advancedFilterState || (0, groupsHelper_1.getDefaultFilterState)(), (0, groupsHelper_1.getDefaultFilterState)())) {
        return true;
    }
    return false;
};
exports.isFilterHasSelection = isFilterHasSelection;
function comparePreparedFilters(f1, f2) {
    if (f1.relation !== f2.relation || (f1.simpleFilterState || []).length !== (f2.simpleFilterState || []).length) {
        return false;
    }
    const isAdvancedFilterTheSame = (0, groupsHelper_1.compareFilters)(f1.advancedFilterState || (0, groupsHelper_1.getDefaultFilterState)(), f2.advancedFilterState || (0, groupsHelper_1.getDefaultFilterState)());
    if (!isAdvancedFilterTheSame) {
        return false;
    }
    if (!(f1.simpleFilterState || []).every(el => (f2.simpleFilterState || []).includes(el))) {
        return false;
    }
    return true;
}
exports.comparePreparedFilters = comparePreparedFilters;
