// @geotab/zenith/dist/groupsFilter/groupsFilter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TRANSLATIONS = exports.GroupsFilter = exports.groupsFilterTotalStateDefault = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const groupsFilterInterfaces_1 = require("./groupsFilterInterfaces");
const groupsFilterMenu_1 = require("./groupsFilterMenu");
const groupsFilterInitialState_1 = require("./groupsFilterInitialState");
const groupsFilterAdjustmentState_1 = require("./groupsFilterAdjustmentState");
const groupsFilterCurrentlySelectedState_1 = require("./groupsFilterCurrentlySelectedState");
const stateReducer_1 = require("./stateReducer/stateReducer");
const stateActionType_1 = require("./stateReducer/stateActionType");
const groupsHelper_1 = require("./groupsHelper");
const groupsFilterStatesHelper_1 = require("./groupsFilterStatesHelper");
const groupsFilterBox_1 = require("./groupsFilterBox");
const groupsFilterHelper_1 = require("./groupsFilterHelper");
const groupsFilterTrigger_1 = require("./groupsFilterTrigger");
const useEscape_1 = require("../commonHelpers/hooks/useEscape");
const focusableSelector_1 = require("../utils/focusableSelector");
const controlledPopup_1 = require("../controlledPopup/controlledPopup");
const stateReducerHelper_1 = require("./stateReducer/stateReducerHelper");
const uiStateReducer_1 = require("./uiStateReducer/uiStateReducer");
const uiStateActionType_1 = require("./uiStateReducer/uiStateActionType");
const classNames_1 = require("../commonHelpers/classNames/classNames");
const advancedGroupsFilter_1 = require("../advancedGroupsFilter/advancedGroupsFilter");
const useLanguage_1 = require("../utils/localization/useLanguage");
const groupsFilterAdvancedValue_1 = require("./groupsFilterAdvancedValue");
const skeletonList_1 = require("../skeletonList/skeletonList");
const chip_1 = require("../chip/chip");
const searchInput_1 = require("../searchInput/searchInput");
const chipStatusProvider_1 = require("../chip/chipStatusProvider");
exports.groupsFilterTotalStateDefault = { groups: (0, groupsHelper_1.getDefaultFilterState)(), sideWide: false };
// eslint-disable-next-line complexity
const GroupsFilter = ({ fullSize, className, dataLoader, errorHandler, isWithColor, onChange, initialFilterState, options = {}, listLimit = 500, chip, chipIcon, chipId, chipName }) => {
    const { translate } = (0, useLanguage_1.useLanguage)();
    const chipStatus = (0, chipStatusProvider_1.useChipStatus)();
    const [groupPlaceHolder, setGroupPlaceHolder] = (0, react_1.useState)(options.placeholder || translate("Groups"));
    const [state, dispatchState] = (0, react_1.useReducer)(stateReducer_1.stateReducer, (0, stateReducer_1.getInitialState)(translate, initialFilterState));
    const [uiState, dispatchUiState] = (0, react_1.useReducer)(uiStateReducer_1.uiStateReducer, uiStateReducer_1.uiInitialGroupsFilterState);
    const [dataItems, setDataItems] = (0, react_1.useState)([]);
    const inputRef = (0, react_1.useRef)(null);
    const triggerRef = (0, react_1.useRef)(null);
    const menuRef = (0, react_1.useRef)(null);
    const prevInitialState = (0, react_1.useRef)(initialFilterState);
    const prevSearchField = (0, react_1.useRef)("");
    const groupsFilterId = (0, react_1.useId)();
    const blocksMap = (0, react_1.useMemo)(() => [["builtInGroups", "userGroups"], ["advancedGroups", "builtInGroups", "userGroups"], ["builtInGroups", "userGroups"]], []);
    const onOptionsToggle = () => { dispatchUiState({ type: uiStateActionType_1.UiStateActionType.ToggleOpen, payload: undefined }); };
    const handleSetStep = (0, react_1.useCallback)((newStep) => {
        var _a;
        uiState.step !== groupsFilterInterfaces_1.FilterMenuStep.CurrentlySelected && dispatchUiState({ type: uiStateActionType_1.UiStateActionType.ClearUncheckedState, payload: undefined });
        dispatchUiState({ type: uiStateActionType_1.UiStateActionType.SetStep, payload: newStep });
        if (((_a = inputRef.current) === null || _a === void 0 ? void 0 : _a.value) && newStep !== groupsFilterInterfaces_1.FilterMenuStep.Initial) {
            inputRef.current.value = "";
            dispatchUiState({ type: uiStateActionType_1.UiStateActionType.SetSearchValue, payload: "" });
        }
    }, [uiState.step]);
    const onSearchChange = (0, react_1.useCallback)((event) => {
        const newValue = event.currentTarget.value;
        dispatchUiState({ type: uiStateActionType_1.UiStateActionType.SetSearchValue, payload: newValue });
        dispatchUiState({ type: uiStateActionType_1.UiStateActionType.SetIsOpen, payload: true });
        uiState.step !== groupsFilterInterfaces_1.FilterMenuStep.Initial && handleSetStep(groupsFilterInterfaces_1.FilterMenuStep.Initial);
    }, [handleSetStep, uiState.step]);
    const onSearchChipChange = (0, react_1.useCallback)((newValue) => {
        dispatchUiState({ type: uiStateActionType_1.UiStateActionType.SetSearchValue, payload: newValue || "" });
        dispatchUiState({ type: uiStateActionType_1.UiStateActionType.SetIsOpen, payload: true });
        uiState.step !== groupsFilterInterfaces_1.FilterMenuStep.Initial && handleSetStep(groupsFilterInterfaces_1.FilterMenuStep.Initial);
    }, [handleSetStep, uiState.step]);
    const handleCancel = () => {
        var _a;
        dispatchUiState({ type: uiStateActionType_1.UiStateActionType.SetIsOpen, payload: false });
        dispatchState({ type: stateActionType_1.StateActionType.SetState, payload: state.frozenState });
        dispatchState({ type: stateActionType_1.StateActionType.SideWide, payload: state.frozenSideWideState });
        handleSetStep(groupsFilterInterfaces_1.FilterMenuStep.Initial);
        (_a = inputRef.current) === null || _a === void 0 ? void 0 : _a.focus();
    };
    const handleApply = () => {
        var _a;
        dispatchUiState({ type: uiStateActionType_1.UiStateActionType.SetIsOpen, payload: false });
        dispatchUiState({ type: uiStateActionType_1.UiStateActionType.SetSearchValue, payload: "" });
        dispatchUiState({ type: uiStateActionType_1.UiStateActionType.ClearUncheckedState, payload: undefined });
        if (inputRef.current) {
            inputRef.current.value = "";
        }
        handleSetStep(groupsFilterInterfaces_1.FilterMenuStep.Initial);
        dispatchState({ type: stateActionType_1.StateActionType.FreezeState, payload: (0, stateReducerHelper_1.getCurrentState)(state) });
        dispatchState({ type: stateActionType_1.StateActionType.FreezeSideWide, payload: state.sideWideState });
        (_a = inputRef.current) === null || _a === void 0 ? void 0 : _a.focus();
    };
    const handleReset = () => {
        dispatchState({ type: stateActionType_1.StateActionType.ResetSelection, payload: undefined });
    };
    const handleChipClose = () => {
        dispatchState({ type: stateActionType_1.StateActionType.ResetSelection, payload: undefined });
        dispatchUiState({ type: uiStateActionType_1.UiStateActionType.SetIsOpen, payload: false });
        dispatchUiState({ type: uiStateActionType_1.UiStateActionType.SetSearchValue, payload: "" });
        dispatchUiState({ type: uiStateActionType_1.UiStateActionType.ClearUncheckedState, payload: undefined });
        if (inputRef.current) {
            inputRef.current.value = "";
        }
        handleSetStep(groupsFilterInterfaces_1.FilterMenuStep.Initial);
        dispatchState({ type: stateActionType_1.StateActionType.FreezeState, payload: stateReducer_1.defaultPreparedFilterState });
        dispatchState({ type: stateActionType_1.StateActionType.FreezeSideWide, payload: false });
    };
    const handleCheckboxChange = (value) => {
        if (value.partial) {
            dispatchState({ type: stateActionType_1.StateActionType.DeselectAllChildren, payload: value.id });
        }
        else {
            dispatchState({ type: stateActionType_1.StateActionType.ChangeSelection, payload: { value: value.state, itemId: value.id } });
        }
    };
    const handleCheckboxItemClick = (id) => {
        dispatchUiState({ type: uiStateActionType_1.UiStateActionType.SetCurrentAdjustmentId, payload: id });
        handleSetStep(id ? groupsFilterInterfaces_1.FilterMenuStep.Adjustment : groupsFilterInterfaces_1.FilterMenuStep.Initial);
    };
    const handleCheckboxItemAdjustmentClick = (value) => {
        if (value.id === groupsFilterBox_1.ADVANCED_FILTER_ID) {
            const currentVal = value.state ? uiState.uncheckedAdvancedState : state.advancedFilterState;
            !value.state && dispatchUiState({ type: uiStateActionType_1.UiStateActionType.SetUncheckedAdvancedState, payload: currentVal });
            !value.state && dispatchState({ type: stateActionType_1.StateActionType.SetAdvancedFilterState, payload: (0, groupsHelper_1.getDefaultFilterState)() });
            value.state && dispatchState({ type: stateActionType_1.StateActionType.SetAdvancedFilterState, payload: currentVal || state.frozenState.advancedFilterState || (0, groupsHelper_1.getDefaultFilterState)() });
            return;
        }
        dispatchUiState({ type: uiStateActionType_1.UiStateActionType.SetUncheckedAdjustmentIds, payload: !value.state ?
                [...uiState.uncheckedAdjustmentIds, value.id] : uiState.uncheckedAdjustmentIds.filter(id => id !== value.id) });
        dispatchState({ type: stateActionType_1.StateActionType.ChangeSelection, payload: { value: value.state, itemId: value.id } });
    };
    const basicHandleBackButtonClick = () => {
        handleSetStep(groupsFilterInterfaces_1.FilterMenuStep.Initial);
    };
    const handleBackButtonClick = () => {
        const currentEl = uiState.currentAdjustmentId ? (0, stateReducerHelper_1.getGroupsTreeItem)(state, uiState.currentAdjustmentId) : undefined;
        if (!uiState.currentAdjustmentId || !(currentEl === null || currentEl === void 0 ? void 0 : currentEl.parent)) {
            basicHandleBackButtonClick();
            return;
        }
        const isNotRoot = currentEl.parent.id && currentEl.parent.id !== groupsHelper_1.ENTIRE_ORGANIZATION_GROUP_ID;
        dispatchUiState({ type: uiStateActionType_1.UiStateActionType.SetCurrentAdjustmentId, payload: isNotRoot ? currentEl.parent.id : undefined });
        handleSetStep(isNotRoot ? groupsFilterInterfaces_1.FilterMenuStep.Adjustment : groupsFilterInterfaces_1.FilterMenuStep.Initial);
    };
    const handleHomeButtonClick = () => {
        handleSetStep(groupsFilterInterfaces_1.FilterMenuStep.Initial);
        uiState.currentAdjustmentId && dispatchUiState({ type: uiStateActionType_1.UiStateActionType.SetCurrentAdjustmentId, payload: undefined });
    };
    const handleCurrentlySelectedClick = () => {
        handleSetStep(groupsFilterInterfaces_1.FilterMenuStep.CurrentlySelected);
    };
    const selectAllHandler = (isAllSelected) => {
        if (!uiState.currentAdjustmentId) {
            return;
        }
        dispatchState({ type: isAllSelected ? stateActionType_1.StateActionType.DeselectAllChildren : stateActionType_1.StateActionType.SelectAllChildren, payload: uiState.currentAdjustmentId });
    };
    const handleTriggerClick = () => {
        dispatchUiState({ type: uiStateActionType_1.UiStateActionType.ToggleOpen, payload: undefined });
    };
    const handleAdvancedDialogOpen = (0, react_1.useCallback)(() => {
        dispatchUiState({ type: uiStateActionType_1.UiStateActionType.SetAdvancedFilterIsOpen, payload: true });
    }, []);
    const handleAdvancedDialogClose = (0, react_1.useCallback)(() => {
        dispatchUiState({ type: uiStateActionType_1.UiStateActionType.SetAdvancedFilterIsOpen, payload: false });
    }, []);
    const handleAdvancedDialogApply = (0, react_1.useCallback)((val) => {
        dispatchUiState({ type: uiStateActionType_1.UiStateActionType.SetAdvancedFilterIsOpen, payload: false });
        dispatchState({ type: stateActionType_1.StateActionType.SetAdvancedFilterState, payload: val.items.length ? val : (0, groupsHelper_1.getDefaultFilterState)() });
    }, []);
    const getAdvancedData = (0, react_1.useCallback)((_, searchValue) => dataLoader().then(data => (searchValue ? data.filter(el => (el.name || el.id).toLocaleLowerCase().indexOf(searchValue.toLocaleLowerCase()) > -1) : data)).catch((e) => {
        errorHandler(e);
        return [];
    }), [dataLoader, errorHandler]);
    (0, react_1.useEffect)(() => {
        dispatchUiState({ type: uiStateActionType_1.UiStateActionType.SetLoadWithError, payload: false });
        dataLoader().then(groups => {
            dispatchState({ type: stateActionType_1.StateActionType.CreateGroupsMap, payload: groups });
            setDataItems(groups);
        }).catch((e) => {
            dispatchUiState({ type: uiStateActionType_1.UiStateActionType.SetLoadWithError, payload: true });
            errorHandler(e);
        });
    }, [dataLoader, errorHandler]);
    (0, react_1.useEffect)(() => {
        const step = uiState.step;
        if (step !== groupsFilterInterfaces_1.FilterMenuStep.CurrentlySelected) {
            return;
        }
        const groupsBlocks = (0, groupsFilterAdvancedValue_1.getAdvancedFilterValue)({
            blocks: blocksMap[uiState.step],
            state,
            isWithColor: isWithColor || false,
            uncheckedItems: uiState.step === groupsFilterInterfaces_1.FilterMenuStep.CurrentlySelected ? uiState.uncheckedAdjustmentIds : [],
            uncheckedAdvancedState: uiState.step === groupsFilterInterfaces_1.FilterMenuStep.CurrentlySelected ? uiState.uncheckedAdvancedState : undefined,
            listLimit
        });
        const label = (0, groupsFilterAdvancedValue_1.getAdvancedLabel)({ groupsBlocks });
        const selectedValues = dataItems.filter((item) => { var _a; return (_a = state.frozenState.simpleFilterState) === null || _a === void 0 ? void 0 : _a.find((id) => id === item.id); }).map((item) => item.name);
        const selectedValuesLabel = selectedValues.join(` ${translate(state.frozenState.relation)} `);
        if (!label && !selectedValuesLabel) {
            setGroupPlaceHolder(options.placeholder || translate("Groups"));
            return;
        }
        const separator = `${label && selectedValuesLabel ? ` ${translate(state.frozenState.relation)} ` : ""}`;
        setGroupPlaceHolder(`${label}${separator}${selectedValuesLabel}`);
    }, [uiState.searchValue, uiState.step, blocksMap, state, isWithColor, listLimit, uiState.currentAdjustmentId,
        uiState.uncheckedAdjustmentIds, uiState.uncheckedAdvancedState, dataItems, options.placeholder, translate]);
    const getGroupsBox = () => (0, jsx_runtime_1.jsx)(groupsFilterBox_1.GroupsFilterBox, { step: uiState.step, currentId: uiState.currentAdjustmentId, isWithColor: isWithColor || false, state: state, search: uiState.searchValue, blocks: uiState.searchValue && uiState.step === groupsFilterInterfaces_1.FilterMenuStep.Initial ? [] : blocksMap[uiState.step], onChange: uiState.step === groupsFilterInterfaces_1.FilterMenuStep.CurrentlySelected ? handleCheckboxItemAdjustmentClick : handleCheckboxChange, onClick: handleCheckboxItemClick, uncheckedItems: uiState.step === groupsFilterInterfaces_1.FilterMenuStep.CurrentlySelected ? uiState.uncheckedAdjustmentIds : [], uncheckedAdvancedState: uiState.step === groupsFilterInterfaces_1.FilterMenuStep.CurrentlySelected ? uiState.uncheckedAdvancedState : undefined, listLimit: listLimit });
    const onSideWideChange = (newState) => { dispatchState({ type: stateActionType_1.StateActionType.SideWide, payload: newState }); };
    const onRelationChange = (newRelation) => { dispatchState({ type: stateActionType_1.StateActionType.Relation, payload: newRelation }); };
    const getSearchResult = () => {
        const result = (0, stateReducerHelper_1.getSearchResultFromObject)(state, uiState.searchValue);
        return result ? (result.builtInGroups.length + result.userGroups.length).toString() : "0";
    };
    const getAdvancedDialog = (0, react_1.useMemo)(() => (0, jsx_runtime_1.jsx)(advancedGroupsFilter_1.AdvancedGroupsFilter, { isOpen: uiState.isOpenAdvancedFilter, onClose: handleAdvancedDialogClose, onApply: handleAdvancedDialogApply, state: state.advancedFilterState, getData: getAdvancedData, errorHandler: errorHandler, dataItems: dataItems }), [uiState.isOpenAdvancedFilter, handleAdvancedDialogClose, handleAdvancedDialogApply, dataItems, state.advancedFilterState, getAdvancedData, errorHandler]);
    (0, react_1.useEffect)(() => {
        if (!state.globalChanges) {
            return;
        }
        const newState = {
            groups: (0, groupsFilterHelper_1.convertPreparedFilters)(state.frozenState),
            sideWide: state.frozenSideWideState
        };
        if ((initialFilterState === null || initialFilterState === void 0 ? void 0 : initialFilterState.sideWide) === state.frozenSideWideState && (0, groupsFilterHelper_1.comparePreparedFilters)((0, groupsFilterHelper_1.prepareFilters)(initialFilterState.groups), state.frozenState)) {
            return;
        }
        const selectedValues = dataItems.filter(item => { var _a; return (_a = state.frozenState.simpleFilterState) === null || _a === void 0 ? void 0 : _a.find(id => id === item.id); }).map(item => item.name);
        setGroupPlaceHolder(selectedValues.length > 0 ? selectedValues.join(", ") : options.placeholder || translate("Groups"));
        onChange(Object.assign({}, newState));
        dispatchState({ type: stateActionType_1.StateActionType.TurnOfGlobalChanges, payload: undefined });
    }, [state.globalChanges, dataItems, translate, state.frozenState, state.frozenSideWideState, options.placeholder, onChange, initialFilterState === null || initialFilterState === void 0 ? void 0 : initialFilterState.sideWide, initialFilterState]);
    (0, react_1.useEffect)(() => {
        var _a;
        if (((_a = prevInitialState.current) === null || _a === void 0 ? void 0 : _a.sideWide) === (initialFilterState === null || initialFilterState === void 0 ? void 0 : initialFilterState.sideWide)
            && (0, groupsFilterHelper_1.comparePreparedFilters)((0, groupsFilterHelper_1.prepareFilters)(initialFilterState ? initialFilterState.groups : (0, groupsHelper_1.getDefaultFilterState)()), (0, groupsFilterHelper_1.prepareFilters)(prevInitialState.current ? prevInitialState.current.groups : (0, groupsHelper_1.getDefaultFilterState)()))) {
            return;
        }
        const stateForDispatch = (initialFilterState === null || initialFilterState === void 0 ? void 0 : initialFilterState.groups) ?
            { value: (0, groupsFilterHelper_1.prepareFilters)(initialFilterState.groups), global: initialFilterState.sideWide } : { value: (0, groupsFilterHelper_1.prepareFilters)((0, groupsHelper_1.getDefaultFilterState)()), global: false };
        if (Object.keys(state.groupsMap).length > 0) {
            prevInitialState.current = initialFilterState ? Object.assign({}, initialFilterState) : undefined;
            dispatchState({ type: stateActionType_1.StateActionType.CreateInitialState, payload: stateForDispatch });
            !(initialFilterState === null || initialFilterState === void 0 ? void 0 : initialFilterState.groups) && dispatchUiState({ type: uiStateActionType_1.UiStateActionType.SetStep, payload: groupsFilterInterfaces_1.FilterMenuStep.Initial });
        }
    }, [initialFilterState, state.groupsMap]);
    (0, react_1.useEffect)(() => {
        var _a, _b, _c;
        if (!uiState.isOpen) {
            return;
        }
        if (uiState.searchValue && uiState.step === groupsFilterInterfaces_1.FilterMenuStep.Initial || uiState.step === groupsFilterInterfaces_1.FilterMenuStep.Initial && uiState.searchValue === ""
            && prevSearchField.current !== uiState.searchValue) {
            (_a = inputRef.current) === null || _a === void 0 ? void 0 : _a.focus();
            prevSearchField.current = uiState.searchValue;
        }
        else {
            const firstFocusable = (_b = menuRef.current) === null || _b === void 0 ? void 0 : _b.querySelector(focusableSelector_1.FOCUSABLE_SELECTOR);
            prevSearchField.current = "";
            if (firstFocusable && !((_c = inputRef.current) === null || _c === void 0 ? void 0 : _c.value)) {
                firstFocusable.focus();
            }
        }
    }, [uiState.searchValue, uiState.isOpen, uiState.step, handleSetStep, chip]);
    (0, react_1.useEffect)(() => {
        var _a;
        uiState.isOpen && ((_a = inputRef.current) === null || _a === void 0 ? void 0 : _a.focus());
    }, [inputRef, uiState.isOpen]);
    (0, react_1.useEffect)(() => {
        if (!uiState.isOpen && uiState.step === groupsFilterInterfaces_1.FilterMenuStep.Initial && (0, groupsFilterHelper_1.isFilterHasSelection)(state.frozenState)) {
            dispatchUiState({ type: uiStateActionType_1.UiStateActionType.SetStep, payload: groupsFilterInterfaces_1.FilterMenuStep.CurrentlySelected });
        }
    }, [uiState.isOpen, state.frozenState, uiState.step]);
    function handleEscape(e) {
        var _a;
        if (e && uiState.isOpen) {
            (_a = e.nativeEvent) === null || _a === void 0 ? void 0 : _a.stopImmediatePropagation();
        }
        if (uiState.isOpenAdvancedFilter) {
            handleAdvancedDialogClose();
            return;
        }
        dispatchUiState({ type: uiStateActionType_1.UiStateActionType.SetIsOpen, payload: false });
        if (triggerRef.current) {
            triggerRef.current.focus();
        }
    }
    (0, useEscape_1.useEscape)(triggerRef, handleEscape, uiState.isOpen);
    const selectedCount = (0, stateReducerHelper_1.getFrozenCount)(state);
    const isStateHasData = Object.keys(state.groupsMap).length > 0;
    return (0, jsx_runtime_1.jsxs)(jsx_runtime_1.Fragment, { children: [chip ? (0, jsx_runtime_1.jsx)(chip_1.Chip, { status: selectedCount ? chipStatus || "active" : undefined, isOpen: uiState.isOpen, disabled: false, onClick: handleTriggerClick, onClose: selectedCount ? handleChipClose : undefined, triggerRef: triggerRef, icon: chipIcon, id: chipId, children: groupPlaceHolder !== translate("Groups") ? groupPlaceHolder : chipName })
                : (0, jsx_runtime_1.jsx)(groupsFilterTrigger_1.GroupsFilterTrigger, { value: uiState.searchValue, onChange: onSearchChange, isOpenPopup: uiState.isOpen, count: selectedCount, withGlobeIcon: state.frozenSideWideState && selectedCount > 0, inputRef: inputRef, handleClick: handleTriggerClick, ref: triggerRef, fullSize: fullSize, placeholder: groupPlaceHolder }), (0, jsx_runtime_1.jsx)(controlledPopup_1.ControlledPopup, { isOpen: uiState.isOpen, className: (0, classNames_1.classNames)(["zen-groups-filter__popup", uiState.isOpenAdvancedFilter ? "zen-groups-filter__popup--inactive" : "", className ? className : ""]), onOpenChange: onOptionsToggle, useTrapFocusWithTrigger: uiState.isOpenAdvancedFilter ? "off" : "withTrigger", shouldHoldScroll: true, triggerRef: triggerRef, ariaLabel: translate("Organization groups filter"), recalculateOnScroll: true, children: (0, jsx_runtime_1.jsxs)("div", { className: "zen-groups-filter-menu", ref: menuRef, children: [chip && (0, jsx_runtime_1.jsx)("div", { className: "zen-groups-filter__input-container", children: (0, jsx_runtime_1.jsx)(searchInput_1.SearchInput, { className: "zen-groups-filter__chip-input", onChange: onSearchChipChange, value: uiState.searchValue, ref: inputRef, id: groupsFilterId, placeholder: translate("Filter by group") }) }), isStateHasData && !uiState.isLoadWithError && (0, jsx_runtime_1.jsxs)(groupsFilterMenu_1.GroupsFilterMenu, { onCancelClick: handleCancel, onApplyClick: handleApply, onResetClick: handleReset, isApplyDisabled: !(0, stateReducerHelper_1.isStateChanged)(state), children: [uiState.step === groupsFilterInterfaces_1.FilterMenuStep.Initial && (0, jsx_runtime_1.jsx)(groupsFilterInitialState_1.GroupsFilterInitialState, { siteWideState: state.sideWideState, onDialogButtonClick: handleAdvancedDialogOpen, count: (0, stateReducerHelper_1.getCurrentlySelectedCount)(state), onSiteWideChange: onSideWideChange, onCurrentlySelectedClick: handleCurrentlySelectedClick, searchResult: uiState.searchValue ? getSearchResult() : "", children: getGroupsBox() }), uiState.step === groupsFilterInterfaces_1.FilterMenuStep.Adjustment && (0, jsx_runtime_1.jsx)(groupsFilterAdjustmentState_1.GroupsFilterAdjustmentState, { onBackButtonClick: handleBackButtonClick, onHomeButtonClick: handleHomeButtonClick, onSelectAllClick: selectAllHandler, backButtonName: (0, groupsFilterStatesHelper_1.getBackButtonName)(state, uiState.currentAdjustmentId), isAllSelected: (0, groupsFilterStatesHelper_1.isAllChildrenSelected)(state, uiState.currentAdjustmentId), children: getGroupsBox() }), uiState.step === groupsFilterInterfaces_1.FilterMenuStep.CurrentlySelected && (0, jsx_runtime_1.jsx)(groupsFilterCurrentlySelectedState_1.GroupsFilterCurrentlySelectedState, { siteWideState: state.sideWideState, relation: state.relation, count: (0, stateReducerHelper_1.getCurrentlySelectedCount)(state), onBackButtonClick: basicHandleBackButtonClick, onSiteWideChange: onSideWideChange, onRelationChange: onRelationChange, children: getGroupsBox() })] }), !uiState.isLoadWithError && !isStateHasData && !Object.keys(state.groupsMap).length && (0, jsx_runtime_1.jsx)(skeletonList_1.SkeletonList, { className: "zen-groups-filter__waiting" })] }) }), uiState.isOpenAdvancedFilter && getAdvancedDialog] });
};
exports.GroupsFilter = GroupsFilter;
exports.TRANSLATIONS = [
    "Organization groups filter",
    ...groupsFilterMenu_1.TRANSLATIONS,
    ...groupsFilterTrigger_1.TRANSLATIONS,
    ...groupsFilterAdjustmentState_1.TRANSLATIONS,
    ...groupsFilterInitialState_1.TRANSLATIONS,
    ...groupsFilterCurrentlySelectedState_1.TRANSLATIONS
];
