// @geotab/zenith/dist/groupsFilter/groupsFilterBox.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TRANSLATIONS = exports.GroupsFilterBox = exports.ADVANCED_FILTER_ID = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const groupsFilterInterfaces_1 = require("./groupsFilterInterfaces");
const groupsFilterStatesHelper_1 = require("./groupsFilterStatesHelper");
const checkboxListWithAction_1 = require("../checkboxListWithAction/checkboxListWithAction");
const useLanguage_1 = require("../utils/localization/useLanguage");
exports.ADVANCED_FILTER_ID = "advancedGroups";
const GroupsFilterBox = ({ blocks, step, state, currentId, onChange, onClick, search, isWithColor, uncheckedItems, uncheckedAdvancedState, listLimit }) => {
    const { translate } = (0, useLanguage_1.useLanguage)();
    const labelMap = {
        builtInGroups: "Built-in Groups",
        userGroups: "User Groups",
        advancedGroups: "Advanced Filter"
    };
    const handleAdvancedChange = (value) => onChange(Object.assign(Object.assign({}, value), { id: exports.ADVANCED_FILTER_ID }));
    const getGroupsMenuValue = (limit, value) => {
        if (step === groupsFilterInterfaces_1.FilterMenuStep.Adjustment && value) {
            return (0, groupsFilterStatesHelper_1.createAdjustmentOptions)(state, isWithColor, limit, currentId, value);
        }
        if (step === groupsFilterInterfaces_1.FilterMenuStep.CurrentlySelected && value) {
            const result = (0, groupsFilterStatesHelper_1.createCurrentlySelectedOptions)(state, value, isWithColor, uncheckedItems, limit, uncheckedAdvancedState);
            return result;
        }
        return (0, groupsFilterStatesHelper_1.createInitialOptions)(state, limit, search, search ? undefined : value);
    };
    const createBlocks = (limit) => {
        const groupsBlocks = blocks.map((val, ind) => {
            const blockOptions = getGroupsMenuValue(limit, val);
            return blockOptions && blockOptions.length ? (0, jsx_runtime_1.jsx)(checkboxListWithAction_1.CheckboxListWithAction, { className: step === groupsFilterInterfaces_1.FilterMenuStep.Adjustment && ind !== blocks.length - 1 ? "zen-groups-filter-menu__groups-list--no-space" : "", options: blockOptions, label: step === groupsFilterInterfaces_1.FilterMenuStep.Adjustment ? "" : translate(labelMap[val]) || "", search: search ? [search] : undefined, onChange: val === "advancedGroups" ? handleAdvancedChange : onChange, onClick: onClick }, `step-${step}_block-${val}`) : null;
        });
        return groupsBlocks.filter(el => el);
    };
    return blocks.length ? (0, jsx_runtime_1.jsx)(jsx_runtime_1.Fragment, { children: createBlocks(listLimit) }) : (0, jsx_runtime_1.jsx)(checkboxListWithAction_1.CheckboxListWithAction, { options: getGroupsMenuValue(listLimit), label: "", search: search ? [search] : undefined, onChange: onChange, onClick: onClick });
};
exports.GroupsFilterBox = GroupsFilterBox;
exports.TRANSLATIONS = [
    "Built-in Groups",
    "User Groups",
    "Advanced Filter"
];
