// @geotab/zenith/dist/toastRaw/toastRaw.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ToastRaw = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const toastAnimation_1 = require("./components/toastAnimation");
const iconCheckRadio_1 = require("../icons/iconCheckRadio");
const iconInfoCircle_1 = require("../icons/iconInfoCircle");
const react_1 = require("react");
const useDrive_1 = require("../utils/theme/useDrive");
const button_1 = require("../button/button");
const iconClose_1 = require("../icons/iconClose");
const useLanguage_1 = require("../utils/localization/useLanguage");
const classNames_1 = require("../commonHelpers/classNames/classNames");
const useMobile_1 = require("../commonHelpers/hooks/useMobile");
const useProgress_1 = require("./hooks/useProgress");
const themeContext_1 = require("../utils/theme/themeContext");
const textIconButton_1 = require("../textIconButton/textIconButton");
const typedIcons = {
    success: iconCheckRadio_1.IconCheckRadio,
    info: iconInfoCircle_1.IconInfoCircle
};
const getIcon = (t) => typedIcons[t];
exports.ToastRaw = (0, react_1.forwardRef)(
// eslint-disable-next-line @typescript-eslint/naming-convention
function ToastRawInner({ isOpen, onClose, animation, type, hasIcon = true, header, id, action, showSeconds = useProgress_1.MIN_SECONDS_TO_SHOW }, ref) {
    const isDrive = (0, useDrive_1.useDrive)();
    const isMobile = (0, useMobile_1.useMobile)();
    const { dark } = (0, react_1.useContext)(themeContext_1.themeContext);
    const { translate } = (0, useLanguage_1.useLanguage)();
    const { progress } = (0, useProgress_1.useProgress)({ isOpen, onClose, showSeconds });
    const toastClasses = (0, classNames_1.classNames)(["zen-toast", `zen-toast--${type || "info"}`, isMobile ? "zen-toast--mobile" : "", isDrive ? "zen-toast--drive" : "", dark ? "zen-dark" : ""]);
    const contentStartClasses = (0, react_1.useMemo)(() => (0, classNames_1.classNames)(["zen-toast__content-start", hasIcon ? "zen-toast__content-start--icon" : ""]), [hasIcon]);
    const contentEndClasses = (0, classNames_1.classNames)(["zen-toast__content-end", action ? "zen-toast__content-end--action" : ""]);
    const onActionClick = (0, react_1.useCallback)(() => {
        var _a;
        (_a = action === null || action === void 0 ? void 0 : action.onClick) === null || _a === void 0 ? void 0 : _a.call(action);
        onClose();
    }, [action, onClose]);
    const actionButton = (0, react_1.useMemo)(() => {
        if (!action) {
            return null;
        }
        if (action.icon) {
            return (0, jsx_runtime_1.jsx)(textIconButton_1.TextIconButton, { className: "zen-toast__action", type: "tertiary", icon: action.icon, iconPosition: action.iconPosition || textIconButton_1.ButtonIconPosition.Start, onClick: onActionClick, link: action.link, title: action.title || action.text, children: action.text });
        }
        return (0, jsx_runtime_1.jsx)(button_1.Button, { className: "zen-toast__action", type: "tertiary", link: action.link, onClick: onActionClick, title: action.title || action.text, children: action.text });
    }, [action, onActionClick]);
    return (0, jsx_runtime_1.jsx)(toastAnimation_1.ToastAnimation, { id: id, animation: animation, isOpen: isOpen, children: (0, jsx_runtime_1.jsxs)("div", { id: id, ref: ref, className: toastClasses, role: "status", "aria-live": "polite", "aria-atomic": "true", children: [(0, jsx_runtime_1.jsx)("div", { style: { width: `${progress}%` }, className: (0, classNames_1.classNames)(["zen-toast__progress", "progress", `zen-toast__progress--${type || "info"}`, progress !== 100 ? "zen-toast__progress--active" : ""]) }), (0, jsx_runtime_1.jsxs)("div", { className: "zen-toast__content", children: [(0, jsx_runtime_1.jsxs)("div", { className: contentStartClasses, children: [hasIcon && (0, react_1.createElement)(getIcon(type || "info"), {
                                    size: isDrive ? "huge" : "large",
                                    className: `zen-toast__icon--${type || "info"}`
                                }), (0, jsx_runtime_1.jsx)("div", { className: (0, classNames_1.classNames)(["zen-toast__header", `zen-toast__header--${type || "info"}`]), children: header })] }), (0, jsx_runtime_1.jsxs)("div", { className: contentEndClasses, children: [actionButton, (0, jsx_runtime_1.jsx)("button", { type: "button", className: (0, classNames_1.classNames)(["zen-toast__close-button", `zen-toast__close-button--${type || "info"}`]), onClick: onClose, title: translate("Close"), "aria-label": translate("Close"), children: (0, jsx_runtime_1.jsx)(iconClose_1.IconClose, { size: isDrive ? "huge" : "large" }) })] })] })] }) });
});
