// @geotab/zenith/dist/table/flexible/columnSettingsStorage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ColumnSettingsStorage = void 0;
const STORAGE_PREFIX = "columnsSettings_";
const createFakeSettings = () => {
    const settings = new Map();
    return {
        getItem: (key) => settings.get(key) || null,
        setItem: (key, s) => settings.set(key, s),
        removeItem: (key) => settings.delete(key)
    };
};
class ColumnSettingsStorage {
    constructor(pageName, storage, defaultSettings) {
        Object.defineProperty(this, "defaultSettings", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: defaultSettings
        });
        Object.defineProperty(this, "storeKey", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "storage", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.storage = storage || createFakeSettings();
        this.storeKey = `${STORAGE_PREFIX}${pageName}`;
        defaultSettings && this.setItem(defaultSettings);
    }
    getItem() {
        const settings = this.storage.getItem(this.storeKey);
        try {
            return (settings ? JSON.parse(settings) : undefined);
        }
        catch (e) {
            this.setItem(this.defaultSettings);
            return this.defaultSettings;
        }
    }
    setItem(s) {
        this.storage.setItem(this.storeKey, JSON.stringify(s));
    }
    clear() {
        this.storage.removeItem(this.storeKey);
    }
}
exports.ColumnSettingsStorage = ColumnSettingsStorage;
