// @geotab/zenith/dist/layout/layoutFullScreenProvider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LayoutFullScreenProvider = exports.useLayoutFullScreen = exports.layoutFullScreenContext = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
exports.layoutFullScreenContext = (0, react_1.createContext)({
    isFullScreen: () => false,
    setFullScreenElement: () => { },
    fullScreenElement: undefined
});
const useLayoutFullScreen = () => {
    const { isFullScreen, setFullScreenElement, fullScreenElement } = (0, react_1.useContext)(exports.layoutFullScreenContext);
    return { isFullScreen, setFullScreenElement, fullScreenElement };
};
exports.useLayoutFullScreen = useLayoutFullScreen;
const LayoutFullScreenProvider = ({ isFullScreen = () => false, setFullScreenElement = () => { }, fullScreenElement = undefined, children }) => (0, jsx_runtime_1.jsx)(exports.layoutFullScreenContext.Provider, { value: { isFullScreen: isFullScreen, setFullScreenElement: setFullScreenElement, fullScreenElement: fullScreenElement }, children: children });
exports.LayoutFullScreenProvider = LayoutFullScreenProvider;
