// @geotab/zenith/dist/miniTabs/miniTabs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MiniTabs = exports.MiniTabsSize = void 0;
const react_1 = require("react");
const jsx_runtime_1 = require("react/jsx-runtime");
const react_2 = require("react");
const classNames_1 = require("../commonHelpers/classNames/classNames");
const miniTabsHeaderItem_1 = require("./miniTabsHeaderItem");
const miniTabsItem_1 = require("./miniTabsItem");
const tabsKeyboard_1 = require("../commonHelpers/tabsKeyboard");
const miniTabsHelper_1 = require("./miniTabsHelper");
var MiniTabsSize;
(function (MiniTabsSize) {
    MiniTabsSize["Small"] = "small";
    MiniTabsSize["Medium"] = "medium";
    MiniTabsSize["Large"] = "large";
    MiniTabsSize["XLarge"] = "xlarge";
})(MiniTabsSize || (exports.MiniTabsSize = MiniTabsSize = {}));
const MiniTabs = ({ activeTabId, onTabChange, children, size = MiniTabsSize.Medium, className }) => {
    const maxTabsValue = 3;
    const [tabs, setTabs] = (0, react_2.useState)([]);
    const activeButtonRef = (0, react_2.useRef)(null);
    const tabsRef = (0, react_2.useRef)(null);
    const handleTabItemClick = (e) => {
        const newState = e.target.id || e.currentTarget.id || undefined;
        newState && newState !== activeTabId && onTabChange(newState);
    };
    const keyDownHandler = (0, react_2.useCallback)((e) => {
        const newActiveTabId = (0, tabsKeyboard_1.tabsKeyboardHandler)(e, tabs, activeTabId);
        if (newActiveTabId && newActiveTabId !== activeTabId) {
            onTabChange(newActiveTabId);
        }
    }, [tabs, activeTabId, onTabChange]);
    (0, react_2.useEffect)(() => setTabs((0, miniTabsHelper_1.getChildrenForRender)(children).slice(0, maxTabsValue)), [children]);
    (0, react_2.useEffect)(() => {
        var _a;
        (_a = activeButtonRef.current) === null || _a === void 0 ? void 0 : _a.focus();
    }, [activeTabId]);
    return (0, jsx_runtime_1.jsx)("div", { className: (0, classNames_1.classNames)(["zen-mini-tabs", `zen-mini-tabs--${size}`, className ? className : ""]), children: tabs.length ? (0, jsx_runtime_1.jsxs)(jsx_runtime_1.Fragment, { children: [(0, jsx_runtime_1.jsx)("div", { className: "zen-mini-tabs__header", role: "tablist", onKeyDown: keyDownHandler, ref: tabsRef, children: tabs.map(tab => ((0, jsx_runtime_1.jsx)(miniTabsHeaderItem_1.MiniTabsHeaderItem, { id: tab.id, isActive: activeTabId === tab.id, handleClick: handleTabItemClick, buttonRef: tab.id === activeTabId ? activeButtonRef : undefined, children: tab.name }, tab.id))) }), (0, jsx_runtime_1.jsx)("div", { className: "zen-mini-tabs__content", children: tabs.map((props) => ((0, react_1.createElement)(miniTabsItem_1.MiniTabsItem, Object.assign({}, props, { key: props.id, activeId: activeTabId })))) })] }) : null });
};
exports.MiniTabs = MiniTabs;
