// @geotab/zenith/dist/columnsSelector/columnsTab/columnsTab.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TRANSLATIONS = exports.ColumnsTab = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const selectList_1 = require("../../selectList/selectList");
const checkboxState_1 = require("../../checkbox/checkboxState");
const useLanguage_1 = require("../../utils/localization/useLanguage");
const ColumnsTab = ({ columns, onColumnToggle, id }) => {
    const { translate } = (0, useLanguage_1.useLanguage)();
    const listItems = columns.map(function (col) {
        return {
            id: col.id,
            state: col.isChecked ? checkboxState_1.CheckboxState.On : checkboxState_1.CheckboxState.Off,
            disabled: col.isBlocked,
            children: col.label
        };
    });
    return (0, jsx_runtime_1.jsx)(selectList_1.SelectList, { title: translate("Columns selector"), multiselect: true, showCheckbox: true, id: `columns-list-${id}`, value: new Map(columns.map(col => [col.id, col.isChecked ? checkboxState_1.CheckboxState.On : checkboxState_1.CheckboxState.Off])), onChange: state => onColumnToggle(columns.map(col => (Object.assign(Object.assign({}, col), { isChecked: state.get(col.id) === checkboxState_1.CheckboxState.On })))), items: listItems });
};
exports.ColumnsTab = ColumnsTab;
exports.TRANSLATIONS = [
    "Columns selector"
];
