// @geotab/zenith/dist/groupsFilter/uiStateReducer/uiStateReducer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.uiStateReducer = exports.uiInitialGroupsFilterState = void 0;
const groupsFilterInterfaces_1 = require("../groupsFilterInterfaces");
const uiStateActionType_1 = require("./uiStateActionType");
exports.uiInitialGroupsFilterState = {
    isOpen: false,
    isLoadWithError: false,
    step: groupsFilterInterfaces_1.FilterMenuStep.Initial,
    currentAdjustmentId: undefined,
    uncheckedAdjustmentIds: [],
    uncheckedAdvancedState: undefined,
    searchValue: "",
    isOpenAdvancedFilter: false
};
// eslint-disable-next-line complexity
function uiStateReducer(state, action) {
    const { type, payload } = action;
    switch (type) {
        case uiStateActionType_1.UiStateActionType.SetIsOpen:
            return Object.assign(Object.assign({}, state), { isOpen: payload });
        case uiStateActionType_1.UiStateActionType.SetAdvancedFilterIsOpen:
            return Object.assign(Object.assign({}, state), { isOpenAdvancedFilter: payload });
        case uiStateActionType_1.UiStateActionType.ToggleOpen:
            return state.isOpenAdvancedFilter ? state : Object.assign(Object.assign({}, state), { isOpen: !state.isOpen });
        case uiStateActionType_1.UiStateActionType.SetLoadWithError:
            return Object.assign(Object.assign({}, state), { isLoadWithError: payload });
        case uiStateActionType_1.UiStateActionType.SetStep:
            return Object.assign(Object.assign({}, state), { step: payload });
        case uiStateActionType_1.UiStateActionType.SetCurrentAdjustmentId:
            return Object.assign(Object.assign({}, state), { currentAdjustmentId: payload });
        case uiStateActionType_1.UiStateActionType.SetUncheckedAdjustmentIds:
            return Object.assign(Object.assign({}, state), { uncheckedAdjustmentIds: [...payload] });
        case uiStateActionType_1.UiStateActionType.SetUncheckedAdvancedState:
            return Object.assign(Object.assign({}, state), { uncheckedAdvancedState: payload });
        case uiStateActionType_1.UiStateActionType.ClearUncheckedState:
            return Object.assign(Object.assign({}, state), { uncheckedAdvancedState: undefined, uncheckedAdjustmentIds: [] });
        case uiStateActionType_1.UiStateActionType.SetSearchValue:
            return Object.assign(Object.assign({}, state), { searchValue: payload });
        default: {
            throw Error("Unknown action: " + type);
        }
    }
}
exports.uiStateReducer = uiStateReducer;
