// @geotab/zenith/dist/table/children/useTableChildren.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useTableChildren = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const deviceType_1 = require("../../commonHelpers/hooks/deviceType");
const bulkActions_1 = require("../../dataGrid/withSelectableRows/components/bulkActions/bulkActions");
const tableBulkActions_1 = require("./tableBulkActions");
const tablePagination_1 = require("./tablePagination");
const pagination_1 = require("../../pagination/pagination");
const paginationType_1 = require("../../pagination/paginationType");
const tableFullscreen_1 = require("./tableFullscreen");
const fullscreenButton_1 = require("../../dataGrid/entitiesListActions/actions/fullscreenButton");
const tableEmpty_1 = require("./tableEmpty");
const emptySearchList_1 = require("../../dataGrid/emptySearchList/emptySearchList");
const dataGrid_1 = require("../../dataGrid/dataGrid");
const tableFooter_1 = require("./tableFooter");
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const isChildElement = (child, expectedType) => {
    if (!child.type) {
        return false;
    }
    return child.type.displayName === expectedType;
};
const useTableChildren = (children, isMobile, selected, allSelected, clearSelection, selectAll, turnOffSelectAll, gridRef) => (0, react_1.useMemo)(() => {
    let bulkActions = undefined;
    let tablePagination = undefined;
    let activePage;
    let tableFullscreen = undefined;
    let emptyList = undefined;
    let footer = undefined;
    const other = [];
    react_1.Children.forEach(children, child => {
        var _a;
        if (!child) {
            return;
        }
        if (isChildElement(child, tableBulkActions_1.TableBulkActions.displayName)) {
            const bulkActionButtons = ((_a = child.props) === null || _a === void 0 ? void 0 : _a.children) || [];
            bulkActions = (0, jsx_runtime_1.jsx)(bulkActions_1.BulkActions, { gridType: isMobile ? deviceType_1.DeviceType.Mobile : deviceType_1.DeviceType.Desktop, selected: selected, allSelected: allSelected, onClearSelection: clearSelection, onSelectAll: selectAll, turnOffSelectAll: turnOffSelectAll, children: bulkActionButtons });
            return;
        }
        if (isChildElement(child, tablePagination_1.TablePagination.displayName)) {
            tablePagination = (0, react_1.createElement)(pagination_1.Pagination, {
                activePage: child.props.activePage,
                total: child.props.total,
                perPage: child.props.perPage,
                onChange: child.props.onChange,
                type: paginationType_1.PaginationType.Compact,
                includeText: true,
                gridRef: gridRef,
                disabled: child.props.disabled
            });
            activePage = child.props.activePage;
            return;
        }
        if (isChildElement(child, tableFullscreen_1.TableFullscreen.displayName)) {
            tableFullscreen = (0, react_1.createElement)(fullscreenButton_1.FullScreenButton, {
                fullScreen: child.props.isFullscreen,
                onClick: child.props.onChange
            });
            return;
        }
        if (isChildElement(child, tableEmpty_1.TableEmpty.displayName)) {
            // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
            emptyList = (0, react_1.createElement)(dataGrid_1.EmptyList, {}, (0, react_1.createElement)(emptySearchList_1.EmptySearchList, {
                description: child.props.description,
                image: child.props.image,
                className: child.props.className
            }));
            return;
        }
        if (isChildElement(child, tableFooter_1.TableFooter.displayName)) {
            footer = (0, react_1.createElement)(dataGrid_1.Footer, {}, child.props.children);
            return;
        }
        other.push(child);
    });
    return {
        bulkActions: bulkActions,
        pagination: tablePagination,
        activePage: activePage,
        fullscreenButton: tableFullscreen,
        emptyList: emptyList,
        footer: footer,
        other: other
    };
}, [children, clearSelection, isMobile, selected, gridRef, allSelected, selectAll, turnOffSelectAll]);
exports.useTableChildren = useTableChildren;
