// @geotab/zenith/dist/chart/barChart/getDefaultOptions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getDefaultOptions = void 0;
const formatDate_1 = require("../../utils/formatDate");
const getDefaultGridStyle = (isDark) => ({
    drawTicks: false,
    color: isDark ? "#C0CCD8" : "#4E677E"
});
const defaultGridBorderStyle = {
    display: false,
    dash: [2, 2]
};
const getDefaultOptions = (yAxis, isTimedData, isDark, translate) => {
    const textColor = isDark ? "#C0CCD8" : "#4E677E";
    const scales = {
        x: {
            title: {
                display: false
            },
            grid: {
                display: false
            },
            type: isTimedData ? "time" : "category",
            ticks: isTimedData ? {
                source: "data",
                autoSkip: true,
                maxRotation: 0,
                maxTicksLimit: 12,
                color: textColor,
                callback: (value) => (0, formatDate_1.formatDate)(new Date(value), "MMM dd", translate)
            } : {
                source: "data",
                autoSkip: false,
                //maxRotation: 0,
                maxTicksLimit: 12,
                color: textColor
            },
            border: defaultGridBorderStyle
        }
    };
    for (let i = 0; i < yAxis.length; i++) {
        const yAxisId = yAxis[i].name;
        scales[yAxisId] = {
            title: {
                display: false
            },
            ticks: {
                color: textColor,
                padding: 10,
                source: "data",
                autoSkip: true
            },
            beginAtZero: true,
            grid: i === 0 ? getDefaultGridStyle(isDark) : { display: false },
            border: defaultGridBorderStyle,
            min: yAxis[i].min,
            max: yAxis[i].max,
            position: i === 0 ? "left" : "right"
        };
    }
    return {
        responsive: true,
        maintainAspectRatio: false,
        scales,
        plugins: {
            legend: {
                display: false
            },
            tooltip: {
                enabled: false
            },
            zenithTooltip: {}
        },
        animation: false
    };
};
exports.getDefaultOptions = getDefaultOptions;
