// @geotab/zenith/dist/pillExpandable/pillExpandable.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TRANSLATIONS = exports.PillExpandable = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const absolute_1 = require("../absolute/absolute");
const button_1 = require("../button/button");
const pillContent_1 = require("./pillContent");
const pillExpandableSimple_1 = require("./pillExpandableSimple");
const pillExpandableAlignments_1 = require("./pillExpandableAlignments");
const useLanguage_1 = require("../utils/localization/useLanguage");
const classNames_1 = require("../commonHelpers/classNames/classNames");
const getProps_1 = require("./getProps");
const skeleton_1 = require("../skeleton/skeleton");
const reactHelpers_1 = require("../commonHelpers/reactHelpers/reactHelpers");
const MAX_PILLS_TO_DISPLAY_IN_POPUP = 2;
const PillContext = (0, react_1.createContext)({
    level: 0,
    onPillExpand: () => { },
    expandId: undefined
});
const paddingX = 26;
const paddingY = 52;
const getDefaultContent = ({ description, date, count, secondaryAction, mainAction, getData }) => {
    if (getData) {
        return {
            description: {
                isLoading: true,
                data: ""
            },
            date: {
                isLoading: true,
                data: ""
            },
            count: {
                isLoading: true,
                data: 0
            },
            secondaryAction: {
                isLoading: true,
                data: {
                    text: ""
                }
            },
            mainAction: {
                isLoading: true,
                data: {
                    text: ""
                }
            },
            getData
        };
    }
    return {
        description,
        date,
        count,
        secondaryAction,
        mainAction,
        getData: undefined
    };
};
const getValue = (parameter, setContent, paramName, errorHandler) => {
    if (parameter instanceof Function) {
        parameter()
            .then((result) => {
            setContent((prevValue) => (Object.assign(Object.assign({}, prevValue), { [paramName]: result })));
        })
            .catch((e) => {
            errorHandler && errorHandler(e);
        });
    }
};
const PillExpandableBase = ({ children, errorHandler, getData, text, description, date, count, mainAction, secondaryAction, viewMoreAction, uniquePills, type, className, popupClassName, isLoading, isFlat, icon, secondaryIcon, isBeta, loadingWidth = 200, hideCounterNumber }) => {
    const popupId = (0, react_1.useId)();
    const pillId = (0, react_1.useId)();
    const { translate } = (0, useLanguage_1.useLanguage)();
    const [content, setContent] = (0, react_1.useState)(getDefaultContent({
        description,
        date,
        count,
        secondaryAction,
        mainAction,
        getData
    }));
    const [isOpen, setIsOpen] = (0, react_1.useState)(false);
    const [isPopupContentReady, setIsPopupContentReady] = (0, react_1.useState)(false);
    const [containerNode, setContainerNode] = (0, react_1.useState)(null);
    (0, react_1.useEffect)(() => {
        if (!isOpen) {
            setIsPopupContentReady(false);
            return;
        }
        if (!containerNode || typeof ResizeObserver === "undefined") {
            return;
        }
        const observer = new ResizeObserver(() => {
            if (containerNode.offsetHeight > 0) {
                setIsPopupContentReady(true);
            }
        });
        observer.observe(containerNode);
        // eslint-disable-next-line consistent-return
        return () => {
            observer.unobserve(containerNode);
        };
    }, [isOpen, containerNode]);
    (0, react_1.useEffect)(() => {
        const fetchData = function () {
            if (!isOpen) {
                return;
            }
            if ((0, reactHelpers_1.isFunction)(getData)) {
                getData().then(result => {
                    setContent(result);
                }).catch(e => {
                    errorHandler && errorHandler(e);
                });
            }
        };
        fetchData();
    }, [getData, isOpen, errorHandler]);
    (0, react_1.useEffect)(() => {
        if (!isOpen) {
            return;
        }
        getValue(description, setContent, "description", errorHandler);
        getValue(date, setContent, "date", errorHandler);
        getValue(count, setContent, "count", errorHandler);
        getValue(secondaryAction, setContent, "secondaryAction", errorHandler);
        getValue(mainAction, setContent, "mainAction", errorHandler);
    }, [content, isOpen, description, date, count, secondaryAction, mainAction, errorHandler]);
    const { level, expandId, onPillExpand } = (0, react_1.useContext)(PillContext);
    const triggerRef = (0, react_1.useRef)(null);
    const isChild = level !== 0;
    const [currentExpand, setCurrentExpand] = (0, react_1.useState)(undefined);
    const onExpandHandler = (0, react_1.useCallback)(() => {
        var _a;
        setIsOpen(!isOpen);
        isOpen && ((_a = triggerRef.current) === null || _a === void 0 ? void 0 : _a.focus());
        onPillExpand(pillId === expandId ? "" : pillId);
    }, [isOpen, onPillExpand, pillId, expandId]);
    const onChange = (0, react_1.useCallback)(() => {
        setIsOpen(!isOpen);
    }, [isOpen]);
    const buttonMoreProps = (0, react_1.useMemo)(() => (0, getProps_1.getButtonMoreProps)(viewMoreAction), [viewMoreAction]);
    const validChildren = (0, react_1.useMemo)(() => react_1.Children.toArray(children).filter(child => { var _a; return ((_a = child === null || child === void 0 ? void 0 : child.type) === null || _a === void 0 ? void 0 : _a.displayName) === "PillExpandable"; }), [children]);
    const childToDisplay = (0, react_1.useMemo)(() => validChildren.slice(0, MAX_PILLS_TO_DISPLAY_IN_POPUP), [validChildren]);
    const childCount = (0, react_1.useMemo)(() => validChildren.length, [validChildren]);
    const flatVsChildCounter = isFlat ? `${childCount + 1}` : `${childCount}`;
    const moreCounter = uniquePills ? `${uniquePills}` : flatVsChildCounter;
    const uniqueChilds = uniquePills ? `${uniquePills}` : `${childCount || ""}`;
    if (isLoading) {
        return (0, jsx_runtime_1.jsx)(skeleton_1.Skeleton, { type: "pill", width: loadingWidth });
    }
    if (isChild) {
        return (0, jsx_runtime_1.jsxs)(PillContext.Provider, { value: { level: level + 1, onPillExpand: setCurrentExpand, expandId: currentExpand }, children: [(0, jsx_runtime_1.jsx)(pillExpandableSimple_1.PillExpandableSimple, { icon: icon, text: text, expanded: expandId === pillId, onExpand: onExpandHandler, expandAriaText: translate("Expand"), type: type, errorHandler: errorHandler }), (0, jsx_runtime_1.jsx)(pillContent_1.PillContent, { onActionClick: onExpandHandler, descriptionText: content.description, date: content.date, count: content.count, mainAction: content.mainAction, secondaryAction: content.secondaryAction, isVisible: expandId === pillId, errorHandler: errorHandler })] });
    }
    return (0, jsx_runtime_1.jsx)(PillContext.Provider, { value: { level: level + 1, onPillExpand: setCurrentExpand, expandId: currentExpand }, children: (0, jsx_runtime_1.jsxs)("div", { className: (0, classNames_1.classNames)(["zen-status-pill", className || ""]), "aria-haspopup": "true", children: [(0, jsx_runtime_1.jsxs)("div", { className: "zen-status-pill__pill-container", children: [(0, jsx_runtime_1.jsx)(pillExpandableSimple_1.PillExpandableSimple, { icon: icon, text: text, expanded: isOpen, onExpand: onExpandHandler, expandAriaText: translate("Expand"), ref: triggerRef, type: type, errorHandler: errorHandler, isBeta: isBeta }), uniqueChilds && (0, jsx_runtime_1.jsx)(pillExpandableSimple_1.PillExpandableSimple, { icon: secondaryIcon || icon, text: hideCounterNumber ? "" : `${isFlat ? "+" : ""}${uniqueChilds}`, onExpand: onExpandHandler, includeExpandIcon: false, type: type, errorHandler: errorHandler })] }), (0, jsx_runtime_1.jsx)(absolute_1.Absolute, { recalculateTrigger: isPopupContentReady, recalculateOnScroll: true, useTrapFocusWithTrigger: "on", className: popupClassName, id: popupId, onOpenChange: onChange, paddingX: paddingX, paddingY: paddingY, isOpen: isOpen, triggerRef: triggerRef, alignmentsFn: pillExpandableAlignments_1.alignments, children: (0, jsx_runtime_1.jsx)("div", { ref: (node) => setContainerNode(node), className: "zen-status-pill-popup", children: (0, jsx_runtime_1.jsxs)("div", { className: "zen-status-pill-popup__container", children: [(0, jsx_runtime_1.jsx)("div", { className: "zen-status-pill-popup__main-pill", children: (0, jsx_runtime_1.jsx)(pillExpandableSimple_1.PillExpandableSimple, { icon: icon, text: text, expanded: true, onExpand: onExpandHandler, expandAriaText: translate("Expand"), type: type, errorHandler: errorHandler, isBeta: isBeta }) }), (0, jsx_runtime_1.jsx)(pillContent_1.PillContent, { onActionClick: onExpandHandler, descriptionText: content.description, date: content.date, count: content.count, mainAction: content.mainAction, secondaryAction: content.secondaryAction, errorHandler: errorHandler }), childToDisplay, moreCounter && viewMoreAction && childCount > MAX_PILLS_TO_DISPLAY_IN_POPUP && (0, jsx_runtime_1.jsxs)(jsx_runtime_1.Fragment, { children: [(0, jsx_runtime_1.jsx)("div", { className: "zen-status-pill-popup__divider" }), (0, jsx_runtime_1.jsxs)(button_1.Button, Object.assign({}, buttonMoreProps, { type: "tertiary", children: [translate("View all"), " (", moreCounter, ")"] }))] })] }) }) })] }) });
};
exports.PillExpandable = (0, react_1.memo)(PillExpandableBase);
exports.PillExpandable.displayName = "PillExpandable";
exports.TRANSLATIONS = [
    "Expand",
    "View all"
];
