// @geotab/zenith/dist/filtersBar/filtersBarSidePanel/sidePanelReducer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sidePanelReducer = exports.StateActionType = void 0;
var StateActionType;
(function (StateActionType) {
    StateActionType["Reset"] = "RESET";
    StateActionType["SetError"] = "SET_ERROR";
    StateActionType["Change"] = "CHANGE";
    StateActionType["Reinitialize"] = "REINITIALIZE";
})(StateActionType || (exports.StateActionType = StateActionType = {}));
function sidePanelReducer(state, action) {
    if (action.type === StateActionType.Reset) {
        const payloadHash = action.payload;
        const updatedState = Object.keys(state).reduce((res, id) => {
            const stateById = payloadHash[id];
            const currStateById = state[id];
            if (stateById && currStateById) {
                res[id] = Object.assign(Object.assign(Object.assign({}, currStateById), stateById), { hasError: false });
            }
            return res;
        }, {});
        return Object.assign(Object.assign({}, state.state), updatedState);
    }
    if (action.type === StateActionType.SetError) {
        const payloadHash = action.payload.id;
        const payloadValue = action.payload.value;
        const updatedState = state[payloadHash] ? Object.assign(Object.assign({}, state), { [payloadHash]: Object.assign(Object.assign({}, state[payloadHash]), { hasError: payloadValue }) }) : {};
        return Object.assign(Object.assign({}, state), updatedState);
    }
    if (action.type === StateActionType.Reinitialize) {
        return Object.assign(Object.assign({}, state), action.payload);
    }
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
    if (action.type === StateActionType.Change) {
        const id = action.payload.id;
        const currentState = state[id];
        if (!currentState) {
            return state;
        }
        const payloadState = action.payload.value;
        const newState = Object.assign(Object.assign({}, state), { [id]: Object.assign(Object.assign({}, currentState), payloadState) });
        return newState;
    }
    return state;
}
exports.sidePanelReducer = sidePanelReducer;
