// @geotab/zenith/dist/overview/overview.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Overview = exports.OverviewOptionsArrow = exports.OverviewOptionsType = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const pill_1 = require("../oldPill/pill");
const classNames_1 = require("../commonHelpers/classNames/classNames");
const iconArrowTop_1 = require("../icons/iconArrowTop");
const iconArrowBottom_1 = require("../icons/iconArrowBottom");
var OverviewOptionsType;
(function (OverviewOptionsType) {
    OverviewOptionsType["Positive"] = "positive";
    OverviewOptionsType["Negative"] = "negative";
    OverviewOptionsType["Neutral"] = "neutral";
})(OverviewOptionsType || (exports.OverviewOptionsType = OverviewOptionsType = {}));
var OverviewOptionsArrow;
(function (OverviewOptionsArrow) {
    OverviewOptionsArrow["Up"] = "up";
    OverviewOptionsArrow["Down"] = "down";
})(OverviewOptionsArrow || (exports.OverviewOptionsArrow = OverviewOptionsArrow = {}));
const Overview = ({ className, title, description, icon, label }) => {
    const getLabel = (0, react_1.useCallback)(() => {
        if (!(label === null || label === void 0 ? void 0 : label.type) || label.type === OverviewOptionsType.Neutral || !label.arrow) {
            return (0, jsx_runtime_1.jsxs)("div", { className: "zen-overview__label-item", children: [label === null || label === void 0 ? void 0 : label.percentage, "%"] });
        }
        return (0, jsx_runtime_1.jsx)(pill_1.Pill, { type: label.type === OverviewOptionsType.Positive ? pill_1.PillType.Success : pill_1.PillType.Error, size: pill_1.PillSize.Small, children: (0, jsx_runtime_1.jsxs)(jsx_runtime_1.Fragment, { children: [label.arrow === OverviewOptionsArrow.Up
                        ? (0, jsx_runtime_1.jsx)(iconArrowTop_1.IconArrowTop, {})
                        : (0, jsx_runtime_1.jsx)(iconArrowBottom_1.IconArrowBottom, {}), (0, jsx_runtime_1.jsxs)("span", { children: [label.percentage, "%"] })] }) });
    }, [label === null || label === void 0 ? void 0 : label.type, label === null || label === void 0 ? void 0 : label.percentage, label === null || label === void 0 ? void 0 : label.arrow]);
    return (0, jsx_runtime_1.jsxs)("div", { className: (0, classNames_1.classNames)(["zen-overview", className ? className : ""]), children: [icon ? (0, jsx_runtime_1.jsx)("div", { className: "zen-overview__icon", children: icon }) : null, (0, jsx_runtime_1.jsxs)("div", { className: (0, classNames_1.classNames)(["zen-overview__content", icon ? "" : "zen-overview__content--full"]), children: [(0, jsx_runtime_1.jsx)("div", { className: "zen-overview__title zen-ellipsis heading-05", children: title }), (0, jsx_runtime_1.jsxs)("div", { className: "zen-overview__additional", children: [(0, jsx_runtime_1.jsx)("div", { className: "zen-overview__description zen-ellipsis body-04", children: description }), label ? (0, jsx_runtime_1.jsx)("div", { className: "zen-overview__label", children: getLabel() }) : null] })] })] });
};
exports.Overview = Overview;
