// @geotab/zenith/dist/dropdown/stateReducer/stateActionType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StateActionType = void 0;
var StateActionType;
(function (StateActionType) {
    StateActionType["CreateMap"] = "CREATE_MAP";
    StateActionType["SetState"] = "SET_STATE";
    StateActionType["SetDefaultValue"] = "SET_DEFAULT_VALUE";
    StateActionType["SetGlobalState"] = "SET_GLOBAL_STATE";
    StateActionType["SetListData"] = "SET_LIST_DATA";
    StateActionType["ChangeSelection"] = "CHANGE_SELECTION";
    StateActionType["ChangeBulkSelection"] = "CHANGE_BULK_SELECTION";
    StateActionType["ChangeSingleSelection"] = "CHANGE_SINGLE_SELECTION";
    StateActionType["DeselectAllChildren"] = "DESELECT_ALL_CHILDREN";
    StateActionType["SelectAllChildren"] = "SELECT_ALL_CHILDREN";
    StateActionType["SetCurrentId"] = "SET_CURRENT_ID";
    StateActionType["ResetSelection"] = "RESET_SELECTION";
    StateActionType["ResetStateToGlobal"] = "RESET_STATE_TO_GLOBAL";
    StateActionType["Show"] = "SHOW";
    StateActionType["NothingToShow"] = "NOTHING_TO_SHOW";
    StateActionType["ShowWaiting"] = "SHOW_WAITING";
    StateActionType["ShowEmptyList"] = "SHOW_EMPTY_LIST";
    StateActionType["ShowList"] = "SHOW_LIST";
    StateActionType["SetInputValue"] = "SET_INPUT_VALUE";
    StateActionType["SetIsOpenCombo"] = "SET_IS_OPEN_COMBO";
    StateActionType["SetDataProblem"] = "SET_DATA_PROBLEM";
    StateActionType["UpdateSelectedItems"] = "UPDATE_SELECTED_ITEMS";
    StateActionType["SetPendingState"] = "SET_PENDING_STATE";
    StateActionType["UpdateNamelessItems"] = "UPDATE_NAMELESS_ITEMS";
    StateActionType["ToggleValueForAllSelected"] = "TOGGLE_VALUE_FOR_ALL_SELECTED";
    StateActionType["DeselectItemsFromAllSelected"] = "DESELECT_ITEMS_FROM_ALL_SELECTED";
    StateActionType["SetDefaultAllSelected"] = "SET_DEFAULT_ALL_SELECTED";
    StateActionType["SetDefaultIsChecked"] = "SET_DEFAULT_IS_CHECKED";
    StateActionType["SetIsChecked"] = "SET_IS_CHECKED";
})(StateActionType || (exports.StateActionType = StateActionType = {}));
