// @geotab/zenith/dist/dateRange/dateRangeUtils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GET_DEFAULT_DATE_OPTIONS = exports.GET_THIS_YEAR_OPTION = exports.GET_LAST_YEAR_OPTION = exports.GET_THIS_QUARTER_OPTION = exports.GET_LAST_QUARTER_OPTION = exports.GET_LAST_THREE_MONTH_OPTION = exports.GET_LAST_TWELVE_MONTHS_OPTION = exports.GET_LAST_THIRTY_DAYS_OPTION = exports.GET_LAST_SEVEN_DAYS_OPTION = exports.GET_LAST_MONTH_OPTION = exports.GET_NEXT_MONTH_OPTION = exports.GET_THIS_MONTH_OPTION = exports.GET_LAST_WEEK_OPTION = exports.GET_NEXT_WEEK_OPTION = exports.GET_THIS_WEEK_OPTION = exports.GET_YESTERDAY_OPTION = exports.GET_TODAY_OPTION = void 0;
const dateUtils_1 = require("../calendar/dateUtils");
const defaultToLocalDateTime = d => typeof d === "string" ? new Date(d) : d;
const defaultToReverseOffsetDateTime = d => typeof d === "string" ? new Date(d) : d;
const defaultTranslate = s => s;
const GET_TODAY_OPTION = (toLocalDateTime = defaultToLocalDateTime, translate = defaultTranslate, toReverseOffsetDateTime = defaultToReverseOffsetDateTime) => {
    const startDate = (0, dateUtils_1.getNewStartDate)(toLocalDateTime);
    const endDate = (0, dateUtils_1.getNewEndDate)(toLocalDateTime);
    const from = toReverseOffsetDateTime(startDate);
    const to = toReverseOffsetDateTime(endDate);
    return {
        label: translate("Today"),
        getRange: () => ({ from, to })
    };
};
exports.GET_TODAY_OPTION = GET_TODAY_OPTION;
const GET_YESTERDAY_OPTION = (toLocalDateTime = defaultToLocalDateTime, translate = defaultTranslate, toReverseOffsetDateTime = defaultToReverseOffsetDateTime) => {
    const startDate = (0, dateUtils_1.addDays)((0, dateUtils_1.getNewStartDate)(toLocalDateTime), -1);
    const endDate = (0, dateUtils_1.addDays)((0, dateUtils_1.getNewEndDate)(toLocalDateTime), -1);
    const from = toReverseOffsetDateTime(startDate);
    const to = toReverseOffsetDateTime(endDate);
    return {
        label: translate("Yesterday"),
        getRange: () => ({ from, to })
    };
};
exports.GET_YESTERDAY_OPTION = GET_YESTERDAY_OPTION;
const GET_THIS_WEEK_OPTION = (toLocalDateTime = defaultToLocalDateTime, translate = defaultTranslate, firstDayOfWeek, toReverseOffsetDateTime = defaultToReverseOffsetDateTime) => {
    const offsetDays = (0, dateUtils_1.getDayOfWeekOffset)(new Date(), firstDayOfWeek);
    const startDate = (0, dateUtils_1.addDays)((0, dateUtils_1.getNewStartDate)(toLocalDateTime), -offsetDays);
    const endDate = (0, dateUtils_1.addDays)((0, dateUtils_1.getNewEndDate)(toLocalDateTime), 6 - offsetDays);
    const from = toReverseOffsetDateTime(startDate);
    const to = toReverseOffsetDateTime(endDate);
    return {
        label: translate("This week"),
        getRange: () => ({
            from,
            to
        })
    };
};
exports.GET_THIS_WEEK_OPTION = GET_THIS_WEEK_OPTION;
const GET_NEXT_WEEK_OPTION = (toLocalDateTime = defaultToLocalDateTime, translate = defaultTranslate, firstDayOfWeek, toReverseOffsetDateTime = defaultToReverseOffsetDateTime) => {
    const offsetDays = (0, dateUtils_1.getDayOfWeekOffset)(new Date(), firstDayOfWeek);
    const startDate = (0, dateUtils_1.addDays)((0, dateUtils_1.getNewStartDate)(toLocalDateTime), -offsetDays + 7);
    const endDate = (0, dateUtils_1.addDays)((0, dateUtils_1.getNewEndDate)(toLocalDateTime), -offsetDays + 13);
    const from = toReverseOffsetDateTime(startDate);
    const to = toReverseOffsetDateTime(endDate);
    return {
        label: translate("Next week"),
        getRange: () => ({
            from,
            to
        })
    };
};
exports.GET_NEXT_WEEK_OPTION = GET_NEXT_WEEK_OPTION;
const GET_LAST_WEEK_OPTION = (toLocalDateTime = defaultToLocalDateTime, translate = defaultTranslate, firstDayOfWeek, toReverseOffsetDateTime = defaultToReverseOffsetDateTime) => {
    const offsetDays = (0, dateUtils_1.getDayOfWeekOffset)(new Date(), firstDayOfWeek);
    const startDate = (0, dateUtils_1.addDays)((0, dateUtils_1.getNewStartDate)(toLocalDateTime), -offsetDays - 7);
    const endDate = (0, dateUtils_1.addDays)((0, dateUtils_1.getNewEndDate)(toLocalDateTime), 6 - offsetDays - 7);
    const from = toReverseOffsetDateTime(startDate);
    const to = toReverseOffsetDateTime(endDate);
    return {
        label: translate("Last week"),
        getRange: () => ({
            from,
            to
        })
    };
};
exports.GET_LAST_WEEK_OPTION = GET_LAST_WEEK_OPTION;
const GET_THIS_MONTH_OPTION = (toLocalDateTime = defaultToLocalDateTime, translate = defaultTranslate, toReverseOffsetDateTime = defaultToReverseOffsetDateTime) => {
    const startDate = (0, dateUtils_1.getNewStartDate)(toLocalDateTime);
    startDate.setDate(1);
    const endDate = (0, dateUtils_1.getNewEndDate)(toLocalDateTime);
    (0, dateUtils_1.addMonths)(endDate, 1);
    endDate.setDate(1);
    endDate.setHours(0);
    endDate.setMinutes(0);
    const from = toReverseOffsetDateTime(startDate);
    const to = toReverseOffsetDateTime((0, dateUtils_1.addMinutes)(endDate, -1));
    return {
        label: translate("This month"),
        getRange: () => ({
            from,
            to
        })
    };
};
exports.GET_THIS_MONTH_OPTION = GET_THIS_MONTH_OPTION;
const GET_NEXT_MONTH_OPTION = (toLocalDateTime = defaultToLocalDateTime, translate = defaultTranslate, toReverseOffsetDateTime = defaultToReverseOffsetDateTime) => {
    const startDate = (0, dateUtils_1.addMonths)((0, dateUtils_1.getNewStartDate)(toLocalDateTime), 1);
    startDate.setDate(1);
    const endDate = (0, dateUtils_1.addMonths)((0, dateUtils_1.getNewStartDate)(toLocalDateTime), 2);
    endDate.setDate(1);
    endDate.setHours(0);
    endDate.setMinutes(0);
    endDate.setSeconds(59);
    const from = toReverseOffsetDateTime(startDate);
    const to = toReverseOffsetDateTime(new Date((0, dateUtils_1.addMinutes)(endDate, -1)));
    return {
        label: translate("Next month"),
        getRange: () => ({
            from,
            to
        })
    };
};
exports.GET_NEXT_MONTH_OPTION = GET_NEXT_MONTH_OPTION;
const GET_LAST_MONTH_OPTION = (toLocalDateTime = defaultToLocalDateTime, translate = defaultTranslate, toReverseOffsetDateTime = defaultToReverseOffsetDateTime) => {
    const startDate = (0, dateUtils_1.addMonths)((0, dateUtils_1.getNewStartDate)(toLocalDateTime), -1);
    startDate.setDate(1);
    const endDate = (0, dateUtils_1.getNewEndDate)(toLocalDateTime);
    endDate.setDate(1);
    endDate.setHours(0);
    endDate.setMinutes(0);
    const from = toReverseOffsetDateTime(startDate);
    const to = toReverseOffsetDateTime(new Date((0, dateUtils_1.addMinutes)(endDate, -1)));
    return {
        label: translate("Last month"),
        getRange: () => ({
            from,
            to
        })
    };
};
exports.GET_LAST_MONTH_OPTION = GET_LAST_MONTH_OPTION;
const GET_LAST_SEVEN_DAYS_OPTION = (toLocalDateTime = defaultToLocalDateTime, translate = defaultTranslate, toReverseOffsetDateTime = defaultToReverseOffsetDateTime) => {
    const startDate = (0, dateUtils_1.addDays)((0, dateUtils_1.getNewStartDate)(toLocalDateTime), -6);
    const endDate = (0, dateUtils_1.getNewEndDate)(toLocalDateTime);
    const from = toReverseOffsetDateTime(startDate);
    const to = toReverseOffsetDateTime(endDate);
    return {
        label: translate("Last 7 days"),
        getRange: () => ({
            from,
            to
        })
    };
};
exports.GET_LAST_SEVEN_DAYS_OPTION = GET_LAST_SEVEN_DAYS_OPTION;
const GET_LAST_THIRTY_DAYS_OPTION = (toLocalDateTime = defaultToLocalDateTime, translate = defaultTranslate, toReverseOffsetDateTime = defaultToReverseOffsetDateTime) => {
    const startDate = (0, dateUtils_1.addDays)((0, dateUtils_1.getNewStartDate)(toLocalDateTime), -29);
    const endDate = (0, dateUtils_1.getNewEndDate)(toLocalDateTime);
    const from = toReverseOffsetDateTime(startDate);
    const to = toReverseOffsetDateTime(endDate);
    return {
        label: translate("Last 30 days"),
        getRange: () => ({
            from,
            to
        })
    };
};
exports.GET_LAST_THIRTY_DAYS_OPTION = GET_LAST_THIRTY_DAYS_OPTION;
const GET_LAST_TWELVE_MONTHS_OPTION = (toLocalDateTime = defaultToLocalDateTime, translate = defaultTranslate, toReverseOffsetDateTime = defaultToReverseOffsetDateTime) => {
    const startDate = (0, dateUtils_1.addMonths)((0, dateUtils_1.getNewStartDate)(toLocalDateTime), -12);
    const endDate = (0, dateUtils_1.getNewEndDate)(toLocalDateTime);
    const from = toReverseOffsetDateTime(startDate);
    const to = toReverseOffsetDateTime(endDate);
    return {
        label: translate("Last 12 months"),
        getRange: () => ({
            from,
            to
        })
    };
};
exports.GET_LAST_TWELVE_MONTHS_OPTION = GET_LAST_TWELVE_MONTHS_OPTION;
const GET_LAST_THREE_MONTH_OPTION = (toLocalDateTime = defaultToLocalDateTime, translate = defaultTranslate, toReverseOffsetDateTime = defaultToReverseOffsetDateTime) => {
    const startDate = (0, dateUtils_1.addMonths)((0, dateUtils_1.getNewStartDate)(toLocalDateTime), -3);
    startDate.setDate(1);
    const endDate = (0, dateUtils_1.getNewEndDate)(toLocalDateTime);
    endDate.setDate(1);
    endDate.setHours(0);
    endDate.setMinutes(0);
    const from = toReverseOffsetDateTime(startDate);
    const to = toReverseOffsetDateTime(new Date((0, dateUtils_1.addMinutes)(endDate, -1)));
    return {
        label: translate("Last 3 months"),
        getRange: () => ({
            from,
            to
        })
    };
};
exports.GET_LAST_THREE_MONTH_OPTION = GET_LAST_THREE_MONTH_OPTION;
const GET_LAST_QUARTER_OPTION = (toLocalDateTime = defaultToLocalDateTime, translate = defaultTranslate, toReverseOffsetDateTime = defaultToReverseOffsetDateTime) => {
    const endDate = (0, dateUtils_1.getNewEndDate)(toLocalDateTime);
    const currentQuarter = Math.floor((endDate.getMonth() + 3) / 3);
    const quarterEndMonth = (currentQuarter - 1) * 3;
    const quarterStartMonth = quarterEndMonth - 3;
    const startDate = new Date(endDate);
    startDate.setMonth(quarterStartMonth);
    startDate.setDate(1);
    startDate.setHours(0);
    startDate.setMinutes(0);
    startDate.setSeconds(0);
    endDate.setMonth(quarterEndMonth);
    endDate.setDate(1);
    endDate.setHours(0);
    endDate.setMinutes(0);
    endDate.setMinutes(-1);
    const from = toReverseOffsetDateTime(startDate);
    const to = toReverseOffsetDateTime(endDate);
    return {
        label: translate("Last quarter"),
        getRange: () => ({
            from,
            to
        })
    };
};
exports.GET_LAST_QUARTER_OPTION = GET_LAST_QUARTER_OPTION;
const GET_THIS_QUARTER_OPTION = (toLocalDateTime = defaultToLocalDateTime, translate = defaultTranslate, toReverseOffsetDateTime = defaultToReverseOffsetDateTime) => {
    const startDate = (0, dateUtils_1.getNewStartDate)(toLocalDateTime);
    const currentQuarter = Math.floor((startDate.getMonth() + 3) / 3);
    const quarterStartMonth = (currentQuarter - 1) * 3;
    startDate.setMonth(quarterStartMonth);
    startDate.setDate(1);
    startDate.setHours(0);
    startDate.setMinutes(0);
    const endDate = new Date(startDate);
    endDate.setMonth(quarterStartMonth + 3);
    endDate.setDate(0);
    endDate.setSeconds(59);
    const from = toReverseOffsetDateTime(startDate);
    const to = toReverseOffsetDateTime(new Date((0, dateUtils_1.addMinutes)(endDate, 23 * 60 + 59)));
    return {
        label: translate("This quarter"),
        getRange: () => ({
            from,
            to
        })
    };
};
exports.GET_THIS_QUARTER_OPTION = GET_THIS_QUARTER_OPTION;
const GET_LAST_YEAR_OPTION = (toLocalDateTime = defaultToLocalDateTime, translate = defaultTranslate, toReverseOffsetDateTime = defaultToReverseOffsetDateTime) => {
    const startDate = (0, dateUtils_1.getNewStartDate)(toLocalDateTime);
    startDate.setFullYear(startDate.getFullYear() - 1);
    startDate.setMonth(0);
    startDate.setDate(1);
    startDate.setHours(0);
    startDate.setMinutes(0);
    const endDate = new Date(startDate);
    endDate.setMonth(11);
    endDate.setDate(31);
    endDate.setSeconds(59);
    const from = toReverseOffsetDateTime(startDate);
    const to = toReverseOffsetDateTime(new Date((0, dateUtils_1.addMinutes)(endDate, 23 * 60 + 59)));
    return {
        label: translate("Last year"),
        getRange: () => ({
            from,
            to
        })
    };
};
exports.GET_LAST_YEAR_OPTION = GET_LAST_YEAR_OPTION;
const GET_THIS_YEAR_OPTION = (toLocalDateTime = defaultToLocalDateTime, translate = defaultTranslate, toReverseOffsetDateTime = defaultToReverseOffsetDateTime) => {
    const startDate = (0, dateUtils_1.getNewStartDate)(toLocalDateTime);
    startDate.setMonth(0);
    startDate.setDate(1);
    startDate.setHours(0);
    startDate.setMinutes(0);
    const endDate = (0, dateUtils_1.getNewEndDate)(toLocalDateTime);
    endDate.setMonth(11);
    endDate.setDate(31);
    endDate.setHours(23);
    endDate.setMinutes(59);
    endDate.setSeconds(59);
    const from = toReverseOffsetDateTime(startDate);
    const to = toReverseOffsetDateTime(endDate);
    return {
        label: translate("This year"),
        getRange: () => ({
            from,
            to
        })
    };
};
exports.GET_THIS_YEAR_OPTION = GET_THIS_YEAR_OPTION;
const GET_DEFAULT_DATE_OPTIONS = (toLocalDateTime, translate, firstDayOfWeek, toReverseOffsetDateTime) => [
    (0, exports.GET_TODAY_OPTION)(toLocalDateTime, translate, toReverseOffsetDateTime),
    (0, exports.GET_YESTERDAY_OPTION)(toLocalDateTime, translate, toReverseOffsetDateTime),
    (0, exports.GET_THIS_WEEK_OPTION)(toLocalDateTime, translate, firstDayOfWeek, toReverseOffsetDateTime),
    (0, exports.GET_LAST_WEEK_OPTION)(toLocalDateTime, translate, firstDayOfWeek, toReverseOffsetDateTime),
    (0, exports.GET_THIS_MONTH_OPTION)(toLocalDateTime, translate, toReverseOffsetDateTime),
    (0, exports.GET_LAST_MONTH_OPTION)(toLocalDateTime, translate, toReverseOffsetDateTime),
    (0, exports.GET_LAST_THREE_MONTH_OPTION)(toLocalDateTime, translate, toReverseOffsetDateTime)
];
exports.GET_DEFAULT_DATE_OPTIONS = GET_DEFAULT_DATE_OPTIONS;
