// @geotab/zenith/dist/table/sortable/sortableStorage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SortableColumnsStorage = void 0;
const STORAGE_PREFIX = "columnsSort_";
class SortableColumnsStorage {
    constructor(pageName, defaultSortState) {
        Object.defineProperty(this, "defaultSortState", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: defaultSortState
        });
        Object.defineProperty(this, "storeKey", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "storage", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.storage = localStorage;
        this.storeKey = `${STORAGE_PREFIX}${pageName}`;
    }
    getItem() {
        const settings = this.storage.getItem(this.storeKey);
        if (!settings) {
            return this.defaultSortState;
        }
        try {
            return JSON.parse(settings);
        }
        catch (e) {
            return this.defaultSortState;
        }
    }
    setItem(s) {
        this.storage.setItem(this.storeKey, JSON.stringify(s));
    }
    clear() {
        this.storage.removeItem(this.storeKey);
    }
}
exports.SortableColumnsStorage = SortableColumnsStorage;
