// @geotab/zenith/dist/dataGrid/columns/mainColumn/mainColumnWrapper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MainColumnWrapper = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const classNames_1 = require("../../../commonHelpers/classNames/classNames");
const primaryIcon_1 = require("../../../primaryIcon/primaryIcon");
class MainColumnWrapper {
    renderCell(elt, entity, isNested, isFeed) {
        var _a, _b, _c, _d, _e, _f, _g;
        const SideItems = this.options.render ? this.options.render(entity) : null;
        const description = this.options.descriptionText ? this.options.descriptionText.createDescription(entity, isNested) : null;
        const mainTextLink = (_b = (_a = this.options.mainText) === null || _a === void 0 ? void 0 : _a.createLink) === null || _b === void 0 ? void 0 : _b.call(_a, entity, isNested);
        const mainTextClassName = (0, classNames_1.classNames)(["zen-main-column__name", ((_c = this.options.mainText) === null || _c === void 0 ? void 0 : _c.className) || "", mainTextLink ? "zen-main-column__name--link" : ""]);
        const descriptionTextClassName = (0, classNames_1.classNames)(["zen-main-column__groups", ((_d = this.options.descriptionText) === null || _d === void 0 ? void 0 : _d.className) || ""]);
        const mainText = ((_f = (_e = this.options.mainText) === null || _e === void 0 ? void 0 : _e.createText) === null || _f === void 0 ? void 0 : _f.call(_e, entity, isNested)) || elt;
        const icon = !isFeed ? this.renderIcon(entity, isNested || false) : null;
        const iconParentClassName = (_g = this.options.icon) === null || _g === void 0 ? void 0 : _g.parentClassName;
        return (0, jsx_runtime_1.jsxs)("div", { className: "zen-main-column", children: [icon ? (0, jsx_runtime_1.jsx)("div", { className: (0, classNames_1.classNames)(["zen-main-column__icon", iconParentClassName || ""]), children: icon }) : null, (0, jsx_runtime_1.jsxs)("div", { className: "zen-main-column__content", children: [mainTextLink ?
                            (0, jsx_runtime_1.jsx)("a", { href: mainTextLink, className: mainTextClassName, children: mainText }) :
                            (0, jsx_runtime_1.jsx)("div", { className: mainTextClassName, children: mainText }), description && (0, jsx_runtime_1.jsx)("div", { className: descriptionTextClassName, children: description })] }), SideItems] });
    }
    constructor(options) {
        Object.defineProperty(this, "options", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: options
        });
    }
    render(elt, entity, isNested) {
        if (!entity) {
            return null;
        }
        return this.renderCell(elt, entity, isNested || false, false);
    }
    renderFeed(elt, entity, isNested) {
        if (!entity) {
            return null;
        }
        return this.renderCell(elt, entity, isNested || false, true);
    }
    renderHeader(title) {
        return title;
    }
    renderIcon(entity, isNested) {
        var _a, _b, _c, _d, _e, _f, _g;
        const Icon = this.options.icon ? this.options.icon.createIcon(entity, isNested) : null;
        if (!Icon) {
            return null;
        }
        const entityActions = ((_a = this.options.icon) === null || _a === void 0 ? void 0 : _a.condition) === undefined || this.options.icon.condition(entity, isNested);
        const isDisabled = !entityActions;
        const onClick = (_b = this.options.icon) === null || _b === void 0 ? void 0 : _b.onClick;
        const handleOnClick = onClick ? (e) => {
            onClick(entity, e);
        } : undefined;
        const link = (_d = (_c = this.options.icon) === null || _c === void 0 ? void 0 : _c.createLink) === null || _d === void 0 ? void 0 : _d.call(_c, entity, isNested);
        const title = (_f = (_e = this.options.icon) === null || _e === void 0 ? void 0 : _e.createTitle) === null || _f === void 0 ? void 0 : _f.call(_e, entity, isNested);
        const iconClassName = (_g = this.options.icon) === null || _g === void 0 ? void 0 : _g.className;
        return (0, jsx_runtime_1.jsx)(primaryIcon_1.PrimaryIcon, { icon: Icon, className: iconClassName, style: isNested ? "secondary" : undefined, onClick: handleOnClick, link: link, title: title, disabled: isDisabled });
    }
}
exports.MainColumnWrapper = MainColumnWrapper;
