// @geotab/zenith/dist/dataGrid/columns/mainColumn/mainColumn.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MainColumn = void 0;
const basicColumn_1 = require("../basicColumn");
const mainColumnWrapper_1 = require("./mainColumnWrapper");
class MainColumn extends basicColumn_1.BasicColumn {
    renderCell(elt, entity, isNested) {
        return this.wrapper.render(elt, entity, isNested);
    }
    renderHeaderCell(title) {
        if (this.isMobile) {
            return "";
        }
        return this.wrapper.renderHeader(title);
    }
    constructor(columnId, isMobile, options, renderHeader) {
        const renderTextDefault = (entity) => entity[this.columnId];
        const renderHeaderDefault = (title) => title;
        super({
            render: renderTextDefault,
            renderHeader: renderHeader ? renderHeader : renderHeaderDefault
        });
        Object.defineProperty(this, "columnId", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: columnId
        });
        Object.defineProperty(this, "isMobile", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: isMobile
        });
        Object.defineProperty(this, "wrapper", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.wrapper = new mainColumnWrapper_1.MainColumnWrapper(options);
    }
    renderFeed(_, entity, isNested) {
        const value = entity[this.columnId];
        return [null, this.wrapper.renderFeed(value, entity, isNested)];
    }
    renderIcon(entity, isNested) {
        return this.wrapper.renderIcon(entity, isNested);
    }
}
exports.MainColumn = MainColumn;
