// @geotab/zenith/dist/dataGrid/withSortableColumns/sortableColumnsStorage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SortableColumnsStorage = void 0;
const STOREAGE_PREFIX = "columnsSort_";
const createFakeSettings = () => {
    const settings = new Map();
    return {
        getItem: (key) => settings.get(key) || null,
        setItem: (key, s) => settings.set(key, s),
        removeItem: (key) => settings.delete(key)
    };
};
class SortableColumnsStorage {
    constructor(pageName, storage, defaultSortState) {
        Object.defineProperty(this, "storeKey", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "storage", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.storage = storage || createFakeSettings();
        this.storeKey = `${STOREAGE_PREFIX}${pageName}`;
        defaultSortState && this.setItem(defaultSortState);
    }
    getItem() {
        const settings = this.storage.getItem(this.storeKey);
        try {
            return (settings ? JSON.parse(settings) : {});
        }
        catch (e) {
            this.storage.setItem(this.storeKey, "[]");
            return {};
        }
    }
    setItem(s) {
        this.storage.setItem(this.storeKey, JSON.stringify(s));
    }
    clear() {
        this.storage.removeItem(this.storeKey);
    }
}
exports.SortableColumnsStorage = SortableColumnsStorage;
