// @geotab/zenith/dist/searchList/searchListHelpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TRANSLATIONS = exports.addAdditionalNavigation = exports.getAdditionalNavigation = exports.allInSuffix = exports.getItemsWithChildren = exports.getParents = exports.getChildren = exports.removeCheckedProp = exports.addParentLinks = exports.searchTree = void 0;
function searchTree(element, elementId) {
    let node = null;
    if (element.id === elementId) {
        return element;
    }
    if (element.children) {
        for (let i = 0; i < element.children.length; i++) {
            node = node ? node : searchTree(element.children[i], elementId);
        }
    }
    return node;
}
exports.searchTree = searchTree;
function addParentLinks(tree, parent) {
    for (const node of tree) {
        if (!node.parent) {
            node.parent = parent;
        }
        if (node.children) {
            addParentLinks(node.children, node);
        }
    }
}
exports.addParentLinks = addParentLinks;
function removeCheckedProp(tree) {
    for (const node of tree) {
        node.checked = false;
        if (node.children) {
            removeCheckedProp(node.children);
        }
    }
}
exports.removeCheckedProp = removeCheckedProp;
const getChildren = (tree, result = []) => {
    for (const node of tree) {
        result.push(node.id);
        if (node.children) {
            (0, exports.getChildren)(node.children, result);
        }
    }
    return result;
};
exports.getChildren = getChildren;
const getParents = (node, result = []) => {
    if (node.parent) {
        result.push(node.parent.id);
        (0, exports.getParents)(node.parent, result);
    }
    return result;
};
exports.getParents = getParents;
const getItemsWithChildren = (tree, ids, result = []) => {
    for (const node of tree) {
        if (ids.some(id => id === node.id) && node.children && node.children.length !== 0) {
            result.push(node.id);
        }
        node.children && (0, exports.getItemsWithChildren)(node.children, result);
    }
    return result;
};
exports.getItemsWithChildren = getItemsWithChildren;
exports.allInSuffix = "allIn";
const getAdditionalNavigation = (id, translate) => [{
        id: `levelUp${id || ""}`,
        icon: "zen-lvl-up",
        path: `${translate("Go back to")} ${id}`,
        name: translate("Level up")
    },
    {
        id: `${exports.allInSuffix}${id || ""}`,
        icon: "zen-circle-ok",
        path: `${translate("All in")} ${id}`,
        name: `${translate("All in")} ${id || ""}`
    }];
exports.getAdditionalNavigation = getAdditionalNavigation;
const addAdditionalNavigation = (tree, translate, path) => {
    for (const node of tree) {
        if (node.children && node.children.length && node.children[0].id.indexOf("levelUp") === -1) {
            node.children = [...(0, exports.getAdditionalNavigation)(node.id || "", translate), ...node.children];
            (0, exports.addAdditionalNavigation)(node.children, translate, `${path ? "/" + path : path || ""}/${node.id}`);
        }
    }
};
exports.addAdditionalNavigation = addAdditionalNavigation;
exports.TRANSLATIONS = [
    "All in", "Go back to", "Level up"
];
