// @geotab/zenith/dist/commonHelpers/arrowNavigation/arrowNavigationHelper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.scrollElement = exports.extractButtonAction = exports.getScrollOptions = exports.clearTimerTimeout = exports.clearTimerInterval = void 0;
const clearTimerInterval = (timerId) => {
    if (timerId !== null) {
        clearInterval(timerId);
    }
};
exports.clearTimerInterval = clearTimerInterval;
const clearTimerTimeout = (timerId) => {
    if (timerId !== null) {
        clearTimeout(timerId);
    }
};
exports.clearTimerTimeout = clearTimerTimeout;
const getScrollOptions = (action, element) => {
    const step = (element === null || element === void 0 ? void 0 : element.offsetWidth) ? element.offsetWidth : 100;
    const scrollTop = (element === null || element === void 0 ? void 0 : element.scrollTop) || 0;
    let scrollLeft = (element === null || element === void 0 ? void 0 : element.scrollLeft) || 0;
    if (action === "left") {
        scrollLeft = scrollLeft - step > 0 ? scrollLeft - step : 0;
    }
    else if (action === "right") {
        const maxWidth = (element === null || element === void 0 ? void 0 : element.scrollWidth) || 0;
        scrollLeft = scrollLeft + step > maxWidth ? maxWidth : scrollLeft + step;
    }
    return {
        top: scrollTop,
        left: scrollLeft
    };
};
exports.getScrollOptions = getScrollOptions;
const extractButtonAction = (event) => {
    const action = event.currentTarget.dataset.action;
    if (!action) {
        return null;
    }
    return action;
};
exports.extractButtonAction = extractButtonAction;
const scrollElement = (element, options) => {
    element.scroll(Object.assign(Object.assign({}, options), { behavior: "smooth" }));
};
exports.scrollElement = scrollElement;
