// @geotab/zenith/dist/selectList/selectList.helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getInitialState = exports.getItemsState = exports.getFirstActiveIndex = exports.getPrevIndex = exports.getNextIndex = exports.getLastIndex = exports.getFirstIndex = exports.searchItemIndex = exports.updateItemState = exports.getItemIndexById = exports.getItemIdByIndex = void 0;
const checkboxState_1 = require("../checkbox/checkboxState");
const reactHelpers_1 = require("../commonHelpers/reactHelpers/reactHelpers");
const getItemIdByIndex = (state, index) => [...state.keys()][index];
exports.getItemIdByIndex = getItemIdByIndex;
const getItemIndexById = (state, id) => [...state.keys()].indexOf(id);
exports.getItemIndexById = getItemIndexById;
const updateItemState = (initialState, multiselect, updatedId, newState) => {
    const temp = new Map(initialState);
    let isUpdated = false;
    if (!temp.has(updatedId)) {
        return [temp, isUpdated];
    }
    let prevState = temp.get(updatedId);
    isUpdated = prevState !== newState;
    temp.set(updatedId, newState);
    if (!multiselect) {
        let onItem = newState !== checkboxState_1.CheckboxState.Off && updatedId;
        temp.forEach((itemState, itemId) => {
            if (itemState !== checkboxState_1.CheckboxState.Off && !onItem) {
                onItem = itemId;
            }
            if (itemId !== onItem) {
                prevState = temp.get(itemId);
                isUpdated = isUpdated || prevState !== checkboxState_1.CheckboxState.Off;
                temp.set(itemId, checkboxState_1.CheckboxState.Off);
            }
        });
    }
    return [temp, isUpdated];
};
exports.updateItemState = updateItemState;
const searchItemIndex = (items, searchStr) => {
    if (!searchStr) {
        return -1;
    }
    return items.findIndex(item => {
        if (item.disabled) {
            return false;
        }
        const content = (0, reactHelpers_1.getComponentText)(item.children);
        return content.toLocaleLowerCase().indexOf(searchStr.toLocaleLowerCase()) > -1;
    });
};
exports.searchItemIndex = searchItemIndex;
const getFirstIndex = (items) => items.findIndex(i => !i.disabled);
exports.getFirstIndex = getFirstIndex;
const getLastIndex = (items) => {
    const reverseLastIndex = [...items].reverse().findIndex(i => !i.disabled);
    return reverseLastIndex === -1 ? reverseLastIndex : items.length - reverseLastIndex - 1;
};
exports.getLastIndex = getLastIndex;
const getNextIndex = (currIndex, items) => {
    const next = (0, exports.getFirstIndex)(items.slice(currIndex + 1));
    return next === -1 ? (0, exports.getFirstIndex)(items) : next + currIndex + 1;
};
exports.getNextIndex = getNextIndex;
const getPrevIndex = (currIndex, items) => {
    const prev = (0, exports.getLastIndex)(items.slice(0, currIndex));
    return prev === -1 ? (0, exports.getLastIndex)(items) : prev;
};
exports.getPrevIndex = getPrevIndex;
const getFirstActiveIndex = (state, items, multiselect) => {
    let firstActiveItemIndex = -1;
    for (let i = 0; i < items.length; i++) {
        const item = items[i];
        if (!item.disabled && firstActiveItemIndex === -1) {
            firstActiveItemIndex = i;
        }
        if (state.get(item.id) === checkboxState_1.CheckboxState.On && !item.disabled) {
            return i;
        }
    }
    return multiselect ? firstActiveItemIndex : -1;
};
exports.getFirstActiveIndex = getFirstActiveIndex;
const getItemsState = (items, value, multiselect) => {
    let forceUncheck = false;
    return items.reduce((res, item) => {
        const itemState = value.get(item.id);
        if (!itemState || itemState === checkboxState_1.CheckboxState.Off || forceUncheck) {
            return res.set(item.id, checkboxState_1.CheckboxState.Off);
        }
        if (!multiselect) {
            forceUncheck = true;
        }
        return res.set(item.id, itemState);
    }, new Map());
};
exports.getItemsState = getItemsState;
const getInitialState = ({ items, value = new Map([]), multiselect }) => ({
    itemsState: (0, exports.getItemsState)(items, value, multiselect),
    isItemsStateChanged: false,
    activeIndex: -1,
    searchStr: "",
    level: ""
});
exports.getInitialState = getInitialState;
