// @geotab/zenith/dist/selectList/selectList.actionTypes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SelectDirection = exports.ListAction = void 0;
var ListAction;
(function (ListAction) {
    ListAction["ChangeItemState"] = "changeItemState";
    ListAction["ToggleCurrent"] = "toggleCurrent";
    ListAction["FocusFirst"] = "focusFirst";
    ListAction["FocusLast"] = "focusLast";
    ListAction["FocusNext"] = "focusNext";
    ListAction["FocusPrev"] = "focusPrev";
    ListAction["FocusNextPage"] = "focusNextPage";
    ListAction["FocusPrevPage"] = "focusPrevPage";
    ListAction["SelectFirst"] = "selectFirst";
    ListAction["SelectLast"] = "selectLast";
    ListAction["SelectNext"] = "selectNext";
    ListAction["SelectPrev"] = "selectPrev";
    ListAction["SearchItem"] = "searchItem";
    ListAction["SelectAllToTheBeginning"] = "selectAllToTheBeginning";
    ListAction["SelectAllToTheEnd"] = "selectAllToTheEnd";
    ListAction["ToggleAll"] = "toggleAll";
    ListAction["Focus"] = "focus";
    ListAction["Blur"] = "blur";
    ListAction["Reset"] = "reset";
    ListAction["SetLevel"] = "level";
})(ListAction || (exports.ListAction = ListAction = {}));
var SelectDirection;
(function (SelectDirection) {
    SelectDirection["Up"] = "Up";
    SelectDirection["Down"] = "Down";
})(SelectDirection || (exports.SelectDirection = SelectDirection = {}));
