// @geotab/zenith/dist/selectList/selectList.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PassiveSelectList = exports.SelectList = exports.defaultListItemHeight = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const checkboxState_1 = require("../checkbox/checkboxState");
const classNames_1 = require("../commonHelpers/classNames/classNames");
const keyboard_1 = require("../commonHelpers/keyboard");
const selectList_actionTypes_1 = require("./selectList.actionTypes");
const selectList_helpers_1 = require("./selectList.helpers");
const selectList_reducer_1 = require("./selectList.reducer");
const selectListItem_1 = require("./selectListItem");
const isChildOf_1 = require("../utils/isChildOf");
const useDriveClassName_1 = require("../utils/theme/useDriveClassName");
const react_2 = require("react");
const getScrollableParent_1 = require("../utils/getScrollableParent");
exports.defaultListItemHeight = 40;
const MemoizedSelectListItem = (0, react_2.memo)(selectListItem_1.SelectListItem);
const SelectListInnerComp = (0, react_2.forwardRef)(({ id, items, value, disabled, activeIndex = -1, title, multiselect = false, showCheckbox = false, standalone, className = "", onClickItem, onChangeItem, onKeyDown, onFocus, onBlur, style = {} }, listRef) => {
    var _a;
    const driveComponentClass = (0, useDriveClassName_1.useDriveClassName)("zen-select-list");
    const state = (0, selectList_helpers_1.getInitialState)({ items, value, multiselect });
    state.activeIndex = activeIndex;
    return (0, jsx_runtime_1.jsx)("ul", { id: id, ref: listRef, className: (0, classNames_1.classNames)(["zen-select-list", driveComponentClass || "", className]), role: "listbox", "aria-multiselectable": multiselect, "aria-activedescendant": standalone ? ((0, selectList_helpers_1.getItemIdByIndex)(state.itemsState, state.activeIndex) || ((_a = items.find(el => !el.disabled)) === null || _a === void 0 ? void 0 : _a.id)) : undefined, "aria-label": title, "aria-disabled": disabled, tabIndex: standalone && !disabled ? 0 : -1, onKeyDown: onKeyDown, onFocus: onFocus, onBlur: onBlur, style: style, children: items.map((item, index) => (0, react_1.createElement)(MemoizedSelectListItem, Object.assign({}, item, { className: (0, classNames_1.classNames)(["zen-select-list__item", item.className || ""]), key: item.id, showCheckbox: showCheckbox, state: state.itemsState.get(item.id), focused: state.activeIndex === index, onChange: onChangeItem, onClick: onClickItem }), item.children)) });
});
SelectListInnerComp.displayName = "SelectListInnerComp";
const SelectListInner = (0, react_2.memo)(SelectListInnerComp);
const SelectList = (_a) => {
    var { items: outerItems, value = new Map(), multiselect = false, selectItemOnFocus = true, forceSelection = false, disabled, onChange, onClick } = _a, rest = __rest(_a, ["items", "value", "multiselect", "selectItemOnFocus", "forceSelection", "disabled", "onChange", "onClick"]);
    const items = (0, react_2.useMemo)(() => outerItems, [outerItems]);
    const listRef = (0, react_2.useRef)(null);
    const [state, dispatchState] = (0, react_2.useReducer)(selectList_reducer_1.selectListReducer, { items, value, multiselect }, selectList_helpers_1.getInitialState);
    const innerValue = (0, react_2.useMemo)(() => JSON.stringify(Array.from((0, selectList_helpers_1.getInitialState)({ items, value, multiselect }).itemsState)), [items, value, multiselect]);
    (0, react_2.useEffect)(() => {
        const parsedValue = JSON.parse(innerValue);
        dispatchState({ type: selectList_actionTypes_1.ListAction.Reset, payload: { items, value: new Map(parsedValue), multiselect } });
    }, [innerValue, multiselect, items]);
    (0, react_2.useEffect)(() => {
        if (!multiselect) {
            dispatchState({ type: selectList_actionTypes_1.ListAction.FocusFirst, payload: { items } });
        }
        dispatchState({ type: selectList_actionTypes_1.ListAction.Focus, payload: { items, multiselect } });
    }, [multiselect, items]);
    (0, react_2.useEffect)(() => {
        onChange && state.isItemsStateChanged && onChange(new Map(state.itemsState));
    }, [state.isItemsStateChanged, state.itemsState, onChange]);
    const onChangeItem = (0, react_2.useCallback)((updatedId, updatedState) => {
        dispatchState({ type: selectList_actionTypes_1.ListAction.ChangeItemState, payload: { updatedId, updatedState, multiselect } });
    }, [multiselect]);
    (0, react_2.useEffect)(() => {
        if (state.activeIndex >= 0 && document.activeElement && listRef.current && (0, isChildOf_1.isChildOf)(document.activeElement, listRef.current)) {
            const itemEl = listRef.current.querySelector(`.zen-select-list__item:nth-child(${state.activeIndex + 1})`);
            itemEl === null || itemEl === void 0 ? void 0 : itemEl.scrollIntoView({ block: "nearest" });
        }
    }, [state.activeIndex]);
    (0, react_2.useEffect)(() => {
        if (!state.level || !onClick) {
            return;
        }
        onClick(state.level);
        dispatchState({ type: selectList_actionTypes_1.ListAction.SetLevel, payload: undefined });
    }, [onClick, state.level]);
    const onKeyDown = (e) => {
        const handlers = {
            "ArrowUp": () => multiselect || !selectItemOnFocus ? dispatchState({ type: selectList_actionTypes_1.ListAction.FocusPrev, payload: { items } }) : dispatchState({ type: selectList_actionTypes_1.ListAction.SelectPrev, payload: { items, multiselect } }),
            "Shift+Tab": () => multiselect || !selectItemOnFocus ? dispatchState({ type: selectList_actionTypes_1.ListAction.FocusPrev, payload: { items } }) : dispatchState({ type: selectList_actionTypes_1.ListAction.SelectPrev, payload: { items, multiselect } }),
            "ArrowDown": () => multiselect || !selectItemOnFocus ? dispatchState({ type: selectList_actionTypes_1.ListAction.FocusNext, payload: { items } }) : dispatchState({ type: selectList_actionTypes_1.ListAction.SelectNext, payload: { items, multiselect } }),
            "Tab": () => multiselect || !selectItemOnFocus ? dispatchState({ type: selectList_actionTypes_1.ListAction.FocusNext, payload: { items } }) : dispatchState({ type: selectList_actionTypes_1.ListAction.SelectNext, payload: { items, multiselect } }),
            "Enter": () => {
                const item = items[state.activeIndex];
                // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
                if (!item) {
                    return;
                }
                if (item.multiLevel) {
                    dispatchState({ type: selectList_actionTypes_1.ListAction.SetLevel, payload: { items } });
                }
                else {
                    dispatchState({ type: selectList_actionTypes_1.ListAction.ChangeItemState, payload: {
                            updatedId: item.id,
                            updatedState: item.state === checkboxState_1.CheckboxState.On ? checkboxState_1.CheckboxState.Off : checkboxState_1.CheckboxState.On,
                            multiselect: multiselect
                        } });
                }
            },
            "Home": () => multiselect || !selectItemOnFocus ? dispatchState({ type: selectList_actionTypes_1.ListAction.FocusFirst, payload: { items } }) : dispatchState({ type: selectList_actionTypes_1.ListAction.SelectFirst, payload: { items, multiselect } }),
            "End": () => multiselect || !selectItemOnFocus ? dispatchState({ type: selectList_actionTypes_1.ListAction.FocusLast, payload: { items } }) : dispatchState({ type: selectList_actionTypes_1.ListAction.SelectLast, payload: { items, multiselect } }),
            "Space": () => dispatchState({ type: selectList_actionTypes_1.ListAction.ToggleCurrent, payload: { multiselect } }),
            "Shift+ArrowUp": () => multiselect && dispatchState({ type: selectList_actionTypes_1.ListAction.SelectPrev, payload: { items, multiselect } }),
            "Shift+ArrowDown": () => multiselect && dispatchState({ type: selectList_actionTypes_1.ListAction.SelectNext, payload: { items, multiselect } }),
            "Ctrl+Shift+Home": () => multiselect && dispatchState({ type: selectList_actionTypes_1.ListAction.SelectAllToTheBeginning, payload: { items, multiselect } }),
            "Ctrl+Shift+End": () => multiselect && dispatchState({ type: selectList_actionTypes_1.ListAction.SelectAllToTheEnd, payload: { items, multiselect } }),
            "Ctrl+KeyA": () => multiselect && dispatchState({ type: selectList_actionTypes_1.ListAction.ToggleAll, payload: { items, multiselect } }),
            "default": () => dispatchState({ type: selectList_actionTypes_1.ListAction.SearchItem, payload: { items, key: e.key } })
        };
        const handlerName = (0, keyboard_1.getKeysPressed)(e);
        const handler = handlers[handlerName] || ((0, keyboard_1.isKeyForSearch)(e.code) && handlers["default"]);
        if (handler) {
            handler();
            e.preventDefault();
            e.stopPropagation();
        }
    };
    const onFocus = () => {
        const scrollableParent = (0, getScrollableParent_1.getScrollableParent)(listRef.current);
        if (!scrollableParent || scrollableParent.scrollTop === 0) {
            dispatchState({ type: selectList_actionTypes_1.ListAction.Focus, payload: { items, multiselect } });
        }
    };
    const onBlur = () => {
        multiselect && dispatchState({ type: selectList_actionTypes_1.ListAction.Blur, payload: { multiselect } });
        !multiselect && forceSelection && dispatchState({ type: selectList_actionTypes_1.ListAction.Blur, payload: { multiselect } });
    };
    return (0, jsx_runtime_1.jsx)(SelectListInner, Object.assign({}, rest, { ref: listRef, value: state.itemsState, items: items, activeIndex: state.activeIndex, multiselect: multiselect, standalone: true, disabled: disabled, onChangeItem: onChangeItem, onClickItem: onClick, onKeyDown: onKeyDown, onFocus: onFocus, onBlur: onBlur }));
};
exports.SelectList = SelectList;
const PassiveSelectList = (props) => (0, jsx_runtime_1.jsx)(SelectListInnerComp, Object.assign({}, props, { standalone: false }));
exports.PassiveSelectList = PassiveSelectList;
