// @geotab/zenith/dist/table/loading/useLoading.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useLoading = exports.DEFAULT_LOADING_ROWS = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const generateId_1 = require("../../commonHelpers/generateId");
const defaultColumnComponent_1 = require("../../dataGrid/extensions/defaultColumnComponent");
const skeleton_1 = require("../../skeleton/skeleton");
exports.DEFAULT_LOADING_ROWS = 50;
const LoadingCell = ({ width, className }) => (0, jsx_runtime_1.jsx)(skeleton_1.Skeleton, { width: width, className: className });
const useLoading = (columns, entities, isLoading, loading) => {
    const sampleEntity = (0, react_1.useMemo)(() => columns.map((column) => ({ [column.id]: column.title })).reduce((acc, el) => (Object.assign(Object.assign({}, acc), el))), [columns]);
    const generateLoadingEntities = (0, react_1.useCallback)(() => (new Array((loading === null || loading === void 0 ? void 0 : loading.rows) || exports.DEFAULT_LOADING_ROWS).fill(undefined)).map(() => (Object.assign(Object.assign({}, sampleEntity), { isLoading: true, id: (0, generateId_1.generateId)() }))), [loading === null || loading === void 0 ? void 0 : loading.rows, sampleEntity]);
    const loadingColumns = (0, react_1.useMemo)(() => columns.map(column => {
        const columnComponent = column.columnComponent || (0, defaultColumnComponent_1.defaultColumnComponent)(column.id);
        // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
        const renderNested = columnComponent.renderNested;
        const renderFeed = columnComponent.renderFeed ? columnComponent.renderFeed.bind(columnComponent) : undefined;
        const renderIcon = ("renderIcon" in columnComponent)
            ? (entity, isNested) => entity.isLoading ? null : columnComponent.renderIcon(entity, isNested)
            : undefined;
        const loadingClassName = "zen-table__cell-loading";
        return Object.assign(Object.assign({}, column), { meta: Object.assign(Object.assign({}, (column.meta || {})), { renderPlaceholder: () => (0, jsx_runtime_1.jsx)(LoadingCell, { className: loadingClassName }), renderFeedPlaceholder: (_) => [
                    (0, jsx_runtime_1.jsx)(LoadingCell, { className: loadingClassName }, `${column.id}_label`),
                    (0, jsx_runtime_1.jsx)(LoadingCell, { className: loadingClassName }, `${column.id}_value`)
                ] }), columnComponent: {
                render: (entity, isNested) => entity.isLoading ? (0, jsx_runtime_1.jsx)(LoadingCell, { className: loadingClassName }) : columnComponent.render(entity, isNested),
                // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
                renderNested: renderNested
                    ? (entity) => entity.isLoading
                        ? (0, jsx_runtime_1.jsx)(LoadingCell, { className: loadingClassName })
                        // eslint-disable-next-line @typescript-eslint/no-explicit-any
                        : renderNested(entity)
                    : undefined,
                renderHeader: columnComponent.renderHeader ? columnComponent.renderHeader.bind(columnComponent) : undefined,
                // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
                renderFeed: renderFeed || renderNested
                    ? (title, entity, isNested) => entity.isLoading
                        ? [
                            (0, jsx_runtime_1.jsx)(LoadingCell, { className: loadingClassName }, `${column.id}_${entity.id}_label`),
                            (0, jsx_runtime_1.jsx)(LoadingCell, { className: loadingClassName }, `${column.id}_${entity.id}_value`)
                        ]
                        : renderFeed
                            ? renderFeed(title, entity, isNested)
                            : [title, isNested
                                    ? renderNested(entity)
                                    : entity[column.id]
                            ]
                    : undefined,
                renderIcon: renderIcon
            } });
    }), [columns]);
    const loadingEntities = (0, react_1.useMemo)(() => {
        if (!isLoading) {
            return entities;
        }
        if (loading === null || loading === void 0 ? void 0 : loading.addToEnd) {
            return [...entities, ...generateLoadingEntities()];
        }
        return generateLoadingEntities();
    }, [entities, generateLoadingEntities, isLoading, loading === null || loading === void 0 ? void 0 : loading.addToEnd]);
    return { loadingColumns, loadingEntities };
};
exports.useLoading = useLoading;
