// @geotab/zenith/dist/gridLayout/hooks/useGridClasses.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useGridClasses = void 0;
const react_1 = require("react");
const classNames_1 = require("../../commonHelpers/classNames/classNames");
const getAllowedComponent_1 = require("../utils/getAllowedComponent");
const updateGridStateFromProps_1 = require("../utils/updateGridStateFromProps");
const updateGridStateFromClassNames_1 = require("../utils/updateGridStateFromClassNames");
const mergeGridStates_1 = require("../utils/mergeGridStates");
const simplifyGridState_1 = require("../utils/simplifyGridState");
const fillGridState_1 = require("../utils/fillGridState");
const adaptGridState_1 = require("../utils/adaptGridState");
const convertStateToClasses_1 = require("../utils/convertStateToClasses");
const adjustGridState_1 = require("../utils/adjustGridState");
const getGridClasses_1 = require("../utils/getGridClasses");
const updateGridStateWithDefaultValues_1 = require("../utils/updateGridStateWithDefaultValues");
const getEmptyGridItemState = () => ({
    gridLg: undefined,
    gridMd: undefined,
    gridSm: undefined,
    gridLgStart: undefined,
    gridMdStart: undefined,
    gridSmStart: undefined,
    lg: undefined,
    md: undefined,
    sm: undefined,
    lgStart: undefined,
    mdStart: undefined,
    smStart: undefined
});
/*
    1. Validation
    2. Convert props to grid item state
    3. Convert className to grid item state
    4. Merge states from props and className
    5. Simplify state: IGridItemFullState -> IGridItemBasicState
    6. Update state with default values (if no props or className provided; example - Card, depended on "size" argument)
    7. Adjust state (example: correct values to be in range of 1 to 12 for desktop)
    8. Fill state (to exclude empty values if possible)
    9. Adapt state (prevent overflowing, corner cases, etc.)
    10. Convert state to className
*/
const useGridClasses = (children, childrenArr) => {
    react_1.Children.forEach(children, (child, indx) => {
        if (!child || typeof child === "string") {
            return;
        }
        const el = (0, getAllowedComponent_1.getAllowedComponent)(child);
        if (!el) {
            return;
        }
        const stateFromProps = getEmptyGridItemState();
        const stateFromClassNames = getEmptyGridItemState();
        const mergedState = getEmptyGridItemState();
        (0, updateGridStateFromProps_1.updateGridStateFromProps)(stateFromProps, el);
        (0, updateGridStateFromClassNames_1.updateGridStateFromClassNames)(stateFromClassNames, el);
        (0, mergeGridStates_1.mergeGridStates)(stateFromProps, stateFromClassNames, mergedState);
        (0, simplifyGridState_1.simplifyGridState)(mergedState);
        (0, updateGridStateWithDefaultValues_1.updateGridStateWithDefaultValues)(el, mergedState);
        (0, adjustGridState_1.adjustGridState)(mergedState);
        (0, fillGridState_1.fillGridState)(mergedState);
        (0, adaptGridState_1.adaptGridState)(mergedState);
        const gridClasses = (0, convertStateToClasses_1.convertStateToClasses)(mergedState);
        const { restClasses } = (0, getGridClasses_1.getGridClasses)(el.props.className || "");
        const itemClassName = (0, classNames_1.classNames)(["zen-grid__item", gridClasses.length ? "zen-grid__item--unset" : "", ...gridClasses, ...restClasses]);
        const clone = (0, react_1.cloneElement)(el, {
            key: el.props.id || el.key || indx,
            className: itemClassName
        });
        const prevElementIndex = childrenArr.findIndex((prevItem) => prevItem.key === clone.key);
        if (prevElementIndex !== -1) {
            childrenArr[prevElementIndex] = clone;
        }
        else {
            childrenArr.push(clone);
        }
    });
};
exports.useGridClasses = useGridClasses;
